{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "49397062",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "import pickle\n",
    "from multiprocessing import Pool"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 184,
   "id": "f668a383",
   "metadata": {},
   "outputs": [],
   "source": [
    "%config Completer.use_jedi = False\n",
    "%config IPCompleter.greedy=True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "0d871c68",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IN-SST_C9ALSvsCTRL',\n",
       " 'L4_C9ALSvsCTRL',\n",
       " 'Inhibitory_sALSvsCTRL',\n",
       " 'L5-6-CC_C9ALSvsCTRL',\n",
       " 'Excitatory_sALSvsCTRL',\n",
       " 'L5-6_C9ALSvsCTRL',\n",
       " 'Excitatory_C9ALSvsCTRL',\n",
       " 'L2-3_sALSvsCTRL',\n",
       " 'AST-FB_sALSvsCTRL',\n",
       " 'Oligodendrocytes_C9ALSvsCTRL',\n",
       " 'sALS_ALL_training_test_data.pkl',\n",
       " 'train_data.npy',\n",
       " 'OPC_C9ALSvsCTRL',\n",
       " 'all_seqs_celltypes_input',\n",
       " 'valid_data.npy',\n",
       " 'REDU_plots',\n",
       " 'tst_train_data.npy',\n",
       " 'IN-VIP_C9ALSvsCTRL',\n",
       " 'L4_sALSvsCTRL',\n",
       " 'AST-PP_C9ALSvsCTRL',\n",
       " 'IN-VIP_sALSvsCTRL',\n",
       " 'AST-PP_sALSvsCTRL',\n",
       " 'L5-6-CC_sALSvsCTRL',\n",
       " 'IN-PV_C9ALSvsCTRL',\n",
       " 'Oligodendrocytes_sALSvsCTRL',\n",
       " 'IN-SST_sALSvsCTRL',\n",
       " 'OPC_sALSvsCTRL',\n",
       " 'L2-3_C9ALSvsCTRL',\n",
       " 'test_data.npy',\n",
       " 'Microglia_sALSvsCTRL',\n",
       " 'Inhibitory_C9ALSvsCTRL',\n",
       " 'C9ALS_ALL_training_test_data.pkl',\n",
       " 'Microglia_C9ALSvsCTRL',\n",
       " 'Endothelial_ALSvsCTRL',\n",
       " 'IN-PV_sALSvsCTRL',\n",
       " 'AST-FB_C9ALSvsCTRL']"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_root = '/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/'\n",
    "os.listdir(data_root)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "67b466ad",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IN-SST_C9ALSvsCTRL',\n",
       " 'L4_C9ALSvsCTRL',\n",
       " 'L5-6-CC_C9ALSvsCTRL',\n",
       " 'L5-6_C9ALSvsCTRL',\n",
       " 'Excitatory_C9ALSvsCTRL',\n",
       " 'Oligodendrocytes_C9ALSvsCTRL',\n",
       " 'OPC_C9ALSvsCTRL',\n",
       " 'IN-VIP_C9ALSvsCTRL',\n",
       " 'AST-PP_C9ALSvsCTRL',\n",
       " 'IN-PV_C9ALSvsCTRL',\n",
       " 'L2-3_C9ALSvsCTRL',\n",
       " 'Inhibitory_C9ALSvsCTRL',\n",
       " 'C9ALS_ALL_training_test_data.pkl',\n",
       " 'Microglia_C9ALSvsCTRL',\n",
       " 'AST-FB_C9ALSvsCTRL']"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## lets focus on C9ALS\n",
    "ct = [ e for e in os.listdir(data_root) if 'C9ALS' in e]\n",
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 224,
   "id": "12b092b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_dict = {}\n",
    "for ct_cn in ct:\n",
    "    inp_fa = data_root + \"/{}/switch_DNA_sequence.fa\".format(ct_cn)\n",
    "    inp_fa = SeqIO.parse(inp_fa, \"fasta\")\n",
    "    for rec in inp_fa:\n",
    "        if rec.id not in sequences_dict:\n",
    "            sequences_dict[rec.id] = str(rec.seq)\n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 225,
   "id": "88b83d28",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sequences_dict = {}\n",
    "for ct_cn in ct:\n",
    "    inp_fa = data_root + \"/{}/switch_DNA_sequence.fa\".format(ct_cn)\n",
    "    inp_fa = SeqIO.parse(inp_fa, \"fasta\")\n",
    "    for rec in inp_fa:\n",
    "        if rec.id not in sequences_dict:\n",
    "            sequences_dict[rec.id] = str(rec.seq)\n",
    "        else:\n",
    "            continue  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 226,
   "id": "b1b34021",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107427"
      ]
     },
     "execution_count": 226,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sequences_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "id": "e4697dfd",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'chr2:AAK1:69461526:69472042:-'"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(sequences_dict.keys())[10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "22f1e6e4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AST-FB',\n",
       " 'AST-PP',\n",
       " 'C9ALS',\n",
       " 'Excitatory',\n",
       " 'IN-PV',\n",
       " 'IN-SST',\n",
       " 'IN-VIP',\n",
       " 'Inhibitory',\n",
       " 'L2-3',\n",
       " 'L4',\n",
       " 'L5-6',\n",
       " 'L5-6-CC',\n",
       " 'Microglia',\n",
       " 'OPC',\n",
       " 'Oligodendrocytes']"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltypes = [e.split('_')[0] for e in ct]\n",
    "celltypes = sorted(celltypes)\n",
    "celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "1898c2a3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C9ALS'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltypes.pop(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "28407869",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AST-FB': 0,\n",
       " 'AST-PP': 1,\n",
       " 'Excitatory': 2,\n",
       " 'IN-PV': 3,\n",
       " 'IN-SST': 4,\n",
       " 'IN-VIP': 5,\n",
       " 'Inhibitory': 6,\n",
       " 'L2-3': 7,\n",
       " 'L4': 8,\n",
       " 'L5-6': 9,\n",
       " 'L5-6-CC': 10,\n",
       " 'Microglia': 11,\n",
       " 'OPC': 12,\n",
       " 'Oligodendrocytes': 13}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltypes_order = {}\n",
    "for e,i in enumerate(celltypes):\n",
    "    celltypes_order[i] = e\n",
    "celltypes_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 230,
   "id": "2b06eb67",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107427"
      ]
     },
     "execution_count": 230,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels = {}\n",
    "for key in sequences_dict.keys():\n",
    "    labels[key] = np.zeros((14,1), dtype=float)\n",
    "len(labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "id": "05a04f37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# def get_lfc_logp(df, name):\n",
    "#     tmp_df =  df.loc[df['switch_name'] == name]\n",
    "#     lfc = round(float(tmp_df['LFC_PA_Usage']),4)\n",
    "#     nlogp = round(float(tmp_df['negative_logFDR']),4)\n",
    "#     return([lfc,nlogp])\n",
    "\n",
    "import scipy.stats as ss\n",
    "\n",
    "def rank_to_normal(rank, c, n):\n",
    "    # Standard quantile function\n",
    "    x = (rank - c) / (n - 2*c + 1)\n",
    "    return ss.norm.ppf(x)\n",
    "\n",
    "def rank_INT(series, c=3.0/8, stochastic=True):\n",
    "    \"\"\" Perform rank-based inverse normal transformation on pandas series.\n",
    "        If stochastic is True ties are given rank randomly, otherwise ties will\n",
    "        share the same value. NaN values are ignored.\n",
    "        Args:\n",
    "            param1 (pandas.Series):   Series of values to transform\n",
    "            param2 (Optional[float]): Constand parameter (Bloms constant)\n",
    "            param3 (Optional[bool]):  Whether to randomise rank of ties\n",
    "        \n",
    "        Returns:\n",
    "            pandas.Series\n",
    "    \"\"\"\n",
    "\n",
    "    # Check input\n",
    "    #assert(isinstance(series, pd.Series))\n",
    "    #assert(isinstance(c, float))\n",
    "    #assert(isinstance(stochastic, bool))\n",
    "\n",
    "    # Set seed\n",
    "    np.random.seed(123)\n",
    "\n",
    "    # Take original series indexes\n",
    "    orig_idx = series.index\n",
    "\n",
    "    # Drop NaNs\n",
    "    series = series.loc[~pd.isnull(series)]\n",
    "\n",
    "    # Get ranks\n",
    "    if stochastic == True:\n",
    "        # Shuffle by index\n",
    "        series = series.loc[np.random.permutation(series.index)]\n",
    "        # Get rank, ties are determined by their position in the series (hence\n",
    "        # why we randomised the series)\n",
    "        rank = ss.rankdata(series, method=\"ordinal\")\n",
    "    else:\n",
    "        # Get rank, ties are averaged\n",
    "        rank = ss.rankdata(series, method=\"average\")\n",
    "\n",
    "    # Convert numpy array back to series\n",
    "    rank = pd.Series(rank, index=series.index)\n",
    "\n",
    "    # Convert rank to normal distribution\n",
    "    transformed = rank.apply(rank_to_normal, c=c, n=len(rank))\n",
    "    \n",
    "    return transformed[orig_idx]\n",
    "\n",
    "def get_sig_lfc(df, name):\n",
    "    tmp_df =  df.loc[df['switch_name'] == name]\n",
    "    lfc = round(float(tmp_df['LFC_rand_INT']),4)\n",
    "    return(lfc)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 235,
   "id": "91f3f5d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aiden/anaconda3/lib/python3.7/site-packages/pandas/core/series.py:1143: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self.loc[key]\n"
     ]
    }
   ],
   "source": [
    "## this will take couple of minutes ~ 10-15 min \n",
    "for ct in celltypes:\n",
    "    df_name = data_root + ct + \"_C9ALSvsCTRL/APAlog_res_metadata_added.tsv\"\n",
    "    inp_df = pd.read_csv(df_name, sep='\\t')\n",
    "    inp_df['LFC_rand_INT'] = rank_INT(inp_df['sig_LFC_PA_Usage'])\n",
    "    ct_idx = celltypes_order[ct]\n",
    "    for key in list(labels.keys()):\n",
    "        if key in inp_df['switch_name'].values:\n",
    "            res = get_sig_lfc(inp_df, key)\n",
    "            labels[key][ct_idx] = res\n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 236,
   "id": "4799048b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr12:AACS:125140928:125143316:+': array([[ 0.    ],\n",
       "        [ 0.3208],\n",
       "        [-1.2717],\n",
       "        [ 0.    ],\n",
       "        [ 0.4777],\n",
       "        [ 0.    ],\n",
       "        [ 0.3739],\n",
       "        [-1.1109],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.4887],\n",
       "        [ 1.0915]]), 'chr4:AADAT:170060271:170060673:-': array([[ 0.    ],\n",
       "        [-0.34  ],\n",
       "        [ 0.    ],\n",
       "        [ 1.3333],\n",
       "        [ 0.869 ],\n",
       "        [ 2.3204],\n",
       "        [ 1.7817],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [-0.7062],\n",
       "        [ 1.6903]]), 'chr2:AAK1:69457997:69461526:-': array([[ 0.    ],\n",
       "        [ 1.724 ],\n",
       "        [-0.1362],\n",
       "        [ 0.9156],\n",
       "        [ 0.6777],\n",
       "        [ 1.4271],\n",
       "        [ 1.0699],\n",
       "        [-0.1143],\n",
       "        [ 0.3483],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 1.4237],\n",
       "        [ 0.4357],\n",
       "        [ 0.6563]]), 'chr2:AAK1:69457997:69464011:-': array([[ 0.    ],\n",
       "        [ 0.7317],\n",
       "        [-0.7626],\n",
       "        [ 0.2778],\n",
       "        [ 0.8448],\n",
       "        [ 1.2505],\n",
       "        [ 1.1759],\n",
       "        [-0.7713],\n",
       "        [-0.0174],\n",
       "        [ 0.    ],\n",
       "        [ 0.    ],\n",
       "        [ 0.723 ],\n",
       "        [ 0.6049],\n",
       "        [ 0.5703]]), 'chr2:AAK1:69457997:69465314:-': array([[0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [1.9975],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ],\n",
       "        [0.    ]])}"
      ]
     },
     "execution_count": 236,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(list(labels.items())[0:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e858ff3",
   "metadata": {},
   "source": [
    "## ok we have the sequences and labels dictionaries.\n",
    "## next lets translate the sequences to the sense RNAs \n",
    "## then make all data dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "3e47bc9d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def transcribe_positive_strand(seq):\n",
    "    \"\"\" input is the 5' to 3' coding squence\n",
    "        so the RNA will be exact sequence except\n",
    "        U instead of T\n",
    "    \"\"\"\n",
    "    return(seq.replace('T','U'))\n",
    "\n",
    "def transcribe_negative_strand(seq):\n",
    "    \"\"\" input is the 5' to 3' template squence\n",
    "        so the function complement and returns\n",
    "        the reverse of sequence\n",
    "    \"\"\"\n",
    "    complement = {'A': 'U', 'C': 'G', 'G': 'C', 'T': 'A'}\n",
    "    return \"\".join(complement.get(base, base) for base in reversed(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 238,
   "id": "5e55f144",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcribed_sequences = {}\n",
    "for key,value in sequences_dict.items():\n",
    "    strand = key.split(':')[-1]\n",
    "    if strand == '+':\n",
    "        transcribed_sequences[key] = transcribe_positive_strand(value)\n",
    "    else:\n",
    "        transcribed_sequences[key] = transcribe_negative_strand(value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 239,
   "id": "fad56b0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "all_data_dict = {}\n",
    "for key,value in transcribed_sequences.items():\n",
    "    all_data_dict[key] = {}\n",
    "    all_data_dict[key][value] = labels[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 240,
   "id": "3dbdbc25",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'chr12:AACS:125140928:125143316:+': {'GUGAGGCGGGACAAACUUGUCUUCCUCACACCCAUCUUACUUCCUCUUAUGAGGAAACCCAGAGAGAUGAGGGGUCUUGCCCAAGGAAGGGGUGUCCAUAGUCAGCUCUGCCUUCUGCUCACCCAGAAUAAAGACCUGGGGACCCCGCGAGGGUCAUGGCCAAGUGGAAUGGACUCCUGGCAUUUGAGGGCUUCCCGACUGCAGCCCUCAGGCAGCCAUGGCUGUCCCAAGUCCAGCGGGCCUUUGCUCGGGUCAUGGCUGGGAUGUCUGGCCCUUCCUGACAGGAGGCUGCUGGGCUCCUGUCUACUUGGGGACGCCUCAUGCAGGAGCUGGUGUGGGGGUGGGCAGGGGGGCGGUGGCUUCUUCCUUUCUCUUUCCCUUUCCUCUACCUUUUCCCCUCUCCCCAGAGGAAAUGGUAGCAGGAUUUCUUUUAAGAGGAUGCUGCUGUAUUUUGCCAGCGGGUGGAAGGUGGCGGUAUUAGCUCCCGUGAGCUGCACGUGGACCCCUGUGUGAAGCGUAGCAGGGCACAGAGCAGGCGAGACGUUUGCAUCUCACAGCGGGAGGGCCGGCGACAUCACAUGAAGUGACAGGCAGGCCCUUGGAAGCCGGUGCUUAGAUCCUUAAUUAGUUCACACGUCGACUGAAUUUUCAAGUGAAUGAAUUUUAAUUACAUCUCAGGUUAAAAAAAAAAAAAGGCGCCAGUGAUCGAGGACUCGUCACUGGGCUCUGUUGCUCCUGAAGUUUCCUAGCCCACAACACACCAACACUGCCAAGGGCUCUUCUGGAUUCAAGGUGAAACACAUGUGCCAUAAAUCUUGGAGCUCUGAAUGUUUGGAAAGGGCCCGACUGUGAGAAGAAGUAACACACCGUCCCGUGCAGAUGGCUGGCUCUGAGGAGGAGUUCAUGGGAGCUUGGGGACACUCUUGCCUCUAGUUCUAGGAAGCUGGGCCACUUCUGAAGUAAUGGCAAUAUCAAUAAAGUAAUGGUCUUUAUCAUAGAAUAACGUGAUAAAAUAUAUAGAGAAGUAAAAAAGUAUAAAUAAAAGUAAAAUCAUCAUAAAACAUAGUAGCUAGGCACUUCUGAAGCUGUGUGUGCACUGAUUCAUUCACCCAGUGACUCACAGCCUUAUAGCCUAGGUGCUGGCACCCCUACUUUCAUUCGAGGAAGUGAACUCAGGUUCAGGAAUUUACCCAGCAUCCCCCAGAUGGGGUGGCAGGAGCCACAUCUUCCCUGAAAACUUUCUUGCCCAGGGUGUCUGCUGGGAUUUAGGAAUGGUCUAUGCCUGCAUUUUUAUCCUGGUCAGGCUGACCCUGAACCCUGAGAGAUACUCUUUUUUAUAUUCCCAUCUGGAAUAUGCACUGCCGGGGUCAGUGGGGUGUCUGGAGGGCCCUCUCGAGGCCAGCUUGGAUGUGACACGUGUCGUGGGUCCCAACGGGGCCCAGUAGAGUGUGCAGCGUUAGAAAAAUGAACAUGCUCGGCUGGGCGCGGUGGCUCACGCCUGUGAUCCUAGCACUUUGGGAGGCCAAGAUGGGUGGAUCAUGAGGUCAGGAGAUCAAGACCAUCCUGGCUAACAUGAGACCAUCCUGGUGAAACCCCAUCUCUACUAAAAAUACAAAAAAUUAGCUGGGCGUGGUGGCAGGUGCCUAUGGUCCCAGCUACUCAGGAGGCUGAGGUAGGAGAAUGGUGUGAACCUGGGAGGGGGAGCUUGCAGUAAGCGGAGAUUGCACCACUGCACUCCAGCCUGGGUGACAGAGUGCGACUCUGUCUCAAAAAAAAAAAAAAGAAAAAAAAAGAAAAAAAGAAAAAUGAACAUGCUCAUGGUGACCUCUGGGCAACUUUGAAAUAGCCUCAGUUUCUCAACAGCCGUGCCAGGGGGCGGGGGUGGGGUGUGGAGAGUGUUAGAGUUGAGUUUCAGCACCAGUCCUUGAAAGUGCCUUUCAGAAGAACAGGUUUUAUCCCCCAAACAGAGAGACAGUGGUCAGUUCACCGUACUUCGGUGCACGUUUGCAGACCUGGUGGGAAGAGGGGCAUCUUAGAGCCGAGACCCAUUCACUCUUGGCACUCCAGGUGGAGCUGGGCCUUUGGGGCCUGGAUAUAUCCAGGGCUGCGGAUUUUCCCCCCUUCAGGUUUAAAUGUUCCUGUUUUUCUACCUUUCCCUCGCAGUAUACGCUCAACGGCAAGAAAGUGGAAGUUGCCGUCAAACAGAUCAUCGCUGGAAAAGCCGUGGAGCAAGGAGGUGCUUUCUCGAACCCCGAGACCCUGGAUCUGUACCGGGACAUCCCUGAGCUGCAGGGCUUCUGAGUCAGACUGGCUGGCGUGUCACUCAGCCGCACCCGUGUGCACUGUAACUUUUGUGUGCUCAAGAAAUUAUACAGAAACCUACAGCUGUUGUAAAAGGAUGCUCGCACCAAGUGUUCUGUAGGCUUGGGGAGGGAUCGUUUCUCUGUUUUGUUAAAUCUGGUGGGUACCUGGAUCUUCCACACGAGUGGGAUUCUGGCCUUCAGAGACCAGGAGGGAGUGUCUGGGCCGCAGGUGUGGCACUGUGGUGAGAGUGUGUGUCUUUGCACACACAGUGCAGCGGGAACGGUGGGGCUGGCUGGUGCUGAAGACAGACACACUCCUGAGCCAAGGUCUUGUCUUCAACCUCCCCGUCCCGUUGUCCCAUUUUGCUCUGUGAAGGUGCAAAUCCCUUUCUUCCCUUCCCAUCUCAGGCUCUCCUGUUUUCCCUCAGGGUCCAGUAUGCCUUUGAGCUUUAGCUGUUAGAAAGGAACCCCCGUGACUUGACACAGCUUUCACAGCUGGCUGCUAGGACCGGCGGGCUGGGUGUUCACGUGUGUCUGUGUCAUGGAUGCAAUGCAGGCCCUGGAGGACUGUGCGUCACCCGUCAACCAGAGCGUGCCUCCGGGCCAGCUUCCCUCCAAGGAAUGAGUGGAUUUCAUACAGGAUCUCUUUAUUGCACAGACUGAAUGGCUUUACAUGUUUCUAAUGUGAAUUAGGCAUGUGAAGCAGUGGGUGUCCACCCGUGUCCCUCAUGGGUGAGCCCUCCAGCUGUGAGCCCAGGCAGUGUGGUCACCGAGUGAGGACCCUCCUCACCAGGAACCGCAUCCCUGUGCUGCCUCCACCUGAGAGUUGCUAGGGGGUUCUUGUCGAGAUCAUGUCAUCAGCACCCCUAAGUCAAGUCACGGGUUUCCAUAGCCAGGCAGUUGGUAUGUACAAUUCAGUUCAGCGUAUGAACUUGUAUCUCUAAUCUGAUGUCCAUUUUUAUAUUUUUUGAAACUGAGCACAAUGAAAUCCUUUCUUGAAUCAUUUUCCUUUUGGAUUAUAAAAAUAUGGGGGAAAGUGCUAUGAUGAAUUUUAUGCAAUAAAUGUAUACAUGUGUGCACAUGCA': array([[ 0.    ],\n",
       "         [ 0.3208],\n",
       "         [-1.2717],\n",
       "         [ 0.    ],\n",
       "         [ 0.4777],\n",
       "         [ 0.    ],\n",
       "         [ 0.3739],\n",
       "         [-1.1109],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [ 0.4887],\n",
       "         [ 1.0915]])},\n",
       " 'chr4:AADAT:170060271:170060673:-': {'UGCAUGAAUAAUGUAUCACAACCUACUGAGCUGCGUAAAUUGUCAUGUUAUGAGGAAAACUUCUUUCCUUCCUUUAUUGAAUUAACAUGAGCCAGAAGUACAAAUUCAUGAUGGCUGUCACUUUCUCUGAAGAAAAAAGUAUCAGCAAGGGUCAAAUAGAAUAAUAUUCUGAAUCACAAUCUAGCUUUAACCAAAAUAAGCUGUGUAUGUGGAAAUCAAACAGAAGAGCUAUAAAUUGAACAGUGCCAUUAACCUUGAGAAGAGUUUAUAUUUAUCCUUGGAAAGGGACAUUUUUGGCAUGUUCUUUAUUUACUUGCUUUCUCAUGUCAUUAAAUGUGUUUUUCAUAAGGUCUCUCUGUUUUAAUGUAUUAGUGGAGUUGGGCAAUUGAAUCUAUAUAAAAUGCAGGUGAGAGAUUAUGUAAAAUCUAAAAUAAUCCAUGUUAUGGAUAAAAUGGAAAAUGAUCUUGAAGGCCACGAGAACAACAACAACAACAAAAUCCUGAAAAUCUGAUGUACUUCAGUCUUCAUAUUUUAAUAGAAAUUGAUGGCAUCCUGCACAGAUAAAGUUGAUGUUAAACUUGACUUUUCAAUGCCUUCUCUCAACCUUAGACAGAUACUCUUUGUUUGGAUUUUUAAAGUUAUUACAGAGUUUUCCAAUAUAGUAUCCUAAUUUAAUUAAUUCUAAUUUUUUUUUUCUGUUUUAGGCCUUCCAGGUAUUAGCACAACUUAUAAAAGAAUCUUUAUGAAGAAAUUAAACUAGGUUGGGCAUGGUGGCUCACACCUAUAAUCCCAGCACUUUGGGAGGCAGAGGAGGGAGGAUCACUUGAACCCAGGAAUUCAAGGCUGCAGUAAGCUACGAUCACACCACUGCACUCUGGCCUGCAUGCACUCUGGCCUGCAUGGCAGAACAAGACCCUGUCUCUAAAAAAAGAGAAAGAAAUCAAACUAAUCAUGCUGCUCAUGGAUUUUUCCAAUAAAUUUCUUGUUUUGGCAGGAAGAAAUGAACACUGGUAUUAGACUUAAAGAUUAAAUUUCCUCAAACAUGUCCUAUCUGUAGUAGUUCAACUAGACACCUUUUAAAGUGCCUCUAAAUUCAUCAGAUGGCCAAACUGUAUUUAUAAUCCACUUAGGCAUUUUGAAAAACUUUCAACCUGUAAAAAGUUACUUUUAUCUUGGAUUUAUUAUGAAGAACUUUGUAGUUGCUUUGUAAUUUCCCAUAAAUUGUCUUUGAAACUAACAUUUUACACUGAAUUAUUUUGAGAUUUUAAAGAAGUAAUUAAGUGCAAAAUGGUAUAUAAUGUGUACUUUUUCUACUUUUAGGAAAAUUUAAUGAGAGCUUAUUGCAAAAAUUGUUAUAAUUUGGUCAUUAUAAGUGACUUUUAGUAAAAGUACCAUAAACCUU': array([[ 0.    ],\n",
       "         [-0.34  ],\n",
       "         [ 0.    ],\n",
       "         [ 1.3333],\n",
       "         [ 0.869 ],\n",
       "         [ 2.3204],\n",
       "         [ 1.7817],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [ 0.    ],\n",
       "         [-0.7062],\n",
       "         [ 1.6903]])}}"
      ]
     },
     "execution_count": 240,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dict(list(all_data_dict.items())[0:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 241,
   "id": "a353e415",
   "metadata": {},
   "outputs": [],
   "source": [
    "outname = data_root + 'C9ALS_ALL_training_test_data.pkl'\n",
    "with open(outname, 'wb') as f:\n",
    "    pickle.dump(all_data_dict, f)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5b75912f",
   "metadata": {},
   "source": [
    "## Okk lets do the same thing for sALS and we are done with data processing :) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "d941d96d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Inhibitory_sALSvsCTRL',\n",
       " 'Excitatory_sALSvsCTRL',\n",
       " 'L2-3_sALSvsCTRL',\n",
       " 'AST-FB_sALSvsCTRL',\n",
       " 'L4_sALSvsCTRL',\n",
       " 'IN-VIP_sALSvsCTRL',\n",
       " 'AST-PP_sALSvsCTRL',\n",
       " 'L5-6-CC_sALSvsCTRL',\n",
       " 'Oligodendrocytes_sALSvsCTRL',\n",
       " 'IN-SST_sALSvsCTRL',\n",
       " 'OPC_sALSvsCTRL',\n",
       " 'Microglia_sALSvsCTRL',\n",
       " 'IN-PV_sALSvsCTRL']"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct = [ e for e in os.listdir(data_root) if 'sALS' in e and 'pkl' not in e]\n",
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "26cc53b7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14"
      ]
     },
     "execution_count": 110,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(ct)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 244,
   "id": "e4ee62da",
   "metadata": {},
   "outputs": [],
   "source": [
    "sequences_dict = {}\n",
    "for ct_cn in ct:\n",
    "    inp_fa = data_root + \"/{}/switch_DNA_sequence.fa\".format(ct_cn)\n",
    "    inp_fa = SeqIO.parse(inp_fa, \"fasta\")\n",
    "    for rec in inp_fa:\n",
    "        if rec.id not in sequences_dict:\n",
    "            sequences_dict[rec.id] = str(rec.seq)\n",
    "        else:\n",
    "            continue  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 245,
   "id": "38fadab2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107834"
      ]
     },
     "execution_count": 245,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sequences_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "9dd34c00",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AST-FB',\n",
       " 'AST-PP',\n",
       " 'Excitatory',\n",
       " 'IN-PV',\n",
       " 'IN-SST',\n",
       " 'IN-VIP',\n",
       " 'Inhibitory',\n",
       " 'L2-3',\n",
       " 'L4',\n",
       " 'L5-6-CC',\n",
       " 'Microglia',\n",
       " 'OPC',\n",
       " 'Oligodendrocytes']"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltypes = [e.split('_')[0] for e in ct]\n",
    "celltypes = sorted(celltypes)\n",
    "celltypes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "id": "23352813",
   "metadata": {},
   "outputs": [],
   "source": [
    "celltypes_order = {'AST-FB': 0,\n",
    " 'AST-PP': 1,\n",
    " 'Excitatory': 2,\n",
    " 'IN-PV': 3,\n",
    " 'IN-SST': 4,\n",
    " 'IN-VIP': 5,\n",
    " 'Inhibitory': 6,\n",
    " 'L2-3': 7,\n",
    " 'L4': 8,\n",
    " 'L5-6': 9,\n",
    " 'L5-6-CC': 10,\n",
    " 'Microglia': 11,\n",
    " 'OPC': 12,\n",
    " 'Oligodendrocytes': 13}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 248,
   "id": "ceb4e7da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107834"
      ]
     },
     "execution_count": 248,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "labels = {}\n",
    "for key in sequences_dict.keys():\n",
    "    labels[key] = np.zeros((13,2), dtype=float)\n",
    "len(labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20b015fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "for ct in celltypes:\n",
    "    df_name = data_root + ct + \"_sALSvsCTRL/APAlog_res_metadata_added.tsv\"\n",
    "    inp_df = pd.read_csv(df_name, sep='\\t')\n",
    "    inp_df['LFC_rand_INT'] = rank_INT(inp_df['sig_LFC_PA_Usage'])\n",
    "    ct_idx = celltypes_order[ct]\n",
    "    for key in list(labels.keys()):\n",
    "        if key in inp_df['switch_name'].values:\n",
    "            res = get_sig_lfc(inp_df, key)\n",
    "            labels[key][ct_idx] = res\n",
    "        else:\n",
    "            continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "099e7167",
   "metadata": {},
   "outputs": [],
   "source": [
    "transcribed_sequences = {}\n",
    "for key,value in sequences_dict.items():\n",
    "    strand = key.split(':')[-1]\n",
    "    if strand == '+':\n",
    "        transcribed_sequences[key] = transcribe_positive_strand(value)\n",
    "    else:\n",
    "        transcribed_sequences[key] = transcribe_negative_strand(value)\n",
    "all_data_dict = {}\n",
    "for key,value in transcribed_sequences.items():\n",
    "    all_data_dict[key] = {}\n",
    "    all_data_dict[key][value] = labels[key]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3da5e6b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "dict(list(all_data_dict.items())[0:2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d9d45c3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "outname = data_root + 'sALS_ALL_training_test_data.pkl'\n",
    "with open(outname, 'wb') as f:\n",
    "    pickle.dump(all_data_dict, f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a89e56e2",
   "metadata": {},
   "outputs": [],
   "source": [
    "len(all_data_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2a5eeaef",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dda87806",
   "metadata": {},
   "source": [
    "## sanity checks :) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "03917611",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(data_root + 'C9ALS_ALL_training_test_data.pkl', 'rb') as inputFile:\n",
    "        C9ALS_data = pickle.load(inputFile)\n",
    "        \n",
    "with open(data_root + 'sALS_ALL_training_test_data.pkl', 'rb') as inputFile:\n",
    "        sALS_data = pickle.load(inputFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "d6e9f9aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107051"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(C9ALS_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "d12b451c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107457"
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(sALS_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "cff57e58",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Run</th>\n",
       "      <th>analyte_type</th>\n",
       "      <th>Assay Type</th>\n",
       "      <th>AvgSpotLen</th>\n",
       "      <th>Bases</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>BioSample</th>\n",
       "      <th>biospecimen_repository</th>\n",
       "      <th>biospecimen_repository_sample_id</th>\n",
       "      <th>body_site</th>\n",
       "      <th>...</th>\n",
       "      <th>study_design</th>\n",
       "      <th>study_name</th>\n",
       "      <th>subject_is_affected</th>\n",
       "      <th>submitted_subject_id</th>\n",
       "      <th>Is_Tumor</th>\n",
       "      <th>sex</th>\n",
       "      <th>alignment_software (exp)</th>\n",
       "      <th>AssemblyName</th>\n",
       "      <th>study_disease</th>\n",
       "      <th>molecular_data_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>SRR10734314</td>\n",
       "      <td>RNA</td>\n",
       "      <td>OTHER</td>\n",
       "      <td>98</td>\n",
       "      <td>141427475196</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN13233895</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>452198_Diag.scRNA</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>452198</td>\n",
       "      <td>Yes</td>\n",
       "      <td>male</td>\n",
       "      <td>Cellranger</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>SRR10734316</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>98</td>\n",
       "      <td>144912081384</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN13234049</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>452198_RelPostAllo.scRNA</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>452198</td>\n",
       "      <td>Yes</td>\n",
       "      <td>male</td>\n",
       "      <td>Cellranger</td>\n",
       "      <td>GCA_000001405.13</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>SRR10734325</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>150</td>\n",
       "      <td>767908950</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN13234284</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_GM-816067-RNA-Seq_816067_17-200nt</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>816067</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>SRR10734326</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>199</td>\n",
       "      <td>14329702448</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN13233382</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_GM-816067-RNA-seq_816067</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>816067</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>SRR10734363</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>198</td>\n",
       "      <td>6472289839</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN13234063</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_KA-242129-RNA-seq_242129</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>242129</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>209</td>\n",
       "      <td>SRR14580604</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>246</td>\n",
       "      <td>8466486982</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN19012589</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_KA-875663-Bulk_RNA_875663_dx_BM</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>875663</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>210</td>\n",
       "      <td>SRR14580605</td>\n",
       "      <td>DNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>264</td>\n",
       "      <td>13928418755</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN19012616</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_KA-894912-894912_TBRS-3_PB_bulkRNA</td>\n",
       "      <td>pheresis_product</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>894912</td>\n",
       "      <td>No</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>211</td>\n",
       "      <td>SRR14580606</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>262</td>\n",
       "      <td>5946431600</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN19012635</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_KA-906708-Bulk_RNA_906708_dx_BM</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>906708</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>212</td>\n",
       "      <td>SRR14580607</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>289</td>\n",
       "      <td>20575367426</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN19012539</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_KA-916462-Bulk_RNA_916462_dx_BM</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>916462</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>213</td>\n",
       "      <td>SRR14580608</td>\n",
       "      <td>RNA</td>\n",
       "      <td>RNA-Seq</td>\n",
       "      <td>271</td>\n",
       "      <td>13280724270</td>\n",
       "      <td>PRJNA74783</td>\n",
       "      <td>SAMN19012592</td>\n",
       "      <td>AML_Genome</td>\n",
       "      <td>H_KA-976838-Bulk_RNA_976838_dx</td>\n",
       "      <td>bone_marrow</td>\n",
       "      <td>...</td>\n",
       "      <td>Tumor vs. Matched-Normal</td>\n",
       "      <td>Genomics of Acute Myeloid Leukemia</td>\n",
       "      <td>Yes</td>\n",
       "      <td>976838</td>\n",
       "      <td>Yes</td>\n",
       "      <td>female</td>\n",
       "      <td>HISAT</td>\n",
       "      <td>JTFH01</td>\n",
       "      <td>Leukemia\\, Myeloid\\, Acute</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>214 rows × 40 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             Run analyte_type Assay Type  AvgSpotLen         Bases  \\\n",
       "0    SRR10734314          RNA      OTHER          98  141427475196   \n",
       "1    SRR10734316          RNA    RNA-Seq          98  144912081384   \n",
       "2    SRR10734325          RNA    RNA-Seq         150     767908950   \n",
       "3    SRR10734326          RNA    RNA-Seq         199   14329702448   \n",
       "4    SRR10734363          RNA    RNA-Seq         198    6472289839   \n",
       "..           ...          ...        ...         ...           ...   \n",
       "209  SRR14580604          RNA    RNA-Seq         246    8466486982   \n",
       "210  SRR14580605          DNA    RNA-Seq         264   13928418755   \n",
       "211  SRR14580606          RNA    RNA-Seq         262    5946431600   \n",
       "212  SRR14580607          RNA    RNA-Seq         289   20575367426   \n",
       "213  SRR14580608          RNA    RNA-Seq         271   13280724270   \n",
       "\n",
       "     BioProject     BioSample biospecimen_repository  \\\n",
       "0    PRJNA74783  SAMN13233895             AML_Genome   \n",
       "1    PRJNA74783  SAMN13234049             AML_Genome   \n",
       "2    PRJNA74783  SAMN13234284             AML_Genome   \n",
       "3    PRJNA74783  SAMN13233382             AML_Genome   \n",
       "4    PRJNA74783  SAMN13234063             AML_Genome   \n",
       "..          ...           ...                    ...   \n",
       "209  PRJNA74783  SAMN19012589             AML_Genome   \n",
       "210  PRJNA74783  SAMN19012616             AML_Genome   \n",
       "211  PRJNA74783  SAMN19012635             AML_Genome   \n",
       "212  PRJNA74783  SAMN19012539             AML_Genome   \n",
       "213  PRJNA74783  SAMN19012592             AML_Genome   \n",
       "\n",
       "         biospecimen_repository_sample_id         body_site  ...  \\\n",
       "0                       452198_Diag.scRNA       bone_marrow  ...   \n",
       "1                452198_RelPostAllo.scRNA       bone_marrow  ...   \n",
       "2     H_GM-816067-RNA-Seq_816067_17-200nt       bone_marrow  ...   \n",
       "3              H_GM-816067-RNA-seq_816067       bone_marrow  ...   \n",
       "4              H_KA-242129-RNA-seq_242129       bone_marrow  ...   \n",
       "..                                    ...               ...  ...   \n",
       "209     H_KA-875663-Bulk_RNA_875663_dx_BM       bone_marrow  ...   \n",
       "210  H_KA-894912-894912_TBRS-3_PB_bulkRNA  pheresis_product  ...   \n",
       "211     H_KA-906708-Bulk_RNA_906708_dx_BM       bone_marrow  ...   \n",
       "212     H_KA-916462-Bulk_RNA_916462_dx_BM       bone_marrow  ...   \n",
       "213        H_KA-976838-Bulk_RNA_976838_dx       bone_marrow  ...   \n",
       "\n",
       "                 study_design                          study_name  \\\n",
       "0    Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "1    Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "2    Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "3    Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "4    Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "..                        ...                                 ...   \n",
       "209  Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "210  Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "211  Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "212  Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "213  Tumor vs. Matched-Normal  Genomics of Acute Myeloid Leukemia   \n",
       "\n",
       "     subject_is_affected submitted_subject_id Is_Tumor     sex  \\\n",
       "0                    Yes               452198      Yes    male   \n",
       "1                    Yes               452198      Yes    male   \n",
       "2                    Yes               816067      Yes  female   \n",
       "3                    Yes               816067      Yes  female   \n",
       "4                    Yes               242129      Yes  female   \n",
       "..                   ...                  ...      ...     ...   \n",
       "209                  Yes               875663      Yes  female   \n",
       "210                  Yes               894912       No  female   \n",
       "211                  Yes               906708      Yes  female   \n",
       "212                  Yes               916462      Yes  female   \n",
       "213                  Yes               976838      Yes  female   \n",
       "\n",
       "    alignment_software (exp)      AssemblyName               study_disease  \\\n",
       "0                 Cellranger  GCA_000001405.13  Leukemia\\, Myeloid\\, Acute   \n",
       "1                 Cellranger  GCA_000001405.13  Leukemia\\, Myeloid\\, Acute   \n",
       "2                        NaN               NaN  Leukemia\\, Myeloid\\, Acute   \n",
       "3                      HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "4                      HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "..                       ...               ...                         ...   \n",
       "209                    HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "210                    HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "211                    HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "212                    HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "213                    HISAT            JTFH01  Leukemia\\, Myeloid\\, Acute   \n",
       "\n",
       "    molecular_data_type  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  \n",
       "..                  ...  \n",
       "209                 NaN  \n",
       "210                 NaN  \n",
       "211                 NaN  \n",
       "212                 NaN  \n",
       "213                 NaN  \n",
       "\n",
       "[214 rows x 40 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## tmp block\n",
    "\n",
    "tmp = pd.read_csv('/data1/tmpfile.tsv', sep=',')\n",
    "tmp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b3869a22",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eb7abbba",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "db5c11f2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>p_devtest</th>\n",
       "      <th>fdr_p_devtest</th>\n",
       "      <th>ref_site</th>\n",
       "      <th>alt_site</th>\n",
       "      <th>b_intercept</th>\n",
       "      <th>p_intercept</th>\n",
       "      <th>b_ConditionALS_pathology</th>\n",
       "      <th>p_ConditionALS_pathology</th>\n",
       "      <th>strand</th>\n",
       "      <th>multiplyer</th>\n",
       "      <th>LFC_PA_Usage</th>\n",
       "      <th>negative_logFDR</th>\n",
       "      <th>switch_width</th>\n",
       "      <th>bed</th>\n",
       "      <th>correction_multiplyer</th>\n",
       "      <th>switch_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>0.261215</td>\n",
       "      <td>4.734204e-02</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>3.376034e-07</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>6.471593</td>\n",
       "      <td>16644</td>\n",
       "      <td>chr12,125124284,125140928</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125140928:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125143316:+</td>\n",
       "      <td>1.641746</td>\n",
       "      <td>4.866807e-52</td>\n",
       "      <td>0.263182</td>\n",
       "      <td>1.246049e-01</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.263182</td>\n",
       "      <td>0.904465</td>\n",
       "      <td>19032</td>\n",
       "      <td>chr12,125124284,125143316</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125143316:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>chr12:125143316:+</td>\n",
       "      <td>1.380531</td>\n",
       "      <td>8.547748e-46</td>\n",
       "      <td>-0.720173</td>\n",
       "      <td>1.476861e-08</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.720173</td>\n",
       "      <td>7.830660</td>\n",
       "      <td>2388</td>\n",
       "      <td>chr12,125140928,125143316</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125140928:125143316:+</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67201672:-</td>\n",
       "      <td>0.936799</td>\n",
       "      <td>7.468705e-06</td>\n",
       "      <td>-0.043726</td>\n",
       "      <td>8.609305e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.043726</td>\n",
       "      <td>0.065032</td>\n",
       "      <td>644</td>\n",
       "      <td>chr15,67201028,67201672</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67201672:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67202710:-</td>\n",
       "      <td>1.470439</td>\n",
       "      <td>7.400884e-14</td>\n",
       "      <td>-1.337591</td>\n",
       "      <td>1.070773e-07</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.337591</td>\n",
       "      <td>6.970303</td>\n",
       "      <td>1682</td>\n",
       "      <td>chr15,67201028,67202710</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67202710:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41174</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>-0.321971</td>\n",
       "      <td>2.534728e-02</td>\n",
       "      <td>0.526766</td>\n",
       "      <td>5.576124e-03</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.526766</td>\n",
       "      <td>2.253668</td>\n",
       "      <td>53729</td>\n",
       "      <td>chr1,77564505,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77618234:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41175</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.016713</td>\n",
       "      <td>8.988251e-01</td>\n",
       "      <td>-0.225352</td>\n",
       "      <td>2.362558e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.225352</td>\n",
       "      <td>0.626618</td>\n",
       "      <td>65335</td>\n",
       "      <td>chr1,77564505,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77629840:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41176</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>-0.221592</td>\n",
       "      <td>1.320936e-01</td>\n",
       "      <td>0.933971</td>\n",
       "      <td>6.105239e-06</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.933971</td>\n",
       "      <td>5.214297</td>\n",
       "      <td>37236</td>\n",
       "      <td>chr1,77580998,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77580998:77618234:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41177</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.117093</td>\n",
       "      <td>3.854144e-01</td>\n",
       "      <td>0.181853</td>\n",
       "      <td>3.789934e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.181853</td>\n",
       "      <td>0.421368</td>\n",
       "      <td>48842</td>\n",
       "      <td>chr1,77580998,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77580998:77629840:-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41178</th>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.338685</td>\n",
       "      <td>1.825640e-02</td>\n",
       "      <td>-0.752118</td>\n",
       "      <td>1.128067e-04</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.752118</td>\n",
       "      <td>3.947665</td>\n",
       "      <td>11606</td>\n",
       "      <td>chr1,77618234,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77618234:77629840:-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41179 rows × 17 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript  p_devtest  fdr_p_devtest           ref_site  \\\n",
       "0           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "1           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "2           AACS   0.006145       0.019492  chr12:125140928:+   \n",
       "3          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "4          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "...          ...        ...            ...                ...   \n",
       "41174       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41175       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41176       ZZZ3   1.000000       1.000000    chr1:77580998:-   \n",
       "41177       ZZZ3   1.000000       1.000000    chr1:77580998:-   \n",
       "41178       ZZZ3   1.000000       1.000000    chr1:77618234:-   \n",
       "\n",
       "                alt_site  b_intercept   p_intercept  b_ConditionALS_pathology  \\\n",
       "0      chr12:125140928:+     0.261215  4.734204e-02                  0.983355   \n",
       "1      chr12:125143316:+     1.641746  4.866807e-52                  0.263182   \n",
       "2      chr12:125143316:+     1.380531  8.547748e-46                 -0.720173   \n",
       "3       chr15:67201672:-     0.936799  7.468705e-06                 -0.043726   \n",
       "4       chr15:67202710:-     1.470439  7.400884e-14                 -1.337591   \n",
       "...                  ...          ...           ...                       ...   \n",
       "41174    chr1:77618234:-    -0.321971  2.534728e-02                  0.526766   \n",
       "41175    chr1:77629840:-     0.016713  8.988251e-01                 -0.225352   \n",
       "41176    chr1:77618234:-    -0.221592  1.320936e-01                  0.933971   \n",
       "41177    chr1:77629840:-     0.117093  3.854144e-01                  0.181853   \n",
       "41178    chr1:77629840:-     0.338685  1.825640e-02                 -0.752118   \n",
       "\n",
       "       p_ConditionALS_pathology strand  multiplyer  LFC_PA_Usage  \\\n",
       "0                  3.376034e-07      +           1      0.983355   \n",
       "1                  1.246049e-01      +           1      0.263182   \n",
       "2                  1.476861e-08      +           1     -0.720173   \n",
       "3                  8.609305e-01      -          -1      0.043726   \n",
       "4                  1.070773e-07      -          -1      1.337591   \n",
       "...                         ...    ...         ...           ...   \n",
       "41174              5.576124e-03      -          -1     -0.526766   \n",
       "41175              2.362558e-01      -          -1      0.225352   \n",
       "41176              6.105239e-06      -          -1     -0.933971   \n",
       "41177              3.789934e-01      -          -1     -0.181853   \n",
       "41178              1.128067e-04      -          -1      0.752118   \n",
       "\n",
       "       negative_logFDR  switch_width                        bed  \\\n",
       "0             6.471593         16644  chr12,125124284,125140928   \n",
       "1             0.904465         19032  chr12,125124284,125143316   \n",
       "2             7.830660          2388  chr12,125140928,125143316   \n",
       "3             0.065032           644    chr15,67201028,67201672   \n",
       "4             6.970303          1682    chr15,67201028,67202710   \n",
       "...                ...           ...                        ...   \n",
       "41174         2.253668         53729     chr1,77564505,77618234   \n",
       "41175         0.626618         65335     chr1,77564505,77629840   \n",
       "41176         5.214297         37236     chr1,77580998,77618234   \n",
       "41177         0.421368         48842     chr1,77580998,77629840   \n",
       "41178         3.947665         11606     chr1,77618234,77629840   \n",
       "\n",
       "       correction_multiplyer                       switch_name  \n",
       "0                          1  chr12:AACS:125124284:125140928:+  \n",
       "1                          1  chr12:AACS:125124284:125143316:+  \n",
       "2                          1  chr12:AACS:125140928:125143316:+  \n",
       "3                          1   chr15:AAGAB:67201028:67201672:-  \n",
       "4                          1   chr15:AAGAB:67201028:67202710:-  \n",
       "...                      ...                               ...  \n",
       "41174                      1     chr1:ZZZ3:77564505:77618234:-  \n",
       "41175                      1     chr1:ZZZ3:77564505:77629840:-  \n",
       "41176                      1     chr1:ZZZ3:77580998:77618234:-  \n",
       "41177                      1     chr1:ZZZ3:77580998:77629840:-  \n",
       "41178                      1     chr1:ZZZ3:77618234:77629840:-  \n",
       "\n",
       "[41179 rows x 17 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## test blocks\n",
    "tst_df = pd.read_csv('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/APAlog_res_metadata_added.tsv', sep='\\t')\n",
    "tst_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cdf3b891",
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'seaborn'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Input \u001b[0;32mIn [4]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mseaborn\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01msns\u001b[39;00m\n\u001b[1;32m      2\u001b[0m sns\u001b[38;5;241m.\u001b[39mdistplot(tst_df[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mLFC_PA_Usage\u001b[39m\u001b[38;5;124m'\u001b[39m])\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'seaborn'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "03e7295b",
   "metadata": {},
   "source": [
    "## lets make celltype specific input data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "2b69c77b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Inhibitory_sALSvsCTRL',\n",
       " 'Excitatory_sALSvsCTRL',\n",
       " 'L2-3_sALSvsCTRL',\n",
       " 'AST-FB_sALSvsCTRL',\n",
       " 'L4_sALSvsCTRL',\n",
       " 'IN-VIP_sALSvsCTRL',\n",
       " 'AST-PP_sALSvsCTRL',\n",
       " 'L5-6-CC_sALSvsCTRL',\n",
       " 'Oligodendrocytes_sALSvsCTRL',\n",
       " 'IN-SST_sALSvsCTRL',\n",
       " 'OPC_sALSvsCTRL',\n",
       " 'Microglia_sALSvsCTRL',\n",
       " 'IN-PV_sALSvsCTRL']"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ct = [ e for e in os.listdir(data_root) if 'sALS' in e and 'pkl' not in e]\n",
    "ct"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "743c8fff",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Inhibitory_sALSvsCTRL\n",
      "24308\n",
      "Inhibitory\n",
      "24308\n",
      "Excitatory_sALSvsCTRL\n",
      "15784\n",
      "Excitatory\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aiden/anaconda3/lib/python3.7/site-packages/pandas/core/series.py:1143: FutureWarning: \n",
      "Passing list-likes to .loc or [] with any missing label will raise\n",
      "KeyError in the future, you can use .reindex() as an alternative.\n",
      "\n",
      "See the documentation here:\n",
      "https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike\n",
      "  return self.loc[key]\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "15784\n",
      "L2-3_sALSvsCTRL\n",
      "7588\n",
      "L2-3\n",
      "7588\n",
      "AST-FB_sALSvsCTRL\n",
      "1022\n",
      "AST-FB\n",
      "1022\n",
      "L4_sALSvsCTRL\n",
      "2324\n",
      "L4\n",
      "2324\n",
      "IN-VIP_sALSvsCTRL\n",
      "5667\n",
      "IN-VIP\n",
      "5667\n",
      "AST-PP_sALSvsCTRL\n",
      "45347\n",
      "AST-PP\n",
      "45347\n",
      "L5-6-CC_sALSvsCTRL\n",
      "61\n",
      "L5-6-CC\n",
      "61\n",
      "Oligodendrocytes_sALSvsCTRL\n",
      "59817\n",
      "Oligodendrocytes\n",
      "59817\n",
      "IN-SST_sALSvsCTRL\n",
      "9234\n",
      "IN-SST\n",
      "9234\n",
      "OPC_sALSvsCTRL\n",
      "38160\n",
      "OPC\n",
      "38160\n",
      "Microglia_sALSvsCTRL\n",
      "27964\n",
      "Microglia\n",
      "27964\n",
      "IN-PV_sALSvsCTRL\n",
      "5305\n",
      "IN-PV\n",
      "5305\n"
     ]
    }
   ],
   "source": [
    "# ct = ['Excitatory_C9ALSvsCTRL']\n",
    "for e in ct:\n",
    "    print(e)\n",
    "    sequences_dict = {}\n",
    "    data_dict = {}\n",
    "    inp_fa = data_root + \"/{}/switch_DNA_sequence.fa\".format(e)\n",
    "    inp_fa = SeqIO.parse(inp_fa, \"fasta\")\n",
    "    for rec in inp_fa:\n",
    "        if rec.id not in sequences_dict:\n",
    "            sequences_dict[rec.id] = str(rec.seq)\n",
    "        else:\n",
    "            continue\n",
    "    print(len(sequences_dict))\n",
    "    celltype = e.split('_')[0]\n",
    "    print(celltype)\n",
    "    df_name = data_root + celltype + \"_sALSvsCTRL/APAlog_res_metadata_added.tsv\"\n",
    "    inp_df = pd.read_csv(df_name, sep='\\t')\n",
    "    inp_df['LFC_rand_INT'] = rank_INT(inp_df['sig_LFC_PA_Usage'])\n",
    "    for key,value in sequences_dict.items():\n",
    "        if key in inp_df['switch_name'].values:\n",
    "            label = get_sig_lfc(inp_df, key)\n",
    "            data_dict[key] = {}\n",
    "            data_dict[key][value] = float(label)\n",
    "    print(len(data_dict))\n",
    "    ids, seqs, labels =[],[],[]\n",
    "    for key, value in data_dict.items():\n",
    "        ids.append(key)\n",
    "        strand = key.split(':')[-1]\n",
    "        for k, v in value.items():\n",
    "            if strand == '+':\n",
    "                seq = transcribe_positive_strand(k)\n",
    "            else:\n",
    "                seq = transcribe_negative_strand(k)\n",
    "            seqs.append(seq)\n",
    "            labels.append(v)\n",
    "    df_data = pd.DataFrame([ids,seqs,labels]).T\n",
    "    df_data.columns = ['site_name','seqs', 'label']\n",
    "    outname = data_root + celltype + \"_sALSvsCTRL/input_data.tsv\"\n",
    "    df_data.to_csv(outname, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "52a403af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "id": "228d7a9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n"
     ]
    },
    {
     "ename": "IndexError",
     "evalue": "list index out of range",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-280-3119d15c106e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mrecids\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrec\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mid\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecids\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrecids\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mIndexError\u001b[0m: list index out of range"
     ]
    }
   ],
   "source": [
    "recids = []\n",
    "for rec in inp_fa:\n",
    "    recids.append(rec.id)\n",
    "print(len(recids))\n",
    "print(recids[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "46e62203",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>site_name</th>\n",
       "      <th>seqs</th>\n",
       "      <th>label</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr12:AACS:125140928:125143316:+</td>\n",
       "      <td>GUGAGGCGGGACAAACUUGUCUUCCUCACACCCAUCUUACUUCCUC...</td>\n",
       "      <td>0.0371</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201672:67202710:-</td>\n",
       "      <td>ACUACAUCAUAAACAUGUCUUUGAAACCCGUCUCCCAUCUUCUAGU...</td>\n",
       "      <td>1.2828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr15:AAGAB:67202710:67217079:-</td>\n",
       "      <td>AUCUAUUAUAGUUCAAUCCCCAGUAAUGCAGAUGGAGGAGGAUUAG...</td>\n",
       "      <td>-0.2141</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>chr2:AAK1:69457997:69461526:-</td>\n",
       "      <td>UUGUAUUCUUAGUAGAGAUGGGGUUUCACCAUGUUGGCCAGGUUGG...</td>\n",
       "      <td>0.0173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>chr2:AAK1:69457997:69467864:-</td>\n",
       "      <td>AAAAACCCACUUUUUAACAAAUAAUUUUGAUGGGUUAGACAUGUUC...</td>\n",
       "      <td>-0.9525</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15779</td>\n",
       "      <td>15779</td>\n",
       "      <td>chr7:ZYX:143382988:143391082:+</td>\n",
       "      <td>AGUCGGGAAUGUGGGGCACGAAUCUUCCCCUCCGGGUCCGUUUUCC...</td>\n",
       "      <td>0.7120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15780</td>\n",
       "      <td>15780</td>\n",
       "      <td>chr17:ZZEF1:4004475:4006020:-</td>\n",
       "      <td>AAUGAGGGAGUGGCUCCCGACAUUGGCGACAACUCUUUUUCCCUCU...</td>\n",
       "      <td>-0.5428</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15781</td>\n",
       "      <td>15781</td>\n",
       "      <td>chr17:ZZEF1:4004475:4008624:-</td>\n",
       "      <td>GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA...</td>\n",
       "      <td>-0.4334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15782</td>\n",
       "      <td>15782</td>\n",
       "      <td>chr17:ZZEF1:4006020:4008624:-</td>\n",
       "      <td>GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA...</td>\n",
       "      <td>0.1435</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15783</td>\n",
       "      <td>15783</td>\n",
       "      <td>chr17:ZZEF1:4008624:4021776:-</td>\n",
       "      <td>AGUUGCUGGUGAAAAAGUGCAGCAAAGGGACUGGCUUUAGUAAAAC...</td>\n",
       "      <td>0.3742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15784 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0                         site_name  \\\n",
       "0               0  chr12:AACS:125140928:125143316:+   \n",
       "1               1   chr15:AAGAB:67201672:67202710:-   \n",
       "2               2   chr15:AAGAB:67202710:67217079:-   \n",
       "3               3     chr2:AAK1:69457997:69461526:-   \n",
       "4               4     chr2:AAK1:69457997:69467864:-   \n",
       "...           ...                               ...   \n",
       "15779       15779    chr7:ZYX:143382988:143391082:+   \n",
       "15780       15780     chr17:ZZEF1:4004475:4006020:-   \n",
       "15781       15781     chr17:ZZEF1:4004475:4008624:-   \n",
       "15782       15782     chr17:ZZEF1:4006020:4008624:-   \n",
       "15783       15783     chr17:ZZEF1:4008624:4021776:-   \n",
       "\n",
       "                                                    seqs   label  \n",
       "0      GUGAGGCGGGACAAACUUGUCUUCCUCACACCCAUCUUACUUCCUC...  0.0371  \n",
       "1      ACUACAUCAUAAACAUGUCUUUGAAACCCGUCUCCCAUCUUCUAGU...  1.2828  \n",
       "2      AUCUAUUAUAGUUCAAUCCCCAGUAAUGCAGAUGGAGGAGGAUUAG... -0.2141  \n",
       "3      UUGUAUUCUUAGUAGAGAUGGGGUUUCACCAUGUUGGCCAGGUUGG...  0.0173  \n",
       "4      AAAAACCCACUUUUUAACAAAUAAUUUUGAUGGGUUAGACAUGUUC... -0.9525  \n",
       "...                                                  ...     ...  \n",
       "15779  AGUCGGGAAUGUGGGGCACGAAUCUUCCCCUCCGGGUCCGUUUUCC...  0.7120  \n",
       "15780  AAUGAGGGAGUGGCUCCCGACAUUGGCGACAACUCUUUUUCCCUCU... -0.5428  \n",
       "15781  GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA... -0.4334  \n",
       "15782  GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA...  0.1435  \n",
       "15783  AGUUGCUGGUGAAAAAGUGCAGCAAAGGGACUGGCUUUAGUAAAAC...  0.3742  \n",
       "\n",
       "[15784 rows x 4 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_name = data_root + 'Excitatory' + \"_sALSvsCTRL/input_data.tsv\"\n",
    "inp_df = pd.read_csv(df_name, sep='\\t')\n",
    "inp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "005d4fdb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdc0b60c0d0>"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXoAAAEGCAYAAABrQF4qAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deXSU933v8fdXo31FQhtCAgkQizA2GBkc7/EW20lw2ms3tpPWaXPrpo2b9KTtbdLkJLfOSZult02buk2c2G3SRCGOs9HENt4dY4NAmFWAQAiQhPZ9QdvMfO8fM8JjIdAILc8s39c5OsyzzXw1aD569Ht+z+8nqooxxpjIFeN0AcYYY+aWBb0xxkQ4C3pjjIlwFvTGGBPhLOiNMSbCxTpdwETZ2dlaXFzsdBnGGBNW9u7d26GqOZNtC7mgLy4upqqqyukyjDEmrIjImYtts6YbY4yJcBb0xhgT4SzojTEmwlnQG2NMhLOgN8aYCGdBb4wxEc6C3hhjIpwFvTHGRDgLemOMiXAhd2esMaGsorL+gnUPbV7iQCXGBM+C3pggtPUP094/wqmOQVISXOSmJTpdkjFBCyroReQu4F8AF/A9Vf3qRfa7D/gpcI2qVvnXfQ74OOABPqWq22ejcGPmUkVlPV5Vjrf2s/NkJyfaBt61PTctgSsLF3BjabZDFRoTvCmDXkRcwOPAHUAjsEdEtqnqkQn7pQGfAioD1pUBDwBrgQLgJRFZqaqe2fsWjJl9o24vP9lTz9GWftISY7l1dS6LMhJJiHXRPjDCocZeXjrayrGWPu66Ip+8dDvDN6ErmDP6TUCtqtYBiMhW4F7gyIT9vgx8HfirgHX3AltVdQQ4JSK1/ufbOdPCjZkrnQMjPLmjjsbuIe5Zt4hrl2URG/NOv4UVuam8Z9lCjjT18XRVA1v+bQdPfewa1hZkOFi1MRcXTK+bxUBDwHKjf915IrIBKFLVX0/3WP/xj4hIlYhUtbe3B1W4MXNhaNTDg9/dRXPvMA9tXsINK7LfFfKBygrS+ZOblzE85uUj36vkyTdOUVFZf/7LmFARTNDLJOv0/EaRGOCfgb+c7rHnV6g+oarlqlqekzPpuPnGzIt/eO4ox1sH+Oi1S4M6Q1+UkcRDm5bQNzTGrw6cRfWCH29jHBdM000jUBSwXAg0BSynAVcAr4kIQD6wTUS2BHGsMY4bP/s+3trPD3ae4frlC1mZlxb08UVZydy6Oo+XjrayKi+NDUsy56pUYy5LMGf0e4BSESkRkXh8F1e3jW9U1V5VzVbVYlUtBnYBW/y9brYBD4hIgoiUAKXA7ln/LoyZoXOjbn72diO5aQncuTZ/2sffsiqH4oXJbDvQxOCIew4qNObyTRn0quoGHgW2A0eBp1W1WkQe85+1X+rYauBpfBdunwc+aT1uTCh6raadgWE3v1deRJxr+jeMx4jwofWLGXV7+e0Ju85kQktQ/ehV9Vng2QnrvniRfW+ZsPwV4CuXWZ8xc65vaIxddZ2sL1pAwYKky36e3PRE1hctYFddJ9evsP71JnTYWDcm6r1a04ZXldvW5M34uW5bk4fHq7xW0zYLlRkzOyzoTVRr6DpH1eluyouzyEqJn/HzZaXEU740iz2numnoOjcLFRozcxb0Jqr92yu1iMB7V+XO2nO+d7XvuZ5689SsPacxM2FBb6JWW/8wv9h3lo1LM8lIipu1581IimNdYQY/rWqkf3hs1p7XmMtlQW+i1g93nmHM652TC6fXLV/IwIibn+1tnPXnNma6LOhNVBoe8/Dfu85w+5o8slMTZv35CzOT2bBkAd/feQav1+6WNc6yoDdR6edvn6X73Bj/+4aSOXuNj11XzKmOQV4/bv3qjbMs6E3U8XqVJ3fUsW5xBptKsubsde6+YhG5aQn851un5+w1jAmGBb2JOm+d7ORk+yB/dEMx/vGZ5kR8bAwPbFrCGyfaaey2rpbGORb0JupU7D5DZnIc96xbNOevdf/GQgCesYuyxkEW9CaqtPUN80J1K/eXF5EQ65rz1yvKSuaGFdn8tKoRj12UNQ6xoDdRo6Kyns//8jBur5KaEDtvE4T8XnkRZ3uGeLO2Y85fy5jJBDWomTGRwKvKnlNdrMhJnZMulRON/xJxe7wkxbn4xvYablppE+uY+Wdn9CZqHG/tp2dojGvmsKfNZGJdMVy9ZAFHmvvoHhyd19c2BizoTRTZfaqLtIRYyhalz/trX700E49X+fWh5nl/bWMs6E1UONszRE1LPxuLM3HFzF2XyotZlJFEXnoCv9x3dt5f25iggl5E7hKRGhGpFZHPTrL9EyJySET2i8gOESnzry8WkSH/+v0i8u3Z/gaMCcZPdvvay68pnt9mm0DrCxew90w39Z3Wp97MrymDXkRcwOPA3UAZ8OB4kAeoUNV1qroe+DrwTwHbTqrqev/XJ2arcGOCNebxsnVPAyvz0shMnvmY85fryqIFAPxqv53Vm/kVzBn9JqBWVetUdRTYCtwbuIOq9gUspgDWYdiEjJePttHWPzKnwx0EIzM5nk0lWfxy/1lU7SNi5k8wQb8YaAhYbvSvexcR+aSInMR3Rv+pgE0lIrJPRF4XkRtnVK0xl+FHlWcoyEhkVX6a06XwofWLOdk+SHVT39Q7GzNLggn6ya5cXXA6oqqPq+py4G+AL/hXNwNLVHUD8BmgQkQu6PIgIo+ISJWIVLW320h/ZvY09w6xo7aD+8qLiJnDcW2Cdc+6fOJcYhdlzbwKJugbgaKA5UKg6RL7bwU+BKCqI6ra6X+8FzgJrJx4gKo+oarlqlqek2M3lJjZ86v9TajC72644I9QRyxIjueGFdk8X91izTdm3gQT9HuAUhEpEZF44AFgW+AOIlIasPh+4IR/fY7/Yi4isgwoBepmo3BjpqKq/OLts1y9ZAHF2SlOl3Pe3VcsorF7yJpvzLyZcggEVXWLyKPAdsAFPKWq1SLyGFClqtuAR0XkdmAM6AYe9h9+E/CYiLgBD/AJVe2ai2/EmEAVlfU09w5R09rPlqsK5mVMm2BUVNYzOOJGgH98oYY7y/IBeGjzEmcLMxEtqLFuVPVZ4NkJ674Y8PjTFznuZ8DPZlKgMZdrX30PLhGuXJzhdCnvkpIQS0l2CtVNfeeD3pi5ZHfGmojkVeVAYw8r89NITgi9sfvWLs6gvX+Etr5hp0sxUcCC3kSkuvZB+ofdrPffpBRqxsfbOdJs7fRm7lnQm4hU3dRLnEtYHQJ95yeTkRTHkqxkDjf1Ol2KiQIW9CbieL3K0eY+SnPTiHOF7o942aJ0mnqG6R0ac7oUE+FC91NgzGU6eLaXvmE3awvmfzji6Vjp/2vjRGu/w5WYSGdBbyLO9uoWYgRW54d20OelJZCRFEeNBb2ZYxb0JuJsr25hWXYqSfFzP/n3TIgIK/NSqW0bYMzjdbocE8Es6E1EqW0boK59kDUh3mwzbmVeGiNuL3vPdDtdiolgFvQmomyvbgFwZLrAy7E8J5UYgddqbDA/M3cs6E1EefVYG1cWZpCRFOd0KUFJjHNRvDCF12ranC7FRDALehMx+ofH2NfQw42l2U6XMi0r89I41tJPc++Q06WYCGVBbyLGrrouPF7lhhXhNdT1eDfLN453OFyJiVQW9CZi7DjRTlKci6uXhuawBxeTl5ZAdmoCb560oDdzI/RGezJmGgKHH/7NoRaKspL42d7wmr1JRLhu+ULeOtmJqiIhMBOWiSx2Rm8iQs+5UToGRliRG5pj20zl+hULae8fobZtwOlSTASyoDcRYTwgV+SmOlzJ5bluue8C8lsnOx2uxEQiC3oTEWrbB0hLjCUvLcHpUi5LUVYyhZlJvGXt9GYOBBX0InKXiNSISK2IfHaS7Z8QkUMisl9EdohIWcC2z/mPqxGR981m8caAb5KRk20DrMhJDdv27YrKevLSE3n9eDs/3HWGisr6kJn+0IS/KYPeP7n348DdQBnwYGCQ+1Wo6jpVXQ98Hfgn/7Fl+CYTXwvcBfz7+GThxsyWtv4RBkc9LMsJz2abcctzUhke89LcY7NOmdkVzBn9JqBWVetUdRTYCtwbuIOqBk6TkwKo//G9wFZVHVHVU0Ct//mMmTWnOwYBKMlOcbiSmVmW46v/ZLtdkDWzK5igXww0BCw3+te9i4h8UkRO4juj/9Q0j31ERKpEpKq93cb8MNNzunOQ9MRYMpPDY9iDi0lPjCMnLcGC3sy6YIJ+skZPvWCF6uOquhz4G+AL0zz2CVUtV9XynJzwuqvROEtVOd0xSHF2Sti2zwdalp3Cmc5zeLwXfEyMuWzBBH0jUBSwXAg0XWL/rcCHLvNYY6al+9wYfcNuiheGd7PNuJLsFEY9Xpp6bNwbM3uCCfo9QKmIlIhIPL6Lq9sCdxCR0oDF9wMn/I+3AQ+ISIKIlAClwO6Zl22Mzyl/+3xxmLfPjxu/zlDn/76MmQ1TDoGgqm4ReRTYDriAp1S1WkQeA6pUdRvwqIjcDowB3cDD/mOrReRp4AjgBj6pqp45+l5MFDrdOUhSnIvcMO0/P1FaYhw5qQmc6hjg5pXWjGlmR1Bj3ajqs8CzE9Z9MeDxpy9x7FeAr1xugcZcyumOQYoXJhMTAe3z40pyUjjQ0GPt9GbW2J2xJmy19Q3TOTgaMc0240qyUxhxe218ejNrLOhN2Np9ugsgYi7Ejhtvpz9l7fRmlljQm7C151QX8a4YChYkOV3KrEpPjCM7NZ66dgt6Mzss6E3Y2n26myVZybhiIqd9flxJdiqnOwetnd7MCgt6E5Z6h8Y41tJHcXay06XMifF2+iNNfVPvbMwULOhNWNp7pgvVyGufHzfeTl95ysanNzNnQW/CUuWpLuJcQlFWZJ7RZyTFsTAlnl11FvRm5izoTVjac6qLKwsXEOeK3B/hkuwUdp/qsnZ6M2OR+ykxEWto1MPBxl6uKc5yupQ5VZKdQt+wm2Mt1k5vZsaC3oSdfQ3duL3K5pLID3qAXXVdDldiwp0FvQk7u091IQJXL810upQ5tSA5niVZyVRaO72ZIQt6E3b2nO5iTX46GUnhPdFIMK5dlsXu0114rZ3ezIAFvQkro24ve890synCm23GbS5ZSM+5MWpa+50uxYQxC3oTVg429jA85uXaZQudLmVebF7m+4Vm3SzNTFjQm7Cy82QnIkT8hdhxhZnJFGUlsfOkBb25fBb0JqzsOtXJ6vx0MlPinS5l3ly3LJtddZ3Wn95ctqCCXkTuEpEaEakVkc9Osv0zInJERA6KyMsisjRgm0dE9vu/tk081phgjbg97D3TzbXLouNsftx1KxbSN+ymuqnX6VJMmJoy6EXEBTwO3A2UAQ+KSNmE3fYB5ap6JfAM8PWAbUOqut7/tWWW6jZR6GBjb1S1zwNUVNbT0jsMwOOvnqSist7hikw4CuaMfhNQq6p1qjoKbAXuDdxBVV9V1XP+xV1A4eyWaQx85/WTCNDQdY6KyvqoCb20xDhy0xKoax9wuhQTpoIJ+sVAQ8Byo3/dxXwceC5gOVFEqkRkl4h8aLIDROQR/z5V7e3tQZRkolFdxyD5GYkkxwc11XFEWZ7rG5/e7fE6XYoJQ8EE/WSzOkx6VUhEPgqUA98IWL1EVcuBh4BvisjyC55M9QlVLVfV8pycnCBKMtFmxO2hvvMcyyJsfthgLc9OZcyj1Hefm3pnYyYIJugbgaKA5UKgaeJOInI78Hlgi6qOjK9X1Sb/v3XAa8CGGdRrotT++h7cXj0//ku0KclOQcCmFzSXJZig3wOUikiJiMQDDwDv6j0jIhuA7+AL+baA9ZkikuB/nA1cDxyZreJN9NhR24Hgm2IvGiXFu1icmcTJNmunN9M3ZdCrqht4FNgOHAWeVtVqEXlMRMZ70XwDSAV+OqEb5RqgSkQOAK8CX1VVC3ozbb890UFRVjJJ8S6nS3HM8pxUGrrP0T885nQpJswEdVVLVZ8Fnp2w7osBj2+/yHFvAetmUqAxvefGONTYwy2rcp0uxVGluam8frydXXVd3FGW53Q5JozYnbEm5L11sgOv+oIumi3JSibeFcNvj1vPNDM9FvQm5P32RAepCbEUZkbm/LDBinXFUJKdwhsnLOjN9FjQm5C3o7ada5ctxBUzWU/f6FKal8rpznPUd1o3SxM8C3oT0s50DtLQNcRNK7OdLiUklOamAfBbO6s302BBb0Lab090AHDDCgt6gOzUeBYvSLLmGzMtFvQmpL1xvJ3FC5Ki9kapiUSEm1Zm81ZtJ2M2HIIJkgW9CVmjbi9v1nZw86ocRKx9ftxNpTn0j7jZ39DjdCkmTFjQm5BVdaaLwVEPt6y08Y8CXbciG1eM8FpN29Q7G4MFvQlhr9W0E+cSrrf2+XfJSIqjfGkmLx+1oDfBib7xXk3IGx9n/pf7zrIkK5lf7b9gDL2od9uaXP7+2WM09QxRsCDJ6XJMiLMzehOSes6N0tY/wsq8NKdLCUm3rvYNB/HKMTurN1OzoDch6Xirb5TGVRb0k1qek8qSrGQLehMUC3oTkmpa+1mQHEdOWoLTpYQkEeHW1bm8WdvB0KjH6XJMiLOgNyHH7fVysn2AlXlp1q3yEm5dncuI28vOug6nSzEhzi7GmpBzuuMco26vNdtcxPjFarfHS7wrhm+/XkdL7wgPbV7icGUmVNkZvQk5x1r6iI0RludE97DEU4l1xVCal8qx5j68Ouk0zsYAQQa9iNwlIjUiUisin51k+2dE5IiIHBSRl0VkacC2h0XkhP/r4dks3kQeVeVYSz/Lc1KJj7XzkKmULUqnb9jN2e4hp0sxIWzKT5KIuIDHgbuBMuBBESmbsNs+oFxVrwSeAb7uPzYL+BKwGdgEfElEMmevfBNp6joG6RocZVW+NdsEY3V+OjECR5r7nC7FhLBgTpk2AbWqWqeqo8BW4N7AHVT1VVUdHyB7F1Dof/w+4EVV7VLVbuBF4K7ZKd1Eolf8d3uutqAPSlK8i5LsFKqbLOjNxQUT9IuBhoDlRv+6i/k48Nx0jhWRR0SkSkSq2ttt+NVo9vKxVvLTE1mQHO90KWGjrCCDjoERatsGnC7FhKhggn6y/m2TXvkRkY8C5cA3pnOsqj6hquWqWp6TYwNYRaveoTH2nO62ZptpKluUDsD26haHKzGhKpigbwSKApYLgQsGHxGR24HPA1tUdWQ6xxoD8MaJdjxetWabacpIiqMwM4kXLOjNRQQT9HuAUhEpEZF44AFgW+AOIrIB+A6+kA+8J3s7cKeIZPovwt7pX2fMBV452kZmchxFWdE9CfjlWLsonQONvTT3Wu8bc6Epg15V3cCj+AL6KPC0qlaLyGMissW/2zeAVOCnIrJfRLb5j+0Cvozvl8Ue4DH/OmPexeNVXq1p45ZVucTY3bDTtqbA13zz4pFWhysxoSioO2NV9Vng2Qnrvhjw+PZLHPsU8NTlFmiiw/6GbrrPjXHr6lz6h91OlxN2ctMSWZ6TwvbqFv7gPcVOl2NCjN2RYkLCK8facMUIN9lsUpftfWvz2VXXRc+5UadLMSHGgt6EhJePtlG+NJOMpDinSwlb71ubj8erNvOUuYANamYcd7ZniGMt/fztPaudLiWsHTrbS3piLE/uOMWI2wtgA50ZwM7oTQgYnzzj1tV5DlcS3mJEKCtI50RbP6P+oDcGLOiNwyoq6/nhzjNkpcRTWdd5fghec3nKFmUw5lFq2/qdLsWEEAt646hRt2+SkVX5NsnIbCjJTiEpzmVj35h3saA3jjrR1o/bq+dv4zcz44oRVuencaylH4/Xxqg3Phb0xlFHm/tIjIuheGGK06VEjLKCdIbGPJzuHHS6FBMiLOiNY9weL8da+lmdn44rxpptZktpbhpxLrHmG3OeBb1xzN4z3Zwb9bDGmm1mVXxsDKW5aRxt7kNtikGDBb1x0EtHW3GJUJprc8POtrKCdHqHxjjY2Ot0KSYEWNAbR6gqLx5pZVlOColxLqfLiTir89OIEXjhiA1dbCzojUNq2wY43XnOmm3mSHJ8LMXZKWyvttEsjQW9ccgL/uF0LejnTtmidGrbBjhjvW+ingW9ccSLR1q5sjDDBjGbQ6vyfDN1jQ8xYaKXBb2Zd239w+xv6OH2NTa2zVxamJrAitxUC3oTXNCLyF0iUiMitSLy2Um23yQib4uIW0Tum7DN45916vzMUya6jQ+je0eZBf1cu3V1LrvqOhkYsclcotmUQS8iLuBx4G6gDHhQRMom7FYPfAyomOQphlR1vf9ryyTbTZR58UgrhZlJNgn4PLh1dS5jHmXHiXanSzEOCuaMfhNQq6p1qjoKbAXuDdxBVU+r6kHAxkY1lzQ44mZHbQd3lOXZIGbzYOPSTNITY635JsoFE/SLgYaA5Ub/umAlikiViOwSkQ9NtoOIPOLfp6q93c48ItkbJ9oZdXu5w9rn50WcK4abVubwyrF2vDbIWdQKJugnO+2azk/MElUtBx4Cvikiyy94MtUnVLVcVctzcmzO0Ej2wpFW0hNjuaYky+lSosZta3LpGBjh0Fm7SzZaBRP0jUBRwHIh0BTsC6hqk//fOuA1YMM06jMRZMTt4cUjrdxRlk+cyzp8zZebV+YSI9bNMpoF82nbA5SKSImIxAMPAEH1nhGRTBFJ8D/OBq4HjlxusSZ8VVTW85XfHKV/2E1KgouKynqbTWqeZKXEs2FJpgV9FJsy6FXVDTwKbAeOAk+rarWIPCYiWwBE5BoRaQTuB74jItX+w9cAVSJyAHgV+KqqWtBHqcNne0mMi2GFDWI2725dncuhs7209Q07XYpxQGwwO6nqs8CzE9Z9MeDxHnxNOhOPewtYN8MaTQRwe7wcae6jbFE6sTHWbDNfxv9qGhnzdYj76nPHKC/O4qHNS5wsy8wz+8SZeVHbPsDwmJcrFmc4XUpUyktPICMpjmMtNml4NLKgN/PCmm2cJeKbS7a2bYAxj93uEm0s6M2cG3Vbs00oWJ2fxqjHy6kOG80y2tinzsy5N2s7rNkmBCzLSSXOJdZ8E4Us6M2c+82hZmu2CQFxrhiW56TaXLJRyILezKlRt5cXqltYk2/NNqHgioIMeofGOGBzyUYV++SZOfVmbQd9w27WFVqzTShYsyidGIHnDjc7XYqZRxb0Zk795lAzaYmx1mwTIpLiXSzPSeX5wy3WfBNFLOjNnBlvtrmjLM+abULI2oIMznSe42izXZSNFvbpM3PmzZO+ZpsPXLnI6VJMgLICX/PN89Z8EzUs6M2c2ba/iYykOK5fke10KSZAakIs1xRn8dzhFqdLMfPEgt7MicERN88fbuH9Vy4iIdbldDlmgruvyOdE2wDHW635JhpY0Js58cKRFobGPPzOhulMRmbmy/uvLMAVI/xi31mnSzHzwILezImfv32WwswkypdmOl2KmUROWgI3lWbzq31nbYrBKGBBb2ZVRWU93379JDtOdFCam8qPdzfYBCMh6neuLqSpd5hdpzqdLsXMMQt6M+sONvSgwPoiO5sPZXeW5ZGWEMvP37bmm0gXVNCLyF0iUiMitSLy2Um23yQib4uIW0Tum7DtYRE54f96eLYKN6FJVdnX0ENhZhI5aQlOl2MuoqKynp+/fZaV+WlsO9DEf7152umSzByaMuhFxAU8DtwNlAEPikjZhN3qgY8BFROOzQK+BGwGNgFfEhE7zYtgTT3DNPcOc/US+28OBxuKFpwfRtpErmDO6DcBtapap6qjwFbg3sAdVPW0qh4EJs5o8D7gRVXtUtVu4EXgrlmo24SoPWe6iHMJVxUucLoUE4Ti7BQyk+OoOt3ldClmDgUT9IuBhoDlRv+6YAR1rIg8IiJVIlLV3t4e5FObUHNu1M2Bhh6uKMggKd76zoeDGBGuKc6irmOQ2rYBp8sxcySYoJdJ1gXbHyuoY1X1CVUtV9XynJycIJ/ahJrfHGxmxO3lmuIsp0sx07BxaSYuEX6823pHRapggr4RKApYLgSagnz+mRxrwsxP9jSQnZrA0oXJTpdipiEtMY6ygnSe2dvI8JjH6XLMHAgm6PcApSJSIiLxwAPAtiCffztwp4hk+i/C3ulfZyLMidZ+qs50c01xJiKT/SFnQtnmkix6h8b49UEb6CwSTRn0quoGHsUX0EeBp1W1WkQeE5EtACJyjYg0AvcD3xGRav+xXcCX8f2y2AM85l9nIsx/vnWa+NgYNlhvm7BUkp3CspwUfrjrjNOlmDkgoTb5QHl5uVZVVTldhpmGnnOjXPsPL7PlqgK7SSqMjbo9/N//OcLP/vQ6NtrQFWFHRPaqavlk2+zOWDNjW/c0MDzm5Q+vL3G6FDMD95cXkZ4Yy5M76pwuxcwyC3ozI26Plx+8dZr3LFvImkXpTpdjZiAlIZaHNi/l+cMtNHSdc7ocM4ss6M2MvHCklabeYf7w+mKnSzGz4GPXFRMjwpM7TjldiplFsU4XYMKXqvKd10+yJCuZ29bkOV2OmaHxUUbXLc6gorKeosxkkuJdPLR5icOVmZmyM3pz2f7uf45woLGXjUsy+ckeG444UtxQms2ox8vOug6nSzGzxILeXBZV5dVjbWQkxbFhqY1rE0kWZSSxJj+NN2s77QaqCGFBby7LzrpOznSd46aVOcTG2I9RpLl1dR5DYx521tmkJJHAPqHmsnzr5VrSEmJtqsAItTgzidX5aew40UH/8JjT5ZgZsqA30/ZWbQc76zq5cWUOcS77EYpUt67OZWjMww922t2y4c4+pWZaVJWvba+hICORzSU2SmUkK8xMZlVeGt9+/STdg6NOl2NmwILeTMv26hYONPTwF3estLP5KPC+K/IZHHHzrVdqnS7FzIB9Uk3Q3B4vX99ew4rcVH53Q7Bzz5hwlp+eyP0bi/jvXaep77S7ZcOVBb2ZUkVlPRWV9fz1Tw9S1z7ItSVZPF3V6HRZZp585s6VxMbE8PXtx5wuxVwmC3oTlJExDy8ebWVpVrKNaRNl8tIT+eMbS/j1wWabWzZMWdCboLx+op2BETf3rFtkE4tEmYrKerJSEshIiuPPf7yP/7ZeOGHHgt5MqefcKDtOdLC+aAFFWTZNYDSKj43h/esW0dw7TOUpu4kq3AQV9CJyl4jUiEitiHx2ku0JIvIT//ZKEV8v1uYAAAz2SURBVCn2ry8WkSER2e//+vbslm/mw/bqFgDuLLOBy6LZ2oJ0VuSm8tLRVtr7R5wux0zDlEEvIi7gceBuoAx4UETKJuz2caBbVVcA/wx8LWDbSVVd7//6xCzVbebJntNdHGjs5cbSbBYkxztdjnGQiPDBKwsYcytf/vURp8sx0xDMGf0moFZV61R1FNgK3Dthn3uB7/sfPwPcJtaQG/Y8XuWLv6omIymOm1fmOl2OCQE5aQncsiqHbQeaePloq9PlmCAFE/SLgYaA5Ub/ukn38U8m3gss9G8rEZF9IvK6iNw4w3rNPKqoPMPR5j7uWbeI+Fi7nGN8bl6Vw6q8NL7wy8M2Dk6YCObTO9mZ+cQZxS+2TzOwRFU3AJ8BKkTkgr55IvKIiFSJSFV7e3sQJZm51jkwwj++cJzrVyzkigLrTmneERsTw9fuu5LWvmG++pz1rQ8HwQR9I1AUsFwINF1sHxGJBTKALlUdUdVOAFXdC5wEVk58AVV9QlXLVbU8Jydn+t+FmXVf2lbN0KiHv9uy1rpTmgusL1rAx28o4UeV9bxW0+Z0OWYKwQT9HqBUREpEJB54ANg2YZ9twMP+x/cBr6iqikiO/2IuIrIMKAVsivkQ9/zhFn59sJlP3baCFblpTpdjQtRf3rmKVXlp/PUzB+myQc9C2pRB729zfxTYDhwFnlbVahF5TES2+Hd7ElgoIrX4mmjGu2DeBBwUkQP4LtJ+QlXt1roQ1nNulC/88jBrC9L5k5uXO12OCVEVlfX8/O2z3Lk2j67BUT76vUp+tMtupApVojqxud1Z5eXlWlVV5XQZUUlV+eC3dnCkuY8/u2UFBQuSnC7JhIE3TrTz3OEWtlxVwL8+uMHpcqKWiOxV1fLJtllXCnPed9+o43BTH3eW5VvIm6BdvyKbVXlp/OZgM3vPdDtdjpmEBb0BfLNGffW5Y1xRkM6NpdlOl2PCSIwIv1deREZyHH/2o71212wIsqA31LT088mKt1mWk8r/urrQetmYaUuKd/HQpiX0nBvjEz/cy9Cox+mSTAAL+ih3orWfh767i/jYGL73B+UkxLmcLsmEqYIFSfzzh9fzdn03f/7jt3F7vE6XZPws6KPY8dZ+HvxuJTExQsUfX0txdorTJZkwd8+6RfzdlrW8dLSNz//iMF5vaHX2iFYW9FHq1Zo2fvff30IEfvzHm1mek+p0SSYCVFTWExsTw3tX5fCTqgbu//ZOPBb2jot1ugAzv1SVJ3ec4iu/OUp+RiK/f+1Sdp/qZvcp6y1hZs/ta/IQEV451saf//htvvnhDTZekoMs6KNI9+Aof/3MAV462kbZonTuLy8kIdba5M3sExFuX5NHYpyLZw81096/i3//yEZy0hKcLi0q2a/YKFFZ18nd//IGrx9v54sfKOMjm5dYyJs5d8OKbP71wQ0cOtvLB7+1g3319pejEyzoI5zHq3z8v/bwwBO7GPN4eeTG5STGuawLpZk3W64q4Od/ej1xscL9397JN186zpj1yJlXNgRCBDvW0sfnfn6IffU9rC9awL1XFVj3SeOYoVEP/3Owif0NPRQsSOQ/PrKRq4oWOF1WxLjUEAgW9BGof3iM/3jtJE/8to70pDhuX5PHevtAmRBR3dTLr/Y3MTDi5v6Nhfz1+1aRm57odFlh71JBbxdjI8jQqIeK3fU8/motXYOj/O7Vi/nC+8t4/nCL06UZc97aggyW56TS2j/MUztOse1AEx/ZvJRP3LzMAn+O2Bl9BGjoOsePKuv5/lunGRrzsCInlTvX5lGYmex0acZcUufACK/VtLOvoRsRYd3iDL70wTLWFy2w60jTZE03EcbrVWpa+/nt8XaePdTMgcZeYgTWLErnuuXZlNgdribMdA2OsqO2g3313Yy4vZTmpvLBqwr44FUF9vMcJAv6MDc06uFwUy/76rvZV9/DntNddAz4ZvS5sjCDe9Yt4oNXFfB6jc23a8LbyJiHuNgYth1oYvcp3xxF6xZn8IErF3Hn2nwL/UuwoA8jbo+X052DHGzsZV99D/saujnW3I/bfxt5UVYSG5dk4oqJYXlOCguS4x2u2Ji50Ts0xqHGHg409nK2ZwiAZTkp3LY6l9vW5LFxaSZxLushPm7GQS8idwH/AriA76nqVydsTwB+AGwEOoEPq+pp/7bPAR8HPMCnVHX7pV4rGoLe41U6BkZo6R2mpW+Yhq5zHG3up6a1j+OtA4y6fX2MU+Jd5GUksiQzmaIs31dqgl0/N9Gna3CUYy191LT0U9c+iEeVhNgYrl+RzeaSLNYVZlC2KD2qT3xm1OvGP7n348AdQCOwR0S2qeqRgN0+DnSr6goReQD4GvBhESnDN5n4WqAAeElEVqpqSAxWrap41fevAqqgKOO/+wKXlXfv5/Z4GRrzMDzmZXjMw4jbQ9+wm76hMXqHxug55/s38KtvaIyuwVE6BkaYOM5TTloCC5Li2FScRX5GIgUZSeSmJxBjF6SMISslnuuWZ3Pd8mxGxjycaBvgRFs/pzsHeeVY2/n9FqbEk5ueSF56ArlpCeSlJ5KRFEdqQizJCbGkxLtISYglJT6WlATXu8bfmeycV8Q3nEOMgCD+5Xcex4gg+P+N4fzjGHln33ft43+++RbM6eEmoFZV6wBEZCtwLxAY9PcC/9f/+Bng38T33dwLbFXVEeCUf/LwTcDO2Sn/HZ0DI9zwtVffFcxcIqjnQ7wrhqR4F0lxLhLjXCTFuyjKSmZtQTppiXFkJMWRnhhHRnKcnakbE6SEOBdXLM7gisUZAAyMuGnqGaKld5jOwRH6h92caB1g75luBobdhFbjtE+MvDv4x3+RXFWUwdZH3jPrrxdMuiwGGgKWG4HNF9tHVd0i0gss9K/fNeHYxRNfQEQeAR7xLw6ISE1Q1c+9bKDD6SJCgL0P77D3wsfeh3fM2ntxDPjJn1z24UsvtiGYoJ/s74yJvyQvtk8wx6KqTwBPBFHLvBKRqou1eUUTex/eYe+Fj70P7wiH9yKYS9aNQFHAciHQdLF9RCQWyAC6gjzWGGPMHAom6PcApSJSIiLx+C6ubpuwzzbgYf/j+4BX1NedZxvwgIgkiEgJUArsnp3SjTHGBGPKpht/m/ujwHZ83SufUtVqEXkMqFLVbcCTwH/7L7Z24ftlgH+/p/FduHUDnwyVHjdBCrnmJIfY+/AOey987H14R8i/FyF3w5QxxpjZZbeVGWNMhLOgN8aYCGdBHyQR+SsRURHJdroWJ4jIN0TkmIgcFJFfiEhUzWQiIneJSI2I1IrIZ52uxykiUiQir4rIURGpFpFPO12Tk0TEJSL7ROTXTtdyKRb0QRCRInxDQNQ7XYuDXgSuUNUrgePA5xyuZ94EDANyN1AGPOgf3iMauYG/VNU1wLXAJ6P4vQD4NHDU6SKmYkEfnH8G/g+T3OwVLVT1BVV1+xd34bsnIlqcHwZEVUeB8WFAoo6qNqvq2/7H/fhC7oK73aOBiBQC7we+53QtU7Ggn4KIbAHOquoBp2sJIX8EPOd0EfNosmFAojLcAolIMbABqHS2Esd8E98JoNfpQqZiI2kBIvISkD/Jps8DfwvcOb8VOeNS74Oq/sq/z+fx/fn+o/mszWFBDeURTUQkFfgZ8Beq2ud0PfNNRD4AtKnqXhG5xel6pmJBD6jq7ZOtF5F1QAlwwD+0aCHwtohsUtWIm3H7Yu/DOBF5GPgAcJtG1w0YNpRHABGJwxfyP1LVnztdj0OuB7aIyD1AIpAuIj9U1Y86XNek7IapaRCR00C5qkbdqH3+yWf+CbhZVaNqzkL/+E3HgduAs/iGBXlIVasdLcwB/uHHvw90qepfOF1PKPCf0f+Vqn7A6VouxtroTbD+DUgDXhSR/SLybacLmi/+i9Djw4AcBZ6OxpD3ux74feBW/8/Bfv9ZrQlhdkZvjDERzs7ojTEmwlnQG2NMhLOgN8aYCGdBb4wxEc6C3hhjIpwFvYl6IjIwxfZiETk8zef8LxG5b2aVGTM7LOiNMSbCWdAb4yciqSLysoi8LSKHRCRwhMpYEfm+fzz+Z0Qk2X/MRhF5XUT2ish2EVnkUPnGXJQFvTHvGAZ+R1WvBt4L/D//Lf8Aq4An/OPx9wF/5h/z5VvAfaq6EXgK+IoDdRtzSTaomTHvEODvReQmfEPPLgby/NsaVPVN/+MfAp8CngeuwDcsBIALaJ7Xio0JggW9Me/4CJADbFTVMf8gdon+bRPHClF8vxiqVfU981eiMdNnTTfGvCMD3xjjYyLyXmBpwLYlIjIe6A8CO4AaIGd8vYjEicjaea3YmCBY0Bvzjh8B5SJShe/s/ljAtqPAwyJyEMgC/sM/reB9wNdE5ACwH7hunms2Zko2eqUxxkQ4O6M3xpgIZ0FvjDERzoLeGGMinAW9McZEOAt6Y4yJcBb0xhgT4SzojTEmwv1/XuRclza65O0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "lab = inp_df['label'].dropna()\n",
    "sns.distplot(lab)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "id": "370d1d51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>p_devtest</th>\n",
       "      <th>fdr_p_devtest</th>\n",
       "      <th>ref_site</th>\n",
       "      <th>alt_site</th>\n",
       "      <th>b_intercept</th>\n",
       "      <th>p_intercept</th>\n",
       "      <th>b_ConditionALS_pathology</th>\n",
       "      <th>p_ConditionALS_pathology</th>\n",
       "      <th>strand</th>\n",
       "      <th>multiplyer</th>\n",
       "      <th>LFC_PA_Usage</th>\n",
       "      <th>negative_logFDR</th>\n",
       "      <th>switch_width</th>\n",
       "      <th>bed</th>\n",
       "      <th>correction_multiplyer</th>\n",
       "      <th>switch_name</th>\n",
       "      <th>sig_LFC_PA_Usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>0.261215</td>\n",
       "      <td>4.734204e-02</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>3.376034e-07</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>6.471593</td>\n",
       "      <td>16644</td>\n",
       "      <td>chr12,125124284,125140928</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125140928:+</td>\n",
       "      <td>0.983355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125143316:+</td>\n",
       "      <td>1.641746</td>\n",
       "      <td>4.866807e-52</td>\n",
       "      <td>0.263182</td>\n",
       "      <td>1.246049e-01</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.263182</td>\n",
       "      <td>0.904465</td>\n",
       "      <td>19032</td>\n",
       "      <td>chr12,125124284,125143316</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125143316:+</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>chr12:125143316:+</td>\n",
       "      <td>1.380531</td>\n",
       "      <td>8.547748e-46</td>\n",
       "      <td>-0.720173</td>\n",
       "      <td>1.476861e-08</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.720173</td>\n",
       "      <td>7.830660</td>\n",
       "      <td>2388</td>\n",
       "      <td>chr12,125140928,125143316</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125140928:125143316:+</td>\n",
       "      <td>-0.720173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67201672:-</td>\n",
       "      <td>0.936799</td>\n",
       "      <td>7.468705e-06</td>\n",
       "      <td>-0.043726</td>\n",
       "      <td>8.609305e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.043726</td>\n",
       "      <td>0.065032</td>\n",
       "      <td>644</td>\n",
       "      <td>chr15,67201028,67201672</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67201672:-</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67202710:-</td>\n",
       "      <td>1.470439</td>\n",
       "      <td>7.400884e-14</td>\n",
       "      <td>-1.337591</td>\n",
       "      <td>1.070773e-07</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.337591</td>\n",
       "      <td>6.970303</td>\n",
       "      <td>1682</td>\n",
       "      <td>chr15,67201028,67202710</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67202710:-</td>\n",
       "      <td>1.337591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41174</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>-0.321971</td>\n",
       "      <td>2.534728e-02</td>\n",
       "      <td>0.526766</td>\n",
       "      <td>5.576124e-03</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.526766</td>\n",
       "      <td>2.253668</td>\n",
       "      <td>53729</td>\n",
       "      <td>chr1,77564505,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77618234:-</td>\n",
       "      <td>-0.526766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41175</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.016713</td>\n",
       "      <td>8.988251e-01</td>\n",
       "      <td>-0.225352</td>\n",
       "      <td>2.362558e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.225352</td>\n",
       "      <td>0.626618</td>\n",
       "      <td>65335</td>\n",
       "      <td>chr1,77564505,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77629840:-</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41176</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>-0.221592</td>\n",
       "      <td>1.320936e-01</td>\n",
       "      <td>0.933971</td>\n",
       "      <td>6.105239e-06</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.933971</td>\n",
       "      <td>5.214297</td>\n",
       "      <td>37236</td>\n",
       "      <td>chr1,77580998,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77580998:77618234:-</td>\n",
       "      <td>-0.933971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41177</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.117093</td>\n",
       "      <td>3.854144e-01</td>\n",
       "      <td>0.181853</td>\n",
       "      <td>3.789934e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.181853</td>\n",
       "      <td>0.421368</td>\n",
       "      <td>48842</td>\n",
       "      <td>chr1,77580998,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77580998:77629840:-</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41178</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.338685</td>\n",
       "      <td>1.825640e-02</td>\n",
       "      <td>-0.752118</td>\n",
       "      <td>1.128067e-04</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.752118</td>\n",
       "      <td>3.947665</td>\n",
       "      <td>11606</td>\n",
       "      <td>chr1,77618234,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77618234:77629840:-</td>\n",
       "      <td>0.752118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41179 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript  p_devtest  fdr_p_devtest           ref_site  \\\n",
       "0           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "1           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "2           AACS   0.006145       0.019492  chr12:125140928:+   \n",
       "3          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "4          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "...          ...        ...            ...                ...   \n",
       "41174       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41175       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41176       ZZZ3   1.000000       1.000000    chr1:77580998:-   \n",
       "41177       ZZZ3   1.000000       1.000000    chr1:77580998:-   \n",
       "41178       ZZZ3   1.000000       1.000000    chr1:77618234:-   \n",
       "\n",
       "                alt_site  b_intercept   p_intercept  b_ConditionALS_pathology  \\\n",
       "0      chr12:125140928:+     0.261215  4.734204e-02                  0.983355   \n",
       "1      chr12:125143316:+     1.641746  4.866807e-52                  0.263182   \n",
       "2      chr12:125143316:+     1.380531  8.547748e-46                 -0.720173   \n",
       "3       chr15:67201672:-     0.936799  7.468705e-06                 -0.043726   \n",
       "4       chr15:67202710:-     1.470439  7.400884e-14                 -1.337591   \n",
       "...                  ...          ...           ...                       ...   \n",
       "41174    chr1:77618234:-    -0.321971  2.534728e-02                  0.526766   \n",
       "41175    chr1:77629840:-     0.016713  8.988251e-01                 -0.225352   \n",
       "41176    chr1:77618234:-    -0.221592  1.320936e-01                  0.933971   \n",
       "41177    chr1:77629840:-     0.117093  3.854144e-01                  0.181853   \n",
       "41178    chr1:77629840:-     0.338685  1.825640e-02                 -0.752118   \n",
       "\n",
       "       p_ConditionALS_pathology strand  multiplyer  LFC_PA_Usage  \\\n",
       "0                  3.376034e-07      +           1      0.983355   \n",
       "1                  1.246049e-01      +           1      0.263182   \n",
       "2                  1.476861e-08      +           1     -0.720173   \n",
       "3                  8.609305e-01      -          -1      0.043726   \n",
       "4                  1.070773e-07      -          -1      1.337591   \n",
       "...                         ...    ...         ...           ...   \n",
       "41174              5.576124e-03      -          -1     -0.526766   \n",
       "41175              2.362558e-01      -          -1      0.225352   \n",
       "41176              6.105239e-06      -          -1     -0.933971   \n",
       "41177              3.789934e-01      -          -1     -0.181853   \n",
       "41178              1.128067e-04      -          -1      0.752118   \n",
       "\n",
       "       negative_logFDR  switch_width                        bed  \\\n",
       "0             6.471593         16644  chr12,125124284,125140928   \n",
       "1             0.904465         19032  chr12,125124284,125143316   \n",
       "2             7.830660          2388  chr12,125140928,125143316   \n",
       "3             0.065032           644    chr15,67201028,67201672   \n",
       "4             6.970303          1682    chr15,67201028,67202710   \n",
       "...                ...           ...                        ...   \n",
       "41174         2.253668         53729     chr1,77564505,77618234   \n",
       "41175         0.626618         65335     chr1,77564505,77629840   \n",
       "41176         5.214297         37236     chr1,77580998,77618234   \n",
       "41177         0.421368         48842     chr1,77580998,77629840   \n",
       "41178         3.947665         11606     chr1,77618234,77629840   \n",
       "\n",
       "       correction_multiplyer                       switch_name  \\\n",
       "0                          1  chr12:AACS:125124284:125140928:+   \n",
       "1                          1  chr12:AACS:125124284:125143316:+   \n",
       "2                          1  chr12:AACS:125140928:125143316:+   \n",
       "3                          1   chr15:AAGAB:67201028:67201672:-   \n",
       "4                          1   chr15:AAGAB:67201028:67202710:-   \n",
       "...                      ...                               ...   \n",
       "41174                      1     chr1:ZZZ3:77564505:77618234:-   \n",
       "41175                      1     chr1:ZZZ3:77564505:77629840:-   \n",
       "41176                      1     chr1:ZZZ3:77580998:77618234:-   \n",
       "41177                      1     chr1:ZZZ3:77580998:77629840:-   \n",
       "41178                      1     chr1:ZZZ3:77618234:77629840:-   \n",
       "\n",
       "       sig_LFC_PA_Usage  \n",
       "0              0.983355  \n",
       "1              0.000000  \n",
       "2             -0.720173  \n",
       "3              0.000000  \n",
       "4              1.337591  \n",
       "...                 ...  \n",
       "41174         -0.526766  \n",
       "41175          0.000000  \n",
       "41176         -0.933971  \n",
       "41177          0.000000  \n",
       "41178          0.752118  \n",
       "\n",
       "[41179 rows x 18 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2_name = data_root + 'Excitatory' + \"_C9ALSvsCTRL/APAlog_res_metadata_added.tsv\"\n",
    "inp_df2 = pd.read_csv(df2_name, sep='\\t')\n",
    "inp_df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "b9d18b70",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdc0bced1d0>"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import seaborn as sns\n",
    "lab2 = inp_df2['sig_LFC_PA_Usage'].dropna()\n",
    "sns.distplot(lab2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "id": "f58bc617",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.axes._subplots.AxesSubplot at 0x7fdc0b77d350>"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXcAAAEHCAYAAABV4gY/AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjEsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy8QZhcZAAAgAElEQVR4nO3deXxc9Xnv8c8zi1bLlozlBe+ELQlhMWZJQy40S6+hacgtSYC8QkOa1m22kt62t1l60zS397a9TZOmJU0uDdyElKYUskHKmhvC0oJBNhhvLMY2WHiTLdmyrHXmPPePc0YeyaPNHnmkc77v10sva84czTySNd/56Xd+5znm7oiISLykKl2AiIiUn8JdRCSGFO4iIjGkcBcRiSGFu4hIDGUq9cRz5szxZcuWVerpRUSmpbVr1+539+ax9qtYuC9btoyWlpZKPb2IyLRkZq+OZz9Ny4iIxJDCXUQkhhTuIiIxpHAXEYkhhbuISAwp3EVEYkjhLiISQwp3EZEYUrhL4n31oRf5yoMvVroMkbKq2BmqIlPFU9vacXTRGokXhbskXt4dXZFM4kbhLomXDzRul/hRuEviBe5o4C5xo3CXxMsHTqBwl5hRuEvi5QON3CV+FO6SeIFr5C7xo3CXxNPIXeJI4S6JF3g4eheJE4W7JF54QFXhLvGicJfEC6dlFO4SLwp3SbzwgKrCXeJF4S6Jp3XuEkdjdoU0sxoze9rM1pvZJjP7sxL7VJvZnWa21czWmNmyyShWZDJo5C5xNJ6Wv33AO9z9POB8YJWZXTpsn48BHe5+OvA14K/KW6bI5MkHTl5Dd4mZMcPdQ13RzWz0MfyVcDXw3ejzu4F3mpmVrUqRSRQ4CneJnXFdrMPM0mb2HLAPeNjd1wzbZSGwE8Ddc8Ah4JQSj7PazFrMrKWtre3EKhcpk0Ajd4mhcYW7u+fd/XxgEXCxmZ0zbJdSo/RjXi3ufou7r3T3lc3NzROvVmQS5N3Ja85dYmZCl9lz94PAL4BVw+5qBRYDmFkGmAW0l6E+kUmXD5xAI3eJmfGslmk2s8bo81rgXcALw3a7B/hI9Pn7gZ+7zgqRaSLQyF1iaDzr3BcA3zWzNOGbwb+6+0/N7MtAi7vfA9wKfM/MthKO2K+btIpFyqzQOMzd0ToAiYsxw93dnwcuKLH9i0Wf9wIfKG9pIpPPi9r95gMnk1a4SzxMaM5dJG6Kp9o1NSNxonCXRCteAhkEFSxEpMwU7pJoxW0HNHKXOFG4S6IVj9zzeYW7xIfCXRItr5G7xJTCXRKt+OQltSCQOFG4S6INOaCqkbvEiMJdEm3ItIxG7hIjCndJtOLljwp3iROFuyRa8chd0zISJwp3STQdUJW4UrhLoumAqsSVwl0SbegB1QoWIlJmCndJtOJpmZyay0iMKNwl0YYcUFW2S4wo3CXRhvSW0Zy7xIjCXRJN69wlrhTukmha5y5xpXCXRMtrnbvElMJdEi0YckBV4S7xoXCXRAt0QFViSuEuiaaukBJXY4a7mS02s0fMbIuZbTKzm0rsc4WZHTKz56KPL05OuSLlVbxaRgdUJU4y49gnB/yBu68zswZgrZk97O6bh+33uLu/p/wlikwetR+QuBpz5O7uu919XfT5YWALsHCyCxM5GYZ2hVS6S3xMaM7dzJYBFwBrStz9VjNbb2b3m9mby1CbyKQbuhSygoWIlNl4pmUAMLMZwA+Az7h757C71wFL3b3LzK4CfgycUeIxVgOrAZYsWXLcRYuUy5BpGc25S4yMa+RuZlnCYL/D3X84/H5373T3rujz+4Csmc0psd8t7r7S3Vc2NzefYOkiJ654Wkbr3CVOxrNaxoBbgS3u/tUR9pkf7YeZXRw97oFyFioyGbQUUuJqPNMybwNuADaY2XPRts8DSwDc/VvA+4GPm1kO6AGuc9ffuDL1qSukxNWY4e7uTwA2xj43AzeXqyiRk0XtBySudIaqJFrxChmN3CVOFO6SaDqgKnGlcJdE0wFViSuFuyRafsgFshXuEh8Kd0m0QFdikphSuEuiqf2AxJXCXRKtONw1cpc4UbhLogU6oCoxpXCXRBuyzl3hLjGicJdE0wFViSuFuyRaYbRuppG7xIvCXRKtEOjZdErtByRWFO6SaIWpmEzK1H5AYkXhLomWD5x0ykib6QxViRWFuyRa3p20GSmN3CVmFO6SaEHgpFKQTpnm3CVWFO6SaPkA0mZhuKv9gMSIwl0SLXAnFc25a1pG4kThLokWeHRAVdMyEjMKd0m0fFA4oKorMUm8KNwl0YqnZTRylzhRuEuiHR25m9oPSKwo3CXR8gGDJzGpcZjEyZjhbmaLzewRM9tiZpvM7KYS+5iZ/Z2ZbTWz581sxeSUK1JegTtm0Tp3jdwlRjLj2CcH/IG7rzOzBmCtmT3s7puL9rkSOCP6uAT4ZvSvyJRWaD+QMoW7xMuYI3d33+3u66LPDwNbgIXDdrsauN1DTwGNZrag7NWKlFmh/YBG7hI3E5pzN7NlwAXAmmF3LQR2Ft1u5dg3AJEpJ2w/UFjnXulqRMpn3OFuZjOAHwCfcffO4XeX+JJjXipmttrMWsyspa2tbWKVikyCwmqZtBqHScyMK9zNLEsY7He4+w9L7NIKLC66vQjYNXwnd7/F3Ve6+8rm5ubjqVekrIasc1e4S4yMZ7WMAbcCW9z9qyPsdg/wG9GqmUuBQ+6+u4x1ikyK8IAqpFLoJCaJlfGslnkbcAOwwcyei7Z9HlgC4O7fAu4DrgK2At3AR8tfqkj55f1oV8jcgNpCSnyMGe7u/gSl59SL93Hgk+UqSuRkKRxQTan9gMSMzlCVRNMBVYkrhbskWl6NwySmFO6SaEFR47CcFrpLjCjcJdHyhYt1qHGYxIzCXRJtyBmqmnOXGFG4S6KFvWXCrpDKdokThbsk2mA/d43cJWYU7pJoQeCkTC1/JX4U7pJogwdUU+iAqsSKwl0STQdUJa4U7pJohYt1pLQUUmJG4S6JVrjMnkbuEjcKd0k0dwYPqOYU7hIjCndJtEI/dzUOk7hRuEuiHV0to8ZhEi8Kd0m04nXuga7VITGicJdEK4zcMxq5S8wo3CXR8oWRu1bLSMwo3CXRguBoy9/CbZE4ULhLouXdSRmkU0dvi8SBwl0SLQgIL5CdCkfumpqRuFC4S6IV2g8MTsto5C4xoXCXRCtuP1C4LRIHY4a7md1mZvvMbOMI919hZofM7Lno44vlL1Ok/AoHTwvr3EHhLvGRGcc+3wFuBm4fZZ/H3f09ZalI5CQpHDzVyF3iaMyRu7s/BrSfhFpETqpCkKeLD6hqzl1iolxz7m81s/Vmdr+ZvXmkncxstZm1mFlLW1tbmZ5a5PgUDp6mLDxDFVALAomNcoT7OmCpu58H/D3w45F2dPdb3H2lu69sbm4uw1OLHL+jI3cGV8to5C5xccLh7u6d7t4VfX4fkDWzOSdcmcgkK4zSC+0Hwm0Kd4mHEw53M5tvFg57zOzi6DEPnOjjiky2oQdUo20Kd4mJMVfLmNn3gSuAOWbWCvwpkAVw928B7wc+bmY5oAe4zl1/28rUN+SAqqZlJGbGDHd3v36M+28mXCopMq0UH1BNa1pGYmY869xFYqkwcl+7o4PaqjQA967fzfxZHXzokiWVLE3khKn9gCRWIdzNIBq4q7eMxIbCXRKreFrG1DhMYkbhLolVauSubJe4ULhLYmnkLnGmcJfEykcnMYUj90K4V7AgkTJSuEti5Ye0/A236RQNiQuFuyTW0WkZiqZlKlmRSPko3CWxCuFuGrlLDCncJbGOTstozl3iR+EuiVU8cjeN3CVmFO6SWPnilr9aCikxo3CXxCo+ickG2w9UsCCRMlK4S2IVn8SkkbvEjcJdEqv4gKqp/YDEjMJdEqtwYQ4DjdwldhTuklhBULzOPQx3ZbvEhcJdEqtU+wGN3CUuFO6SWEfXuR9tP6Bsl7hQuEtiDV3nHn4eoHSXeFC4S2LlS4zctc5d4kLhLokVlJpzV7pLTCjcJbFKNQ5TbxmJizHD3cxuM7N9ZrZxhPvNzP7OzLaa2fNmtqL8ZYqUX75E4zAN3CUuxjNy/w6wapT7rwTOiD5WA9888bJEJl+gkbvE2Jjh7u6PAe2j7HI1cLuHngIazWxBuQoUmSwlR+4VrEeknMox574Q2Fl0uzXadgwzW21mLWbW0tbWVoanFjl+pUbuOolJ4qIc4W4ltpV8hbj7Le6+0t1XNjc3l+GpRY5fXu0HJMbKEe6twOKi24uAXWV4XJFJlY+CvLAMMmUauUt8lCPc7wF+I1o1cylwyN13l+FxRSZV8Tp3CEfwynaJi8xYO5jZ94ErgDlm1gr8KZAFcPdvAfcBVwFbgW7go5NVrEg5FZ+hChq5S7yMGe7ufv0Y9zvwybJVJHKS5DVylxjTGaqSWEFw7Mg9r3SXmFC4S2LlfdjIHdNJTBIbCndJrEKrgcJa3nDOvWLliJSVwl0SKwgc42i735Rp5C7xoXCXxMq7D07JQDj3rpG7xIXCXRIrCJyibCeV0shd4kPhLomVD4aO3FNmGrlLbCjcJbHyPnTkbugkJokPhbskVlBi5K5sl7hQuEtiHTNyV/sBiRGFuyRWPkBz7hJbCndJrGNWy5gusyfxoXCXxDp2nbtpWkZiQ+EuiVV65F65ekTKSeEuiaWRu8SZwl0SKx/1lilQ4zCJE4W7JFbgpda5K90lHhTuklj5Y+bctRRS4kPhLok1fJ27aSmkxIjCXRIrnJY5elsjd4kThbskVjgtM7yfu9Jd4kHhLolVauSubJe4ULhLYmnkLnE2rnA3s1Vm9qKZbTWzz5a4/0YzazOz56KP3yp/qSLlFV6s4+jt0ebce/rzfOKOtezYf+TkFCdygsYMdzNLA98ArgTeBFxvZm8qseud7n5+9PHtMtcpUnaBHztyH2m1zJY9ndy3YQ8/fu71k1WeyAkZz8j9YmCru29z937gX4CrJ7cskclXeuReOtz3dfYCsGZb+8koTeSEjSfcFwI7i263RtuGu8bMnjezu81scakHMrPVZtZiZi1tbW3HUa5I+eT92H7uI0257+3sA2Ddax305fInozyREzKecLcS24a/BO4Flrn7ucDPgO+WeiB3v8XdV7r7yubm5olVKlJmpbpCjjhyPxyO3PtyAc+3HgLghT2drL69he7+3KTXKjJR4wn3VqB4JL4I2FW8g7sfcPe+6OY/AheWpzyRyZMPju0KOdrIvaE6A8CabQcA+Pufb+WhzXt5erumamTqGU+4PwOcYWbLzawKuA64p3gHM1tQdPO9wJbylSgyOYYfUB195N7H8uZ6zp7fwJrt7ezr7OXBjXsAWPtqx0mpV2QiMmPt4O45M/sU8CCQBm5z901m9mWgxd3vAX7PzN4L5IB24MZJrFmkLAJ3sqnh/dxL77uvs5dFTXV09eVYs62dP/7B8+QCZ1ZtlpYdCneZesYMdwB3vw+4b9i2LxZ9/jngc+UtTWRy5QOnKjP+kfuKpU00ezVPbTvA4y/v54y5M5jTUM2zr3UwkA/IpnVOoEwd+m2U2AoCZ0N08LPk/c642g/05wLaj/Qzr6GGZafUAZALnEtPO4Wls+voHQjYvKuz3OWLnBCFu8TWvc/v4tdufoKt+7pK3n/sAdXSI/e2rnCtwNyZ1TTUZGluqKaxLstZ8xtYeko9AC2ad5cpRuEusfXEy/sB2D5Cy4BjL7NXeuS+NzqBad7MagCuXbmYGy5dSsqMWbVZFjXV0rLj6IqZ/lxQnm9A5AQo3CW21kRLFFs7ukveP/wyeyON3PdFJzDNbagB4NTGWhbMqh28f+XSJlpe7SCXD/jt21tY9bePEagxvFSYwl1iafehHl5rD0O9taOn5D6lLrPnHNtfpnAC09xo5D7chctm03a4j9XfW8vDm/eybf8RNrw+8ly/yMmgcJdY6B0IuzZu2R0e2CycWFSVSfH6COF+7AWyw3+Hj7n3dfaRMjilvnS4X7SsCYCfv7CPG39pGWbwyIv7TuC7ETlxCneJhTXb27lvwx7+5qEXAXhqWzsNNRkuWT6b1oOlp2VKjdzh2KmZvZ29NDdUk06V6sQBLTs6mF1fxSXLZ3PG3Bksaqzl7rWtZfiuRI7fuNa5i0x1hZYAP9uyj799+CV+tnkvp86qpS8XsK1tlAOqw9oPAMccVN13uG9wvr2UlBn/9d1nDr45nDW/gf+3ZR/7u/qYM6P0aF9ksmnkLrHw9PZ2Tp87g0zKeGDTHtq6+lg+p56muiq6+/Mc6Tu2udex69wL248duc8bYb796NcefaCz5s3EgcdeUudTqRyFu0x7Pf151rce5F1vnMeKpU28sOcwAMvn1NNYlwXg9YPhvPtAPuBQzwBQunEYQBDADbeu4V9bwk7XbYf7mDtz5JH7cAsaa5hRneEXLyrcpXIU7jLtPbuzg4G8c8ny2Vx2+hwMqEqnOLWxlqa6KuDocshvPLKVd3/1UYLAyfuxLX8BXj1whMdf3s83HtlKXy7PgSP9zG0Y//RKyowz5zXw6Ett5LUkUipEc+4y7a3Z1k7K4MJlTew+1MtFy2fjDumUDY7cC8shH32pjX2H+9jT2Uswwsh9fetBAF490M1P1+8GYN4ERu4Qzruve62Dq77+ODVVaa4+71R+87LlJ/y9ioyXRu4y7T29vZ03nTqTmTVhkL/v/IX8lwvCi4XNqM6QSRmvd/TQ059nY7T+/JW2rhFH7lt2H+b0uTOoyab4i/tfCLft6uSf17w27prOmtfANSsWsaipls6eAb768Esl5/1FJovCXaal25/cwQMbd9OXy7PutQ4uWX5Kyf1SFo7eWzt6WN96kIF8OE2ydV8XPvwye1Ezgv58wIoljZy3qJH9UV+ZhtrshOqryqT4mw+ex603XsRXPnAuXX057l2/a+wvFCkTTcvItPPqgSN88SebAFh2Sh19uYC+gfyII+vGuipaD/YMXlSjNpvmpb1hM7HikXvh80zKOHv+TOY21Ay2MGiomfhLpVCPuzO3oZq///lWAocPXbJkwo8lMlEaucu0c+czOzHgstPnDLYYKHRnLKWpLsvrHd08s6OdM+bO4OwFDby8N1xRM/wC2QBnzmugJpvm1MZaTp1VgxFO7xwvM+Pi5bN5/WDP4KodkcmmkbtMKwP5gLvWtnLW/AauessCzlvUSHt3P/WjhG9jXRX7uzp4Zns77z1/IQP5gIc2hZfIKx7dpKIbb1k4a3DbqnMWsG1/15A3geNxweImHti4h2e2t3Owu5/Nuzp5y6JZNNRMbLpHZLwU7jLlffc/drDrUA9/9Ctn8cgL+2g73MeqN88HYGFTLQubakf9+qZoxcyR/jwXLWtib2cfnb3hwc3iM1SXz5nBJctn88YFMwe3nT53BqfPnXHC30NtVZpzF83imR3tnP/lhwFY9eb5fOsGXUteJofCXaa01w508+f/tpmBvLNj/xG6+/PMbajmzHkN436Mxtqqwc9XLp3NC3uOXjWp+AzVWbVZrj5/YVnqLuWKM+cSOLz7TfPYc6iX7z31Kk+8vJ/Lzpgzac8pyaVwlyntU99fB8A7zp7Lg5v2AnDFmc0jNvEqpak+DPeGmgyPv9zG/q7+wfvsBKdbJmJOQzUfXLmYD12yhN6BPI++1MaX7t3E/Te9nWw6xeHeAU3TSNko3GVKuX/Dbu5s2TnYiOv51kNcfmYz73rjPGbXV/H4y21ctHz2hB6zoSZD2oylp9RjZsyur4ouhj105H6yFFbRXH5mM9976lV+/R/+g47uflo7evjz953Dhy9devKLkthRuEtFuR/tzLizvZs/uvt5uvpyPPpSG/Nn1lBXlebyM5sBWLGkiRVLmib8HCkzfn3FQubPCs8yTaeMU+qraevqO6kj9+HOnt/AWfMa2PD6IRbMquH8xY186Z5NnDW/gYuWzSYInM7eARrrqsZ+MJFhFO5SMT9c18qf/9sWPn75G6jJprn1iW0M5AM+/Y7TeWZHB2u2HeDXzjuVmmz6hJ/rgmFvCs0NYbif6CqYE2FmXHfxYg735pgzo5pfPXcBV9/8BB//p3X87uWn8c9Pv8ZrB7r52rXn82vnnVqxOmV6suGXFCu5k9kq4OtAGvi2u//lsPurgduBC4EDwLXuvmO0x1y5cqW3tLQcZ9ky3TzfepA/+fFGLjt9Dp/45dN55IV93PQvzzK7vpr9XX001mU52D3ANSsWceHSMIh7B/JlCfZSHty0h0dfauPXL1jIymUTm+aZTHs7e/nmL16hPx+wqKkWI+xo+TcfPI8z5jZw+5M76B0I+OyVZ3NqY7hK6EhfjkzaqM5Mzs9KphYzW+vuK8fab8yRu5mlgW8A7wZagWfM7B5331y028eADnc/3cyuA/4KuPb4Sh9dEF0953j+nC6eAjjRry28KRZvK1VbPvAhB//cnYG8U5VJDdknHwzd1juQJ2U2uM3d6ezJUVOVGnwR56L2tTNrs2TT4X49/XkO9w1wSn145aAgcNq6+ujPBSyYVUMmnaIvl2dnew9V6RQLm2pJp4yOI/1s23+EU+qrWDy7DoCX9x1me9sR3jB3Bm9onkF3f46nt7ez62APFy6dzdnzG3h2Zwd3tbTSM5DnfRcsZOXSJu58Zie3PbGdpvoqPnbZcgKHL/xoA7VVaf7hF6/w3Sdfpac/x5LZddz4S8vZuOsQ967fxTmnzmTFksbBn8FkBTtAc3QRjUpOy5Qyb2YNv3P5aeTyzuLZdfTnAm5/cge/f+d6IOx2CfDQ5j388aqz2fD6IX76/G5m1mS46Z1n8L4LFvLw5r3ct2EPb2iu5wMrF7OwsZYntu7nuZ0dvGVhI28/Yw7ZdIoX9nSy51Avb1wwk0VNtfTnA17Zd4TeXJ4z5zUwozoz+LtSnQm7bKZTRu9Anl0He5hVm2V2fRVmRk9/nv1dfTQ3VA/+vw3kA3oG8jRUZwZ/zr0DeQJ36qqORk/vQJ50ygZ/hwH6cwHZtI35WnKHVNG2XD7AYchjncjrfjobc+RuZm8FvuTu/zm6/TkAd/+Lon0ejPZ50swywB6g2Ud58OMduT+wcTefuGMdM6ozzKjOMBA4Xb05BvIBM2oy1Fdl6B3Iczhq0tRQnaEmm6arL8fh3gGy6RQNNVmq0sbhvhzd/XlqMuE2xzncm6N3IE99dYaG6gz9+XDeM5cPmFmbpb4qw5H+HJ09A5gZM2sy1GbTHO7NcbgvRzZtzKqtIps2DnYP0DOQpzqTorEuS+BwsLufgbxTm03TWJeldyDPwZ4B3MOzIGfWZOjszdEV1T+zJkNdVYb2I/305wMAGuvCMN/f1Yd7eNr8KfXV5IKAg91hr/JwXrmKgz0D9OeCIdv2d/VR6ERblU4xoyZ8/IKabIqUGd39+cFt9VVpenPBkBa21ZkUfbmAuqo01ZkUHd0Dgwcql8+p50hfjn2Hw94sy+fUc/3FS+g40s/9G/dgBjdcunQwCPpzAZm0nbRpkp3t3Xzz0Vf4wIWLjpmymWr6cwEPbt5DU10VFy5pors/x11rW3mtvZuqTIrzFzfSdriP7fuPYITXgJ1Vm+Vw7wBB1B2z+P+t8AZR+H2C8PfsSH9+yH5zZlTRfqR/yO/KzNrsYL8dCH9nqzKpIb8/p9RX4TC4rSqdYnZ9FUf6cxyOzi+oq0rTWJvlYM/A4O/ZrNosdVXpwddNJmU01lWRTsGhngF6BwJqs2lm1WbJBU5nzwD9+YCGmgwza7J09eUGe/UXthVeq5l0iplFr/sjfTlqsmkaajIYxuHe8DnrqzI01IS5crh3gFzeh+RKZ2/h8bPUZtMc6c/R1ZsjG72OsqnSudLVm6M3eq3MrMlyw1uX8ruXv+G4fh/GO3IfT7i/H1jl7r8V3b4BuMTdP1W0z8Zon9bo9ivRPvuHPdZqYHV08yzgxfF/S5NuDrB/zL2mhulS63SpE1TrZJkutU6XOgHOcvcxT/QYzwHVUkOp4e8I49kHd78FuGUcz3nSmVnLeN4Np4LpUut0qRNU62SZLrVOlzohrHU8+42ncVgrsLjo9iJgeO/SwX2iaZlZQPt4ChARkfIbT7g/A5xhZsvNrAq4Drhn2D73AB+JPn8/8PPR5ttFRGRyjTkt4+45M/sU8CDhUsjb3H2TmX0ZaHH3e4Bbge+Z2VbCEft1k1n0JJmS00UjmC61Tpc6QbVOlulS63SpE8ZZ67jWuYuIyPSii3WIiMSQwl1EJIYU7sOY2afN7EUz22Rm/7vS9YzFzP7QzNzMpmRTcDP7azN7wcyeN7MfmVnj2F91cpnZquj/fKuZfbbS9ZRiZovN7BEz2xL9bt5U6ZrGYmZpM3vWzH5a6VpGY2aNZnZ39Hu6JTpxc8oxs9+P/u83mtn3zaxmtP0V7kXM7JeBq4Fz3f3NwFcqXNKozGwxYVuI0leGnhoeBs5x93OBl4DPVbieIYraa1wJvAm43szeVNmqSsoBf+DubwQuBT45RessdhOwpdJFjMPXgQfc/WzgPKZgzWa2EPg9YKW7n0O4uGXUhSsK96E+Dvylu/cBuPu+Ctczlq8B/40SJ4xNFe7+kLvnoptPEZ4nMZVcDGx1923u3g/8C+Eb/JTi7rvdfV30+WHCAJq8y0adIDNbBPwq8O1K1zIaM5sJ/CfCFX+4e7+7H6xsVSPKALXRuUR1HHu+0RAK96HOBN5uZmvM7FEzu6jSBY3EzN4LvO7u6ytdywT8JnB/pYsYZiGws+h2K1M4NAHMbBlwAbCmspWM6m8JBx7BWDtW2GlAG/B/oymkb5tZfaWLGs7dXyecSXgN2A0ccveHRvuaxPVzN7OfAfNL3PUFwp9HE+GfvRcB/2pmp1XqhKwxav088Csnt6LSRqvT3X8S7fMFwqmFO05mbeMwrtYZU4WZzQB+AHzG3TvH2r8SzOw9wD53X2tmV1S6njFkgBXAp919jZl9Hfgs8N8rW9ZQZtZE+BflcuAgcJeZfdjd/2mkr0lcuLv7u0a6z8w+DvwwCvOnzSwgbCjUdrLqKzZSrWb2FsL/5PVRK9NFwDozu9jd95zEEoHRf6YAZvYR4D3AO6fgmcvjaa8xJZhZljDY73D3H1a6nlG8DXivmV0F1AAzzeyf3P3DFa6rlM1jwoMAAAT9SURBVFag1d0LfwXdTRjuU827gO3u3gZgZj8EfgkYMdw1LTPUj4F3AJjZmUAVU7BTnLtvcPe57r7M3ZcR/oKuqESwjyW60MsfA+919+5K11PCeNprVJyF7+K3Alvc/auVrmc07v45d18U/W5eR9iOZCoGO9FrZqeZnRVteieweZQvqZTXgEvNrC76XXgnYxz4TdzIfQy3AbdFLYz7gY9MwZHmdHMzUA08HP2V8ZS7/25lSzpqpPYaFS6rlLcBNwAbzOy5aNvn3f2+CtYUF58G7oje3LcBH61wPceIpozuBtYRTm8+yxhtCNR+QEQkhjQtIyISQwp3EZEYUriLiMSQwl1EJIYU7iIiMaRwFxGJIYW7nHRR/44JdzQ0sy+Z2R+W2J43s+eKPpZF2y82s8eidr4vRM9bN8Jj32hmbdHXbzaz3x52/0/M7Mlx1PgdM3v/sG1dE/k+RcpBJzHJSefuv1Xmh+xx9/OLN5jZPOAu4Dp3fzI6q+8aoAEY6UzZO939U2Y2F9hkZve4+96oB/0KoMvMlrv79jLXL1J2GrnLpDKzejP7NzNbH11k4Foz+4WZrYzu/5iZvRRt+0czu7lMT/1J4Lvu/iSAh+52971jfWHU6vkVYGm06RrgXsJ2wMd98XczWxD9JfFc9LN4e7T9m2bWEl2I4c+K9r8q+ovjCTP7u8JFL6Kf6W1m9kzUyXDKtSiWylO4y2RbBexy9/Oiiww8ULjDzE4l7L53KeFFR84+zueoLZqS+VG07Rxg7fE8mJmdRtgKdmu06Xrg+9HH9cdZI8CHgAejvzLOAwptBL7g7iuBc4HLzezc6Co7/we40t0vA5qLHucLhP1aLgJ+GfjrqdimVipL0zIy2TYAXzGzvwJ+6u6PRz1mILxQxqPu3g5gZncR9tSfqGOmZY7TtWZ2GdAH/I67t0fTO6cDT7i7m1nOzM5x940jPEapfh6Fbc8Q9i7KAj9290K4f9DMVhO+HhcQXhEqBWwrmgL6PrA6+vxXCLsuFo4/1ABLmIJXEJLKUbjLpHL3l8zsQuAq4C/MrPgCA6V6qZfLJuBC4CcT+Jo73f1Tw7ZdS9jjf3v0pjSTcGrmT0Z4jAPR/gCY2WyizqLu/piZ/SfCKxR9z8z+Gngc+EPgInfvMLPvEIb1aD8bA65x9xcn8L1JwmhaRiZVNPXSHV1U4CuEByYLniachmiy8NJh15TxqW8GPmJmlxTV8mEzK3VRkdFcD6wqaq98IaPPu/+C8C+Aquj2jcAj0fMvJbyIxT8Stu9dQfhmcQQ4FP2VcGX0dS8ApxVW/hC+yRQ8CHw6OkiMmV0wwe9JEkAjd5lsbyGcEw6AAcLr1H4FwkuHmdn/Irxc3C7CPtqHxni8PzGzzxRuuHvJa7JGq1yuI5wSmkt4ubfHgHFf5CIK1iWE134tPO52M+s0s0uKLvBQ/Lw/jf5SWWtmecIDs4UWx1cAf2RmA0AX8BvR4z1L+JfGNuDfo8fpMbNPAA+Y2X7CN8KC/0F4Gbvno4DfQXgxFJFBavkrFWVmM9y9Kxq5/4iwn/qPxvq6JCj62RjwDeBld/9apeuS6UHTMlJpX4ouPrER2E54NSwJ/Xb0s9kEzCJcPSMyLhq5y5Rj4cW0PzBs813u/j/L8NgfBW4atvnf3f2TE3ycSatRpBwU7iIiMaRpGRGRGFK4i4jEkMJdRCSGFO4iIjH0/wF+PrpBdBkYtAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "sns.distplot(lab2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "id": "65c070b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.preprocessing import OneHotEncoder\n",
    "\n",
    "encoder = OneHotEncoder(categories='auto')\n",
    "inp = [[float(i)] for i in range(14)]\n",
    "output = encoder.fit_transform(inp)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "ef0b1c1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[[0.0],\n",
       " [1.0],\n",
       " [2.0],\n",
       " [3.0],\n",
       " [4.0],\n",
       " [5.0],\n",
       " [6.0],\n",
       " [7.0],\n",
       " [8.0],\n",
       " [9.0],\n",
       " [10.0],\n",
       " [11.0],\n",
       " [12.0],\n",
       " [13.0]]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "id": "6e212634",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]),\n",
       " array([0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# np.tile(output.toarray()[1],(20,1)).tolist()\n",
    "[output.toarray()[1]]*2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "76ea098a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>site_name</th>\n",
       "      <th>seqs</th>\n",
       "      <th>label</th>\n",
       "      <th>ct</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>chr12:AACS:125140928:125143316:+</td>\n",
       "      <td>GUGAGGCGGGACAAACUUGUCUUCCUCACACCCAUCUUACUUCCUC...</td>\n",
       "      <td>0.0371</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201672:67202710:-</td>\n",
       "      <td>ACUACAUCAUAAACAUGUCUUUGAAACCCGUCUCCCAUCUUCUAGU...</td>\n",
       "      <td>1.2828</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>chr15:AAGAB:67202710:67217079:-</td>\n",
       "      <td>AUCUAUUAUAGUUCAAUCCCCAGUAAUGCAGAUGGAGGAGGAUUAG...</td>\n",
       "      <td>-0.2141</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>chr2:AAK1:69457997:69461526:-</td>\n",
       "      <td>UUGUAUUCUUAGUAGAGAUGGGGUUUCACCAUGUUGGCCAGGUUGG...</td>\n",
       "      <td>0.0173</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>chr2:AAK1:69457997:69467864:-</td>\n",
       "      <td>AAAAACCCACUUUUUAACAAAUAAUUUUGAUGGGUUAGACAUGUUC...</td>\n",
       "      <td>-0.9525</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15779</td>\n",
       "      <td>15779</td>\n",
       "      <td>chr7:ZYX:143382988:143391082:+</td>\n",
       "      <td>AGUCGGGAAUGUGGGGCACGAAUCUUCCCCUCCGGGUCCGUUUUCC...</td>\n",
       "      <td>0.7120</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15780</td>\n",
       "      <td>15780</td>\n",
       "      <td>chr17:ZZEF1:4004475:4006020:-</td>\n",
       "      <td>AAUGAGGGAGUGGCUCCCGACAUUGGCGACAACUCUUUUUCCCUCU...</td>\n",
       "      <td>-0.5428</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15781</td>\n",
       "      <td>15781</td>\n",
       "      <td>chr17:ZZEF1:4004475:4008624:-</td>\n",
       "      <td>GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA...</td>\n",
       "      <td>-0.4334</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15782</td>\n",
       "      <td>15782</td>\n",
       "      <td>chr17:ZZEF1:4006020:4008624:-</td>\n",
       "      <td>GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA...</td>\n",
       "      <td>0.1435</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>15783</td>\n",
       "      <td>15783</td>\n",
       "      <td>chr17:ZZEF1:4008624:4021776:-</td>\n",
       "      <td>AGUUGCUGGUGAAAAAGUGCAGCAAAGGGACUGGCUUUAGUAAAAC...</td>\n",
       "      <td>0.3742</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15784 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Unnamed: 0                         site_name  \\\n",
       "0               0  chr12:AACS:125140928:125143316:+   \n",
       "1               1   chr15:AAGAB:67201672:67202710:-   \n",
       "2               2   chr15:AAGAB:67202710:67217079:-   \n",
       "3               3     chr2:AAK1:69457997:69461526:-   \n",
       "4               4     chr2:AAK1:69457997:69467864:-   \n",
       "...           ...                               ...   \n",
       "15779       15779    chr7:ZYX:143382988:143391082:+   \n",
       "15780       15780     chr17:ZZEF1:4004475:4006020:-   \n",
       "15781       15781     chr17:ZZEF1:4004475:4008624:-   \n",
       "15782       15782     chr17:ZZEF1:4006020:4008624:-   \n",
       "15783       15783     chr17:ZZEF1:4008624:4021776:-   \n",
       "\n",
       "                                                    seqs   label  \\\n",
       "0      GUGAGGCGGGACAAACUUGUCUUCCUCACACCCAUCUUACUUCCUC...  0.0371   \n",
       "1      ACUACAUCAUAAACAUGUCUUUGAAACCCGUCUCCCAUCUUCUAGU...  1.2828   \n",
       "2      AUCUAUUAUAGUUCAAUCCCCAGUAAUGCAGAUGGAGGAGGAUUAG... -0.2141   \n",
       "3      UUGUAUUCUUAGUAGAGAUGGGGUUUCACCAUGUUGGCCAGGUUGG...  0.0173   \n",
       "4      AAAAACCCACUUUUUAACAAAUAAUUUUGAUGGGUUAGACAUGUUC... -0.9525   \n",
       "...                                                  ...     ...   \n",
       "15779  AGUCGGGAAUGUGGGGCACGAAUCUUCCCCUCCGGGUCCGUUUUCC...  0.7120   \n",
       "15780  AAUGAGGGAGUGGCUCCCGACAUUGGCGACAACUCUUUUUCCCUCU... -0.5428   \n",
       "15781  GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA... -0.4334   \n",
       "15782  GUCACCGAGAACCGUGCCCAGGUGAGGCCUGAGGUCUGGGCAGGGA...  0.1435   \n",
       "15783  AGUUGCUGGUGAAAAAGUGCAGCAAAGGGACUGGCUUUAGUAAAAC...  0.3742   \n",
       "\n",
       "                                               ct  \n",
       "0      [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "1      [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "2      [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "3      [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "4      [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "...                                           ...  \n",
       "15779  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "15780  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "15781  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "15782  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "15783  [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]  \n",
       "\n",
       "[15784 rows x 5 columns]"
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "idx = celltypes_order['Excitatory']\n",
    "inp_df['ct'] = [np.array(output.toarray()[idx], dtype=np.int8)]* inp_df.shape[0]\n",
    "inp_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "c3f1276d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'AST-FB': 0,\n",
       " 'AST-PP': 1,\n",
       " 'Excitatory': 2,\n",
       " 'IN-PV': 3,\n",
       " 'IN-SST': 4,\n",
       " 'IN-VIP': 5,\n",
       " 'Inhibitory': 6,\n",
       " 'L2-3': 7,\n",
       " 'L4': 8,\n",
       " 'L5-6': 9,\n",
       " 'L5-6-CC': 10,\n",
       " 'Microglia': 11,\n",
       " 'OPC': 12,\n",
       " 'Oligodendrocytes': 13}"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "celltypes_order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "id": "4a113c53",
   "metadata": {},
   "outputs": [],
   "source": [
    "## first get rid of Exci from celltypes list as you already have it in inp_df\n",
    "acc =inp_df.shape[0]\n",
    "for celltype in celltypes:\n",
    "    if celltype != 'Excitatory':\n",
    "        df_name = data_root + celltype + \"_sALSvsCTRL/input_data.tsv\"\n",
    "        df = pd.read_csv(df_name, sep='\\t')\n",
    "        acc += df.shape[0]\n",
    "        idx = celltypes_order[celltype]\n",
    "        df['ct'] = [np.array(output.toarray()[idx], dtype=np.float32)] * df.shape[0]\n",
    "        inp_df = pd.concat([inp_df,df], ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "10373101",
   "metadata": {},
   "outputs": [],
   "source": [
    "outname = data_root + \"sals_all_seqs_celltypes_input.pkl\"\n",
    "inp_df.to_pickle(outname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "35a078ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(275741, 5)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "inp_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "f69d7022",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pickle, os\n",
    "import pandas as pd\n",
    "import random\n",
    "import numpy as np\n",
    "\n",
    "def clean_nans(arr):\n",
    "    \"\"\"\n",
    "    takes in the label np.array and then zeros the nan values\n",
    "    \n",
    "    \"\"\"\n",
    "    arr[np.isnan(arr)] = 0\n",
    "    out = arr\n",
    "    return out\n",
    "\n",
    "\n",
    "def split_data(df):\n",
    "    \"\"\"\n",
    "    takes in the pandas dataframe and split to train, test and valid datasets\n",
    "    \"\"\"\n",
    "    df = df[df['label'].notna()]\n",
    "    test_index = np.array(random.sample(range(df.shape[0]), int(float(df.shape[0])*0.15)))\n",
    "    mask = np.zeros(df.shape[0],dtype=bool)\n",
    "    mask[test_index] = True\n",
    "    df_test = df[mask]\n",
    "    df_test.head()\n",
    "    tmp_df = df[~mask]\n",
    "    valid_index = np.array(random.sample(range(tmp_df.shape[0]), int(float(tmp_df.shape[0])*0.05)))\n",
    "    mask = np.zeros(tmp_df.shape[0],dtype=bool)\n",
    "    mask[valid_index] = True\n",
    "    df_valid = tmp_df[mask]\n",
    "    df_train = tmp_df[~mask]\n",
    "    \n",
    "    return (df_train, df_valid, df_test)\n",
    "\n",
    "def change_arr_dtype(arr):\n",
    "    return arr.astype(np.float32)\n",
    "\n",
    "def get_np_input_for_model(df):\n",
    "    # add one hot reps to the df_train\n",
    "    df['seqs_one_hot'] = df['seqs'].apply(makes_seqs_ready, args=(16000,10))\n",
    "    df_train_all_num = pd.DataFrame([df['index'],\n",
    "                                 df['seqs_one_hot'],df['label'],df['celltype']]).T\n",
    "    df_train_all_num.columns = ['index', 'seqs_one_hot','label','celltype']\n",
    "    # make the np array of data\n",
    "    idx = np.array(df_train_all_num['index'].values, dtype=np.float32)\n",
    "    seqs = np.array(df_train_all_num['seqs_one_hot'].values)\n",
    "    labels = np.array(df_train_all_num['label'].values, dtype=np.float32)\n",
    "    celltype = df_train_all_num['celltype'].values\n",
    "    df_train_all_num_numpy = pd.DataFrame([idx,seqs,labels,celltype]).T.to_numpy()\n",
    "    \n",
    "    return df_train_all_num_numpy\n",
    "\n",
    "\n",
    "def makes_seqs_ready(seq,max_len,left_pad_max):\n",
    "    \"\"\"\n",
    "    takes in the sequence, max len and max left pad\n",
    "    and returns the one hot representation of sequence with maximum pad\n",
    "    \"\"\"\n",
    "    if len(seq) < max_len:\n",
    "        diff = max_len - len(seq)\n",
    "        lp = random.randint(0, left_pad_max)\n",
    "        lp = min(lp, diff)\n",
    "        seq = get_1h_seq(\n",
    "            \"N\" * (lp) + str(seq) + \"N\" * (max_len - len(seq) - lp)\n",
    "        )\n",
    "    else:\n",
    "        seq = get_1h_seq(str(seq))\n",
    "    seq = np.array(seq, dtype=np.float32).swapaxes(0, 1)\n",
    "    \n",
    "    return seq\n",
    "\n",
    "def get_1h_seq(seq):\n",
    "    \"\"\"\n",
    "    This is a simple code to get one-hot representaton of the\n",
    "    senquences.\n",
    "    --------------\n",
    "    Arguments:\n",
    "    seq: RNA sequence in fasta format\n",
    "    \"\"\"\n",
    "    nt_code = {\n",
    "        \"N\": [0, 0, 0, 0],\n",
    "        \"A\": [1, 0, 0, 0],\n",
    "        \"C\": [0, 1, 0, 0],\n",
    "        \"G\": [0, 0, 1, 0],\n",
    "        \"U\": [0, 0, 0, 1],\n",
    "    }\n",
    "\n",
    "    seq_1h = [nt_code[nt] for nt in seq]\n",
    "    return np.array(seq_1h)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 382,
   "id": "b00acb5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def makit_nparr(lst):\n",
    "    lst = list(lst)\n",
    "    print(lst)\n",
    "    arr = np.array(lst, dtype=np.float32)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 383,
   "id": "a3465fd3",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['[', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '1', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ' ', '0', '.', ']']\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "could not convert string to float: '['",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-383-6cc2b07fc968>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0minput_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'celltype'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0minput_data\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'celltype'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mapply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmakit_nparr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/pandas/core/series.py\u001b[0m in \u001b[0;36mapply\u001b[0;34m(self, func, convert_dtype, args, **kwds)\u001b[0m\n\u001b[1;32m   4040\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4041\u001b[0m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mastype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobject\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4042\u001b[0;31m                 \u001b[0mmapped\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmap_infer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mconvert\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mconvert_dtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4043\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4044\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0misinstance\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmapped\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mSeries\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/lib.pyx\u001b[0m in \u001b[0;36mpandas._libs.lib.map_infer\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32m<ipython-input-382-aa24986afb9b>\u001b[0m in \u001b[0;36mmakit_nparr\u001b[0;34m(lst)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0mlst\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlst\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlst\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0marr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marray\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlst\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfloat32\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m: could not convert string to float: '['"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 486,
   "id": "56e2918b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ## read in the all_seqs_all_ct dataframe and prepare numpy input\n",
    "# dtype_in = {}\n",
    "# input_data = pd.read_csv(outname, sep='\\t')\n",
    "# input_data = input_data.iloc[:,1:]\n",
    "# input_data.columns = ['index','site_name', 'seqs','label','celltype']\n",
    "# input_data['index'] = input_data.index.values\n",
    "# input_data\n",
    "##### I had problem with reading the pd.dataframe of the input! it reads in the celltype arrays as string\n",
    "##### to avoid this you should pass in the dtype dictionary to your pd.read_csv, for now lets skip it"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "5b38f5d0",
   "metadata": {},
   "outputs": [],
   "source": [
    "inp_df.columns = ['index','site_name', 'seqs','label','celltype']\n",
    "df_train, df_valid, df_test = split_data(inp_df)\n",
    "\n",
    "df_valid.columns = ['index','site_name', 'seqs','label','celltype']\n",
    "df_valid = get_np_input_for_model(df_valid)\n",
    "\n",
    "\n",
    "df_train.columns = ['index','site_name', 'seqs','label','celltype']\n",
    "df_train = get_np_input_for_model(df_train)\n",
    "\n",
    "df_test.columns = ['index','site_name', 'seqs','label','celltype']\n",
    "df_test = get_np_input_for_model(df_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "d6ceb9c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/valid_data', df_valid)\n",
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/train_data', df_train)\n",
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/test_data', df_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "40def979",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 509,
   "id": "e198ccc4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(11715, 4)\n",
      "(41350, 4)\n"
     ]
    }
   ],
   "source": [
    "print(df_valid.shape)\n",
    "print(df_test.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a38a8a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/valid_data', df_valid)\n",
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/train_data', df_train)\n",
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/test_data', df_test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "23727fac",
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/tst_train_data', df_valid[1:33])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "8530c11e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(242568, 4)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_train.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a655fb89",
   "metadata": {},
   "source": [
    "## Lets get positive control\n",
    "## basically do motif analysis on seqs that are lengthened (positive LFC) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c0902aa6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>p_devtest</th>\n",
       "      <th>fdr_p_devtest</th>\n",
       "      <th>ref_site</th>\n",
       "      <th>alt_site</th>\n",
       "      <th>b_intercept</th>\n",
       "      <th>p_intercept</th>\n",
       "      <th>b_ConditionALS_pathology</th>\n",
       "      <th>p_ConditionALS_pathology</th>\n",
       "      <th>strand</th>\n",
       "      <th>multiplyer</th>\n",
       "      <th>LFC_PA_Usage</th>\n",
       "      <th>negative_logFDR</th>\n",
       "      <th>switch_width</th>\n",
       "      <th>bed</th>\n",
       "      <th>correction_multiplyer</th>\n",
       "      <th>switch_name</th>\n",
       "      <th>sig_LFC_PA_Usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>0.261215</td>\n",
       "      <td>4.734204e-02</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>3.376034e-07</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>6.471593</td>\n",
       "      <td>16644</td>\n",
       "      <td>chr12,125124284,125140928</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125140928:+</td>\n",
       "      <td>0.983355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125143316:+</td>\n",
       "      <td>1.641746</td>\n",
       "      <td>4.866807e-52</td>\n",
       "      <td>0.263182</td>\n",
       "      <td>1.246049e-01</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.263182</td>\n",
       "      <td>0.904465</td>\n",
       "      <td>19032</td>\n",
       "      <td>chr12,125124284,125143316</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125143316:+</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>chr12:125143316:+</td>\n",
       "      <td>1.380531</td>\n",
       "      <td>8.547748e-46</td>\n",
       "      <td>-0.720173</td>\n",
       "      <td>1.476861e-08</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>-0.720173</td>\n",
       "      <td>7.830660</td>\n",
       "      <td>2388</td>\n",
       "      <td>chr12,125140928,125143316</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125140928:125143316:+</td>\n",
       "      <td>-0.720173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67201672:-</td>\n",
       "      <td>0.936799</td>\n",
       "      <td>7.468705e-06</td>\n",
       "      <td>-0.043726</td>\n",
       "      <td>8.609305e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.043726</td>\n",
       "      <td>0.065032</td>\n",
       "      <td>644</td>\n",
       "      <td>chr15,67201028,67201672</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67201672:-</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67202710:-</td>\n",
       "      <td>1.470439</td>\n",
       "      <td>7.400884e-14</td>\n",
       "      <td>-1.337591</td>\n",
       "      <td>1.070773e-07</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.337591</td>\n",
       "      <td>6.970303</td>\n",
       "      <td>1682</td>\n",
       "      <td>chr15,67201028,67202710</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67202710:-</td>\n",
       "      <td>1.337591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41174</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>-0.321971</td>\n",
       "      <td>2.534728e-02</td>\n",
       "      <td>0.526766</td>\n",
       "      <td>5.576124e-03</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.526766</td>\n",
       "      <td>2.253668</td>\n",
       "      <td>53729</td>\n",
       "      <td>chr1,77564505,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77618234:-</td>\n",
       "      <td>-0.526766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41175</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.016713</td>\n",
       "      <td>8.988251e-01</td>\n",
       "      <td>-0.225352</td>\n",
       "      <td>2.362558e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.225352</td>\n",
       "      <td>0.626618</td>\n",
       "      <td>65335</td>\n",
       "      <td>chr1,77564505,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77629840:-</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41176</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>-0.221592</td>\n",
       "      <td>1.320936e-01</td>\n",
       "      <td>0.933971</td>\n",
       "      <td>6.105239e-06</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.933971</td>\n",
       "      <td>5.214297</td>\n",
       "      <td>37236</td>\n",
       "      <td>chr1,77580998,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77580998:77618234:-</td>\n",
       "      <td>-0.933971</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41177</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.117093</td>\n",
       "      <td>3.854144e-01</td>\n",
       "      <td>0.181853</td>\n",
       "      <td>3.789934e-01</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>-0.181853</td>\n",
       "      <td>0.421368</td>\n",
       "      <td>48842</td>\n",
       "      <td>chr1,77580998,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77580998:77629840:-</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41178</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.338685</td>\n",
       "      <td>1.825640e-02</td>\n",
       "      <td>-0.752118</td>\n",
       "      <td>1.128067e-04</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.752118</td>\n",
       "      <td>3.947665</td>\n",
       "      <td>11606</td>\n",
       "      <td>chr1,77618234,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77618234:77629840:-</td>\n",
       "      <td>0.752118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>41179 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript  p_devtest  fdr_p_devtest           ref_site  \\\n",
       "0           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "1           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "2           AACS   0.006145       0.019492  chr12:125140928:+   \n",
       "3          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "4          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "...          ...        ...            ...                ...   \n",
       "41174       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41175       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41176       ZZZ3   1.000000       1.000000    chr1:77580998:-   \n",
       "41177       ZZZ3   1.000000       1.000000    chr1:77580998:-   \n",
       "41178       ZZZ3   1.000000       1.000000    chr1:77618234:-   \n",
       "\n",
       "                alt_site  b_intercept   p_intercept  b_ConditionALS_pathology  \\\n",
       "0      chr12:125140928:+     0.261215  4.734204e-02                  0.983355   \n",
       "1      chr12:125143316:+     1.641746  4.866807e-52                  0.263182   \n",
       "2      chr12:125143316:+     1.380531  8.547748e-46                 -0.720173   \n",
       "3       chr15:67201672:-     0.936799  7.468705e-06                 -0.043726   \n",
       "4       chr15:67202710:-     1.470439  7.400884e-14                 -1.337591   \n",
       "...                  ...          ...           ...                       ...   \n",
       "41174    chr1:77618234:-    -0.321971  2.534728e-02                  0.526766   \n",
       "41175    chr1:77629840:-     0.016713  8.988251e-01                 -0.225352   \n",
       "41176    chr1:77618234:-    -0.221592  1.320936e-01                  0.933971   \n",
       "41177    chr1:77629840:-     0.117093  3.854144e-01                  0.181853   \n",
       "41178    chr1:77629840:-     0.338685  1.825640e-02                 -0.752118   \n",
       "\n",
       "       p_ConditionALS_pathology strand  multiplyer  LFC_PA_Usage  \\\n",
       "0                  3.376034e-07      +           1      0.983355   \n",
       "1                  1.246049e-01      +           1      0.263182   \n",
       "2                  1.476861e-08      +           1     -0.720173   \n",
       "3                  8.609305e-01      -          -1      0.043726   \n",
       "4                  1.070773e-07      -          -1      1.337591   \n",
       "...                         ...    ...         ...           ...   \n",
       "41174              5.576124e-03      -          -1     -0.526766   \n",
       "41175              2.362558e-01      -          -1      0.225352   \n",
       "41176              6.105239e-06      -          -1     -0.933971   \n",
       "41177              3.789934e-01      -          -1     -0.181853   \n",
       "41178              1.128067e-04      -          -1      0.752118   \n",
       "\n",
       "       negative_logFDR  switch_width                        bed  \\\n",
       "0             6.471593         16644  chr12,125124284,125140928   \n",
       "1             0.904465         19032  chr12,125124284,125143316   \n",
       "2             7.830660          2388  chr12,125140928,125143316   \n",
       "3             0.065032           644    chr15,67201028,67201672   \n",
       "4             6.970303          1682    chr15,67201028,67202710   \n",
       "...                ...           ...                        ...   \n",
       "41174         2.253668         53729     chr1,77564505,77618234   \n",
       "41175         0.626618         65335     chr1,77564505,77629840   \n",
       "41176         5.214297         37236     chr1,77580998,77618234   \n",
       "41177         0.421368         48842     chr1,77580998,77629840   \n",
       "41178         3.947665         11606     chr1,77618234,77629840   \n",
       "\n",
       "       correction_multiplyer                       switch_name  \\\n",
       "0                          1  chr12:AACS:125124284:125140928:+   \n",
       "1                          1  chr12:AACS:125124284:125143316:+   \n",
       "2                          1  chr12:AACS:125140928:125143316:+   \n",
       "3                          1   chr15:AAGAB:67201028:67201672:-   \n",
       "4                          1   chr15:AAGAB:67201028:67202710:-   \n",
       "...                      ...                               ...   \n",
       "41174                      1     chr1:ZZZ3:77564505:77618234:-   \n",
       "41175                      1     chr1:ZZZ3:77564505:77629840:-   \n",
       "41176                      1     chr1:ZZZ3:77580998:77618234:-   \n",
       "41177                      1     chr1:ZZZ3:77580998:77629840:-   \n",
       "41178                      1     chr1:ZZZ3:77618234:77629840:-   \n",
       "\n",
       "       sig_LFC_PA_Usage  \n",
       "0              0.983355  \n",
       "1              0.000000  \n",
       "2             -0.720173  \n",
       "3              0.000000  \n",
       "4              1.337591  \n",
       "...                 ...  \n",
       "41174         -0.526766  \n",
       "41175          0.000000  \n",
       "41176         -0.933971  \n",
       "41177          0.000000  \n",
       "41178          0.752118  \n",
       "\n",
       "[41179 rows x 18 columns]"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exc_res = pd.read_csv('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/APAlog_res_metadata_added.tsv', sep='\\t')\n",
    "exc_res"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "2a0a1972",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript</th>\n",
       "      <th>p_devtest</th>\n",
       "      <th>fdr_p_devtest</th>\n",
       "      <th>ref_site</th>\n",
       "      <th>alt_site</th>\n",
       "      <th>b_intercept</th>\n",
       "      <th>p_intercept</th>\n",
       "      <th>b_ConditionALS_pathology</th>\n",
       "      <th>p_ConditionALS_pathology</th>\n",
       "      <th>strand</th>\n",
       "      <th>multiplyer</th>\n",
       "      <th>LFC_PA_Usage</th>\n",
       "      <th>negative_logFDR</th>\n",
       "      <th>switch_width</th>\n",
       "      <th>bed</th>\n",
       "      <th>correction_multiplyer</th>\n",
       "      <th>switch_name</th>\n",
       "      <th>sig_LFC_PA_Usage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>AACS</td>\n",
       "      <td>0.006145</td>\n",
       "      <td>0.019492</td>\n",
       "      <td>chr12:125124284:+</td>\n",
       "      <td>chr12:125140928:+</td>\n",
       "      <td>0.261215</td>\n",
       "      <td>4.734204e-02</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>3.376034e-07</td>\n",
       "      <td>+</td>\n",
       "      <td>1</td>\n",
       "      <td>0.983355</td>\n",
       "      <td>6.471593</td>\n",
       "      <td>16644</td>\n",
       "      <td>chr12,125124284,125140928</td>\n",
       "      <td>1</td>\n",
       "      <td>chr12:AACS:125124284:125140928:+</td>\n",
       "      <td>0.983355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67202710:-</td>\n",
       "      <td>1.470439</td>\n",
       "      <td>7.400884e-14</td>\n",
       "      <td>-1.337591</td>\n",
       "      <td>1.070773e-07</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.337591</td>\n",
       "      <td>6.970303</td>\n",
       "      <td>1682</td>\n",
       "      <td>chr15,67201028,67202710</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67202710:-</td>\n",
       "      <td>1.337591</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201028:-</td>\n",
       "      <td>chr15:67217079:-</td>\n",
       "      <td>1.098249</td>\n",
       "      <td>8.045112e-08</td>\n",
       "      <td>-1.226330</td>\n",
       "      <td>3.590218e-06</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.226330</td>\n",
       "      <td>5.444879</td>\n",
       "      <td>16051</td>\n",
       "      <td>chr15,67201028,67217079</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201028:67217079:-</td>\n",
       "      <td>1.226330</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>6</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201672:-</td>\n",
       "      <td>chr15:67202710:-</td>\n",
       "      <td>0.533640</td>\n",
       "      <td>1.343992e-04</td>\n",
       "      <td>-1.293865</td>\n",
       "      <td>1.237304e-11</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.293865</td>\n",
       "      <td>10.907524</td>\n",
       "      <td>1038</td>\n",
       "      <td>chr15,67201672,67202710</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201672:67202710:-</td>\n",
       "      <td>1.293865</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>7</td>\n",
       "      <td>AAGAB</td>\n",
       "      <td>0.000253</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>chr15:67201672:-</td>\n",
       "      <td>chr15:67217079:-</td>\n",
       "      <td>0.161450</td>\n",
       "      <td>2.848487e-01</td>\n",
       "      <td>-1.182604</td>\n",
       "      <td>1.250607e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.182604</td>\n",
       "      <td>7.902879</td>\n",
       "      <td>15407</td>\n",
       "      <td>chr15,67201672,67217079</td>\n",
       "      <td>1</td>\n",
       "      <td>chr15:AAGAB:67201672:67217079:-</td>\n",
       "      <td>1.182604</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41170</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77562475:-</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>1.235521</td>\n",
       "      <td>2.309093e-09</td>\n",
       "      <td>-1.937261</td>\n",
       "      <td>1.751085e-14</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.937261</td>\n",
       "      <td>13.756693</td>\n",
       "      <td>18523</td>\n",
       "      <td>chr1,77562475,77580998</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77562475:77580998:-</td>\n",
       "      <td>1.937261</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41171</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77562475:-</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>1.013929</td>\n",
       "      <td>1.827069e-06</td>\n",
       "      <td>-1.003291</td>\n",
       "      <td>3.643041e-05</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.003291</td>\n",
       "      <td>4.438536</td>\n",
       "      <td>55759</td>\n",
       "      <td>chr1,77562475,77618234</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77562475:77618234:-</td>\n",
       "      <td>1.003291</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41172</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77562475:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>1.352614</td>\n",
       "      <td>3.506398e-11</td>\n",
       "      <td>-1.755409</td>\n",
       "      <td>5.237627e-13</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>1.755409</td>\n",
       "      <td>12.280865</td>\n",
       "      <td>67365</td>\n",
       "      <td>chr1,77562475,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77562475:77629840:-</td>\n",
       "      <td>1.755409</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41173</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77564505:-</td>\n",
       "      <td>chr1:77580998:-</td>\n",
       "      <td>-0.100379</td>\n",
       "      <td>4.586096e-01</td>\n",
       "      <td>-0.407205</td>\n",
       "      <td>4.407777e-02</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.407205</td>\n",
       "      <td>1.355780</td>\n",
       "      <td>16493</td>\n",
       "      <td>chr1,77564505,77580998</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77564505:77580998:-</td>\n",
       "      <td>0.407205</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>41178</td>\n",
       "      <td>ZZZ3</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>chr1:77618234:-</td>\n",
       "      <td>chr1:77629840:-</td>\n",
       "      <td>0.338685</td>\n",
       "      <td>1.825640e-02</td>\n",
       "      <td>-0.752118</td>\n",
       "      <td>1.128067e-04</td>\n",
       "      <td>-</td>\n",
       "      <td>-1</td>\n",
       "      <td>0.752118</td>\n",
       "      <td>3.947665</td>\n",
       "      <td>11606</td>\n",
       "      <td>chr1,77618234,77629840</td>\n",
       "      <td>1</td>\n",
       "      <td>chr1:ZZZ3:77618234:77629840:-</td>\n",
       "      <td>0.752118</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>15743 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      transcript  p_devtest  fdr_p_devtest           ref_site  \\\n",
       "0           AACS   0.006145       0.019492  chr12:125124284:+   \n",
       "4          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "5          AAGAB   0.000253       0.001115   chr15:67201028:-   \n",
       "6          AAGAB   0.000253       0.001115   chr15:67201672:-   \n",
       "7          AAGAB   0.000253       0.001115   chr15:67201672:-   \n",
       "...          ...        ...            ...                ...   \n",
       "41170       ZZZ3   1.000000       1.000000    chr1:77562475:-   \n",
       "41171       ZZZ3   1.000000       1.000000    chr1:77562475:-   \n",
       "41172       ZZZ3   1.000000       1.000000    chr1:77562475:-   \n",
       "41173       ZZZ3   1.000000       1.000000    chr1:77564505:-   \n",
       "41178       ZZZ3   1.000000       1.000000    chr1:77618234:-   \n",
       "\n",
       "                alt_site  b_intercept   p_intercept  b_ConditionALS_pathology  \\\n",
       "0      chr12:125140928:+     0.261215  4.734204e-02                  0.983355   \n",
       "4       chr15:67202710:-     1.470439  7.400884e-14                 -1.337591   \n",
       "5       chr15:67217079:-     1.098249  8.045112e-08                 -1.226330   \n",
       "6       chr15:67202710:-     0.533640  1.343992e-04                 -1.293865   \n",
       "7       chr15:67217079:-     0.161450  2.848487e-01                 -1.182604   \n",
       "...                  ...          ...           ...                       ...   \n",
       "41170    chr1:77580998:-     1.235521  2.309093e-09                 -1.937261   \n",
       "41171    chr1:77618234:-     1.013929  1.827069e-06                 -1.003291   \n",
       "41172    chr1:77629840:-     1.352614  3.506398e-11                 -1.755409   \n",
       "41173    chr1:77580998:-    -0.100379  4.586096e-01                 -0.407205   \n",
       "41178    chr1:77629840:-     0.338685  1.825640e-02                 -0.752118   \n",
       "\n",
       "       p_ConditionALS_pathology strand  multiplyer  LFC_PA_Usage  \\\n",
       "0                  3.376034e-07      +           1      0.983355   \n",
       "4                  1.070773e-07      -          -1      1.337591   \n",
       "5                  3.590218e-06      -          -1      1.226330   \n",
       "6                  1.237304e-11      -          -1      1.293865   \n",
       "7                  1.250607e-08      -          -1      1.182604   \n",
       "...                         ...    ...         ...           ...   \n",
       "41170              1.751085e-14      -          -1      1.937261   \n",
       "41171              3.643041e-05      -          -1      1.003291   \n",
       "41172              5.237627e-13      -          -1      1.755409   \n",
       "41173              4.407777e-02      -          -1      0.407205   \n",
       "41178              1.128067e-04      -          -1      0.752118   \n",
       "\n",
       "       negative_logFDR  switch_width                        bed  \\\n",
       "0             6.471593         16644  chr12,125124284,125140928   \n",
       "4             6.970303          1682    chr15,67201028,67202710   \n",
       "5             5.444879         16051    chr15,67201028,67217079   \n",
       "6            10.907524          1038    chr15,67201672,67202710   \n",
       "7             7.902879         15407    chr15,67201672,67217079   \n",
       "...                ...           ...                        ...   \n",
       "41170        13.756693         18523     chr1,77562475,77580998   \n",
       "41171         4.438536         55759     chr1,77562475,77618234   \n",
       "41172        12.280865         67365     chr1,77562475,77629840   \n",
       "41173         1.355780         16493     chr1,77564505,77580998   \n",
       "41178         3.947665         11606     chr1,77618234,77629840   \n",
       "\n",
       "       correction_multiplyer                       switch_name  \\\n",
       "0                          1  chr12:AACS:125124284:125140928:+   \n",
       "4                          1   chr15:AAGAB:67201028:67202710:-   \n",
       "5                          1   chr15:AAGAB:67201028:67217079:-   \n",
       "6                          1   chr15:AAGAB:67201672:67202710:-   \n",
       "7                          1   chr15:AAGAB:67201672:67217079:-   \n",
       "...                      ...                               ...   \n",
       "41170                      1     chr1:ZZZ3:77562475:77580998:-   \n",
       "41171                      1     chr1:ZZZ3:77562475:77618234:-   \n",
       "41172                      1     chr1:ZZZ3:77562475:77629840:-   \n",
       "41173                      1     chr1:ZZZ3:77564505:77580998:-   \n",
       "41178                      1     chr1:ZZZ3:77618234:77629840:-   \n",
       "\n",
       "       sig_LFC_PA_Usage  \n",
       "0              0.983355  \n",
       "4              1.337591  \n",
       "5              1.226330  \n",
       "6              1.293865  \n",
       "7              1.182604  \n",
       "...                 ...  \n",
       "41170          1.937261  \n",
       "41171          1.003291  \n",
       "41172          1.755409  \n",
       "41173          0.407205  \n",
       "41178          0.752118  \n",
       "\n",
       "[15743 rows x 18 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exc_res_lengthened = exc_res[exc_res['sig_LFC_PA_Usage'] > 0]\n",
    "exc_res_lengthened"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "194d98e4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(15743, 18)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exc_res_lengthened.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "0a19fbe2",
   "metadata": {},
   "outputs": [],
   "source": [
    "out_name = \"/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/Exc_lengthened_PAs.fa\"\n",
    "output_handle = open(out_name, 'w')\n",
    "inp_fa = \"/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/switch_DNA_sequence.fa\"\n",
    "inp_fa = SeqIO.parse(inp_fa, \"fasta\")\n",
    "Exc_lengthened_RNAs = {}\n",
    "for rec in inp_fa:\n",
    "    if rec.id in list(exc_res_lengthened['switch_name']):\n",
    "        strand = rec.id .split(':')[-1]\n",
    "        if strand == '+':\n",
    "            seq = transcribe_positive_strand(str(rec.seq))\n",
    "        else:\n",
    "            seq = transcribe_negative_strand(str(rec.seq))\n",
    "        if len(seq) > 7000:\n",
    "            Exc_lengthened_RNAs[rec.id] = seq\n",
    "               "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "00d3deea",
   "metadata": {},
   "outputs": [],
   "source": [
    "Exc_lengthened_RNAs_df = pd.DataFrame([list(Exc_lengthened_RNAs.keys()), list(Exc_lengthened_RNAs.values())]).T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "dd89cda6",
   "metadata": {},
   "outputs": [],
   "source": [
    "Exc_lengthened_RNAs_df.to_csv('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/Exc_lengthened_PAs.tsv',\n",
    "                              sep='\\t', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "45d9767a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2186"
      ]
     },
     "execution_count": 65,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SeqIO.convert('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/Exc_lengthened_PAs.tsv',\n",
    "              'tab', '/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/Excitatory_C9ALSvsCTRL/Exc_lengthened_PAs_seq.fa', 'fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "efa3e264",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
