{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f930985c",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "c8ee6a41",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAS_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Mean RPM</th>\n",
       "      <th>Intron/exon location</th>\n",
       "      <th>Ensemble ID</th>\n",
       "      <th>RefSeq Gene ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>FAMTOM ID</th>\n",
       "      <th>FANTOM Category</th>\n",
       "      <th>Extension</th>\n",
       "      <th>PAS_type</th>\n",
       "      <th>PSE</th>\n",
       "      <th>PAS Signal</th>\n",
       "      <th>Conservation</th>\n",
       "      <th>intergenic_TE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:564599:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564599</td>\n",
       "      <td>+</td>\n",
       "      <td>3.333397</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>49.53%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:564629:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564629</td>\n",
       "      <td>+</td>\n",
       "      <td>1.201574</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:564664:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564664</td>\n",
       "      <td>+</td>\n",
       "      <td>14.856402</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:564708:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564708</td>\n",
       "      <td>+</td>\n",
       "      <td>1.872293</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:564952:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564952</td>\n",
       "      <td>+</td>\n",
       "      <td>1.991760</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>YES</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>32.71%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311589</th>\n",
       "      <td>chrY:27546122:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27546122</td>\n",
       "      <td>-</td>\n",
       "      <td>5.550364</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>11.21%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311590</th>\n",
       "      <td>chrY:28536584:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28536584</td>\n",
       "      <td>-</td>\n",
       "      <td>1.619023</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>AUUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311591</th>\n",
       "      <td>chrY:28548425:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28548425</td>\n",
       "      <td>-</td>\n",
       "      <td>0.809998</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.54%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311592</th>\n",
       "      <td>chrY:28671979:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28671979</td>\n",
       "      <td>-</td>\n",
       "      <td>0.985659</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>8.41%</td>\n",
       "      <td>Arich</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311593</th>\n",
       "      <td>chrY:28772221:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28772221</td>\n",
       "      <td>-</td>\n",
       "      <td>11.041395</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>60.75%</td>\n",
       "      <td>AAUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311594 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PAS_ID Chromosome  Position Strand   Mean RPM  \\\n",
       "0         chr1:564599:+       chr1    564599      +   3.333397   \n",
       "1         chr1:564629:+       chr1    564629      +   1.201574   \n",
       "2         chr1:564664:+       chr1    564664      +  14.856402   \n",
       "3         chr1:564708:+       chr1    564708      +   1.872293   \n",
       "4         chr1:564952:+       chr1    564952      +   1.991760   \n",
       "...                 ...        ...       ...    ...        ...   \n",
       "311589  chrY:27546122:-       chrY  27546122      -   5.550364   \n",
       "311590  chrY:28536584:-       chrY  28536584      -   1.619023   \n",
       "311591  chrY:28548425:-       chrY  28548425      -   0.809998   \n",
       "311592  chrY:28671979:-       chrY  28671979      -   0.985659   \n",
       "311593  chrY:28772221:-       chrY  28772221      -  11.041395   \n",
       "\n",
       "       Intron/exon location      Ensemble ID RefSeq Gene ID Gene Symbol  \\\n",
       "0               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "1               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "2               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "3               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "4               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "...                     ...              ...            ...         ...   \n",
       "311589           intergenic              NaN            NaN         NaN   \n",
       "311590           intergenic              NaN            NaN         NaN   \n",
       "311591           intergenic              NaN            NaN         NaN   \n",
       "311592           intergenic              NaN            NaN         NaN   \n",
       "311593           intergenic              NaN            NaN         NaN   \n",
       "\n",
       "                                                Gene Name FAMTOM ID  \\\n",
       "0       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "1       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "2       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "3       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "4       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "...                                                   ...       ...   \n",
       "311589                                                NaN        na   \n",
       "311590                                                NaN        na   \n",
       "311591                                                NaN        na   \n",
       "311592                                                NaN        na   \n",
       "311593                                                NaN        na   \n",
       "\n",
       "       FANTOM Category Extension    PAS_type     PSE PAS Signal Conservation  \\\n",
       "0                   na        NO  Pseudogene  49.53%      NoPAS           No   \n",
       "1                   na        NO  Pseudogene   3.74%      NoPAS           No   \n",
       "2                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "3                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "4                   na       YES  Pseudogene  32.71%      NoPAS           No   \n",
       "...                ...       ...         ...     ...        ...          ...   \n",
       "311589              na        NO  intergenic  11.21%   OtherPAS           No   \n",
       "311590              na        NO  intergenic   3.74%     AUUAAA           No   \n",
       "311591              na        NO  intergenic   6.54%   OtherPAS           No   \n",
       "311592              na        NO  intergenic   8.41%      Arich           No   \n",
       "311593              na        NO  intergenic  60.75%     AAUAAA           No   \n",
       "\n",
       "       intergenic_TE  \n",
       "0                 NO  \n",
       "1                 NO  \n",
       "2                 NO  \n",
       "3                 NO  \n",
       "4                 NO  \n",
       "...              ...  \n",
       "311589           YES  \n",
       "311590            NO  \n",
       "311591           YES  \n",
       "311592            NO  \n",
       "311593            NO  \n",
       "\n",
       "[311594 rows x 18 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tst = pd.read_csv('/data/APAproject/pre_qual/apatlas/human.PAS.txt', sep='\\t')\n",
    "tst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "04dd9783",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Intron             78002\n",
       "3'UTR(M)           76167\n",
       "intergenic         61668\n",
       "LncRNA(FANTOM5)    51535\n",
       "3'UTR(F)           13623\n",
       "3'UTR(L)           13623\n",
       "CDS                 6717\n",
       "Pseudogene          6257\n",
       "3'UTR(S)            3158\n",
       "5UTR                 844\n",
       "Name: PAS_type, dtype: int64"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tst['PAS_type'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "ff1438bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# hg19_coordinates = pd.DataFrame([tst['Chromosome'], tst['Position'], tst['Position']])\n",
    "hg19_coordinates = hg19_coordinates.T \n",
    "hg19_coordinates.to_csv('/data/APAproject/pre_qual/apatlas/PolyAdb_3_hg38_coordinates.bed', sep='\\t' , index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "8b61e8e8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>pos1</th>\n",
       "      <th>pos2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120908277</td>\n",
       "      <td>120908277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120908277</td>\n",
       "      <td>120908277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147919834</td>\n",
       "      <td>147919834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149665879</td>\n",
       "      <td>149665879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150219070</td>\n",
       "      <td>150219070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142582600</td>\n",
       "      <td>142582600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142723951</td>\n",
       "      <td>142723951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142715897</td>\n",
       "      <td>142715897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149574315</td>\n",
       "      <td>149574315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14919272</td>\n",
       "      <td>14919272</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome       pos1       pos2\n",
       "0        chr1  120908277  120908277\n",
       "1        chr1  120908277  120908277\n",
       "2        chr1  147919834  147919834\n",
       "3        chr1  149665879  149665879\n",
       "4        chr1  150219070  150219070\n",
       "..        ...        ...        ...\n",
       "71       chrX  142582600  142582600\n",
       "72       chrX  142723951  142723951\n",
       "73       chrX  142715897  142715897\n",
       "74       chrX  149574315  149574315\n",
       "75       chrY   14919272   14919272\n",
       "\n",
       "[76 rows x 3 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## ok read in the failed positions to get rid of them then read in the new positions to attach to the table\n",
    "\n",
    "failed_positions = pd.read_csv('/data/APAproject/pre_qual/apatlas/failed_hg19_to_hg38_positions.bed', sep='\\t')\n",
    "failed_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "2bd41057",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chromosome</th>\n",
       "      <th>pos1</th>\n",
       "      <th>pos2</th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120908277</td>\n",
       "      <td>120908277</td>\n",
       "      <td>chr1120908277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>120908277</td>\n",
       "      <td>120908277</td>\n",
       "      <td>chr1120908277</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>147919834</td>\n",
       "      <td>147919834</td>\n",
       "      <td>chr1147919834</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>149665879</td>\n",
       "      <td>149665879</td>\n",
       "      <td>chr1149665879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>150219070</td>\n",
       "      <td>150219070</td>\n",
       "      <td>chr1150219070</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142582600</td>\n",
       "      <td>142582600</td>\n",
       "      <td>chrX142582600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142723951</td>\n",
       "      <td>142723951</td>\n",
       "      <td>chrX142723951</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>chrX</td>\n",
       "      <td>142715897</td>\n",
       "      <td>142715897</td>\n",
       "      <td>chrX142715897</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>chrX</td>\n",
       "      <td>149574315</td>\n",
       "      <td>149574315</td>\n",
       "      <td>chrX149574315</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75</th>\n",
       "      <td>chrY</td>\n",
       "      <td>14919272</td>\n",
       "      <td>14919272</td>\n",
       "      <td>chrY14919272</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>76 rows Ã— 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   chromosome       pos1       pos2     identifier\n",
       "0        chr1  120908277  120908277  chr1120908277\n",
       "1        chr1  120908277  120908277  chr1120908277\n",
       "2        chr1  147919834  147919834  chr1147919834\n",
       "3        chr1  149665879  149665879  chr1149665879\n",
       "4        chr1  150219070  150219070  chr1150219070\n",
       "..        ...        ...        ...            ...\n",
       "71       chrX  142582600  142582600  chrX142582600\n",
       "72       chrX  142723951  142723951  chrX142723951\n",
       "73       chrX  142715897  142715897  chrX142715897\n",
       "74       chrX  149574315  149574315  chrX149574315\n",
       "75       chrY   14919272   14919272   chrY14919272\n",
       "\n",
       "[76 rows x 4 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "failed_positions['pos1'] = failed_positions['pos1'].map(str)\n",
    "failed_positions['identifier'] = failed_positions.chromosome.str.cat(failed_positions.pos1) \n",
    "failed_positions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "46918d0e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAS_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Mean RPM</th>\n",
       "      <th>Intron/exon location</th>\n",
       "      <th>Ensemble ID</th>\n",
       "      <th>RefSeq Gene ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>FAMTOM ID</th>\n",
       "      <th>FANTOM Category</th>\n",
       "      <th>Extension</th>\n",
       "      <th>PAS_type</th>\n",
       "      <th>PSE</th>\n",
       "      <th>PAS Signal</th>\n",
       "      <th>Conservation</th>\n",
       "      <th>intergenic_TE</th>\n",
       "      <th>pos1</th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:564599:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564599</td>\n",
       "      <td>+</td>\n",
       "      <td>3.333397</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>49.53%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564599</td>\n",
       "      <td>chr1564599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:564629:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564629</td>\n",
       "      <td>+</td>\n",
       "      <td>1.201574</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564629</td>\n",
       "      <td>chr1564629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:564664:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564664</td>\n",
       "      <td>+</td>\n",
       "      <td>14.856402</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564664</td>\n",
       "      <td>chr1564664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:564708:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564708</td>\n",
       "      <td>+</td>\n",
       "      <td>1.872293</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564708</td>\n",
       "      <td>chr1564708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:564952:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564952</td>\n",
       "      <td>+</td>\n",
       "      <td>1.991760</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>YES</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>32.71%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564952</td>\n",
       "      <td>chr1564952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311589</th>\n",
       "      <td>chrY:27546122:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27546122</td>\n",
       "      <td>-</td>\n",
       "      <td>5.550364</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>11.21%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>27546122</td>\n",
       "      <td>chrY27546122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311590</th>\n",
       "      <td>chrY:28536584:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28536584</td>\n",
       "      <td>-</td>\n",
       "      <td>1.619023</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>AUUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28536584</td>\n",
       "      <td>chrY28536584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311591</th>\n",
       "      <td>chrY:28548425:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28548425</td>\n",
       "      <td>-</td>\n",
       "      <td>0.809998</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.54%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>28548425</td>\n",
       "      <td>chrY28548425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311592</th>\n",
       "      <td>chrY:28671979:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28671979</td>\n",
       "      <td>-</td>\n",
       "      <td>0.985659</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>8.41%</td>\n",
       "      <td>Arich</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28671979</td>\n",
       "      <td>chrY28671979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311593</th>\n",
       "      <td>chrY:28772221:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28772221</td>\n",
       "      <td>-</td>\n",
       "      <td>11.041395</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>60.75%</td>\n",
       "      <td>AAUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28772221</td>\n",
       "      <td>chrY28772221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311594 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PAS_ID Chromosome  Position Strand   Mean RPM  \\\n",
       "0         chr1:564599:+       chr1    564599      +   3.333397   \n",
       "1         chr1:564629:+       chr1    564629      +   1.201574   \n",
       "2         chr1:564664:+       chr1    564664      +  14.856402   \n",
       "3         chr1:564708:+       chr1    564708      +   1.872293   \n",
       "4         chr1:564952:+       chr1    564952      +   1.991760   \n",
       "...                 ...        ...       ...    ...        ...   \n",
       "311589  chrY:27546122:-       chrY  27546122      -   5.550364   \n",
       "311590  chrY:28536584:-       chrY  28536584      -   1.619023   \n",
       "311591  chrY:28548425:-       chrY  28548425      -   0.809998   \n",
       "311592  chrY:28671979:-       chrY  28671979      -   0.985659   \n",
       "311593  chrY:28772221:-       chrY  28772221      -  11.041395   \n",
       "\n",
       "       Intron/exon location      Ensemble ID RefSeq Gene ID Gene Symbol  \\\n",
       "0               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "1               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "2               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "3               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "4               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "...                     ...              ...            ...         ...   \n",
       "311589           intergenic              NaN            NaN         NaN   \n",
       "311590           intergenic              NaN            NaN         NaN   \n",
       "311591           intergenic              NaN            NaN         NaN   \n",
       "311592           intergenic              NaN            NaN         NaN   \n",
       "311593           intergenic              NaN            NaN         NaN   \n",
       "\n",
       "                                                Gene Name FAMTOM ID  \\\n",
       "0       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "1       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "2       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "3       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "4       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "...                                                   ...       ...   \n",
       "311589                                                NaN        na   \n",
       "311590                                                NaN        na   \n",
       "311591                                                NaN        na   \n",
       "311592                                                NaN        na   \n",
       "311593                                                NaN        na   \n",
       "\n",
       "       FANTOM Category Extension    PAS_type     PSE PAS Signal Conservation  \\\n",
       "0                   na        NO  Pseudogene  49.53%      NoPAS           No   \n",
       "1                   na        NO  Pseudogene   3.74%      NoPAS           No   \n",
       "2                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "3                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "4                   na       YES  Pseudogene  32.71%      NoPAS           No   \n",
       "...                ...       ...         ...     ...        ...          ...   \n",
       "311589              na        NO  intergenic  11.21%   OtherPAS           No   \n",
       "311590              na        NO  intergenic   3.74%     AUUAAA           No   \n",
       "311591              na        NO  intergenic   6.54%   OtherPAS           No   \n",
       "311592              na        NO  intergenic   8.41%      Arich           No   \n",
       "311593              na        NO  intergenic  60.75%     AAUAAA           No   \n",
       "\n",
       "       intergenic_TE      pos1    identifier  \n",
       "0                 NO    564599    chr1564599  \n",
       "1                 NO    564629    chr1564629  \n",
       "2                 NO    564664    chr1564664  \n",
       "3                 NO    564708    chr1564708  \n",
       "4                 NO    564952    chr1564952  \n",
       "...              ...       ...           ...  \n",
       "311589           YES  27546122  chrY27546122  \n",
       "311590            NO  28536584  chrY28536584  \n",
       "311591           YES  28548425  chrY28548425  \n",
       "311592            NO  28671979  chrY28671979  \n",
       "311593            NO  28772221  chrY28772221  \n",
       "\n",
       "[311594 rows x 20 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tst['pos1'] = tst['Position'].map(str)\n",
    "tst['identifier'] = tst.Chromosome.str.cat(tst.pos1) \n",
    "tst"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "ef2d2cab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAS_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Mean RPM</th>\n",
       "      <th>Intron/exon location</th>\n",
       "      <th>Ensemble ID</th>\n",
       "      <th>RefSeq Gene ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>FAMTOM ID</th>\n",
       "      <th>FANTOM Category</th>\n",
       "      <th>Extension</th>\n",
       "      <th>PAS_type</th>\n",
       "      <th>PSE</th>\n",
       "      <th>PAS Signal</th>\n",
       "      <th>Conservation</th>\n",
       "      <th>intergenic_TE</th>\n",
       "      <th>pos1</th>\n",
       "      <th>identifier</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:564599:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564599</td>\n",
       "      <td>+</td>\n",
       "      <td>3.333397</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>49.53%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564599</td>\n",
       "      <td>chr1564599</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:564629:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564629</td>\n",
       "      <td>+</td>\n",
       "      <td>1.201574</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564629</td>\n",
       "      <td>chr1564629</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:564664:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564664</td>\n",
       "      <td>+</td>\n",
       "      <td>14.856402</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564664</td>\n",
       "      <td>chr1564664</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:564708:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564708</td>\n",
       "      <td>+</td>\n",
       "      <td>1.872293</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564708</td>\n",
       "      <td>chr1564708</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:564952:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564952</td>\n",
       "      <td>+</td>\n",
       "      <td>1.991760</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>YES</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>32.71%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564952</td>\n",
       "      <td>chr1564952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311589</th>\n",
       "      <td>chrY:27546122:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27546122</td>\n",
       "      <td>-</td>\n",
       "      <td>5.550364</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>11.21%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>27546122</td>\n",
       "      <td>chrY27546122</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311590</th>\n",
       "      <td>chrY:28536584:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28536584</td>\n",
       "      <td>-</td>\n",
       "      <td>1.619023</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>AUUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28536584</td>\n",
       "      <td>chrY28536584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311591</th>\n",
       "      <td>chrY:28548425:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28548425</td>\n",
       "      <td>-</td>\n",
       "      <td>0.809998</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.54%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>28548425</td>\n",
       "      <td>chrY28548425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311592</th>\n",
       "      <td>chrY:28671979:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28671979</td>\n",
       "      <td>-</td>\n",
       "      <td>0.985659</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>8.41%</td>\n",
       "      <td>Arich</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28671979</td>\n",
       "      <td>chrY28671979</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311593</th>\n",
       "      <td>chrY:28772221:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28772221</td>\n",
       "      <td>-</td>\n",
       "      <td>11.041395</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>60.75%</td>\n",
       "      <td>AAUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28772221</td>\n",
       "      <td>chrY28772221</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311518 rows Ã— 20 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PAS_ID Chromosome  Position Strand   Mean RPM  \\\n",
       "0         chr1:564599:+       chr1    564599      +   3.333397   \n",
       "1         chr1:564629:+       chr1    564629      +   1.201574   \n",
       "2         chr1:564664:+       chr1    564664      +  14.856402   \n",
       "3         chr1:564708:+       chr1    564708      +   1.872293   \n",
       "4         chr1:564952:+       chr1    564952      +   1.991760   \n",
       "...                 ...        ...       ...    ...        ...   \n",
       "311589  chrY:27546122:-       chrY  27546122      -   5.550364   \n",
       "311590  chrY:28536584:-       chrY  28536584      -   1.619023   \n",
       "311591  chrY:28548425:-       chrY  28548425      -   0.809998   \n",
       "311592  chrY:28671979:-       chrY  28671979      -   0.985659   \n",
       "311593  chrY:28772221:-       chrY  28772221      -  11.041395   \n",
       "\n",
       "       Intron/exon location      Ensemble ID RefSeq Gene ID Gene Symbol  \\\n",
       "0               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "1               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "2               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "3               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "4               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "...                     ...              ...            ...         ...   \n",
       "311589           intergenic              NaN            NaN         NaN   \n",
       "311590           intergenic              NaN            NaN         NaN   \n",
       "311591           intergenic              NaN            NaN         NaN   \n",
       "311592           intergenic              NaN            NaN         NaN   \n",
       "311593           intergenic              NaN            NaN         NaN   \n",
       "\n",
       "                                                Gene Name FAMTOM ID  \\\n",
       "0       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "1       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "2       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "3       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "4       mitochondrially encoded NADH:ubiquinone oxidor...        na   \n",
       "...                                                   ...       ...   \n",
       "311589                                                NaN        na   \n",
       "311590                                                NaN        na   \n",
       "311591                                                NaN        na   \n",
       "311592                                                NaN        na   \n",
       "311593                                                NaN        na   \n",
       "\n",
       "       FANTOM Category Extension    PAS_type     PSE PAS Signal Conservation  \\\n",
       "0                   na        NO  Pseudogene  49.53%      NoPAS           No   \n",
       "1                   na        NO  Pseudogene   3.74%      NoPAS           No   \n",
       "2                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "3                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "4                   na       YES  Pseudogene  32.71%      NoPAS           No   \n",
       "...                ...       ...         ...     ...        ...          ...   \n",
       "311589              na        NO  intergenic  11.21%   OtherPAS           No   \n",
       "311590              na        NO  intergenic   3.74%     AUUAAA           No   \n",
       "311591              na        NO  intergenic   6.54%   OtherPAS           No   \n",
       "311592              na        NO  intergenic   8.41%      Arich           No   \n",
       "311593              na        NO  intergenic  60.75%     AAUAAA           No   \n",
       "\n",
       "       intergenic_TE      pos1    identifier  \n",
       "0                 NO    564599    chr1564599  \n",
       "1                 NO    564629    chr1564629  \n",
       "2                 NO    564664    chr1564664  \n",
       "3                 NO    564708    chr1564708  \n",
       "4                 NO    564952    chr1564952  \n",
       "...              ...       ...           ...  \n",
       "311589           YES  27546122  chrY27546122  \n",
       "311590            NO  28536584  chrY28536584  \n",
       "311591           YES  28548425  chrY28548425  \n",
       "311592            NO  28671979  chrY28671979  \n",
       "311593            NO  28772221  chrY28772221  \n",
       "\n",
       "[311518 rows x 20 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "poly_atlas = tst[~tst.identifier.isin(failed_positions.identifier.values)]\n",
    "poly_atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "059a057f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos1</th>\n",
       "      <th>pos2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629219</td>\n",
       "      <td>629219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629249</td>\n",
       "      <td>629249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629284</td>\n",
       "      <td>629284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629328</td>\n",
       "      <td>629328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>629572</td>\n",
       "      <td>629572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311513</th>\n",
       "      <td>chrY</td>\n",
       "      <td>25399975</td>\n",
       "      <td>25399975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311514</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26390437</td>\n",
       "      <td>26390437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311515</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26402278</td>\n",
       "      <td>26402278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311516</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26525832</td>\n",
       "      <td>26525832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311517</th>\n",
       "      <td>chrY</td>\n",
       "      <td>26626074</td>\n",
       "      <td>26626074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311518 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       Chromosome      pos1      pos2\n",
       "0            chr1    629219    629219\n",
       "1            chr1    629249    629249\n",
       "2            chr1    629284    629284\n",
       "3            chr1    629328    629328\n",
       "4            chr1    629572    629572\n",
       "...           ...       ...       ...\n",
       "311513       chrY  25399975  25399975\n",
       "311514       chrY  26390437  26390437\n",
       "311515       chrY  26402278  26402278\n",
       "311516       chrY  26525832  26525832\n",
       "311517       chrY  26626074  26626074\n",
       "\n",
       "[311518 rows x 3 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "##read in the converted positions\n",
    "hg38_coordinates = pd.read_csv('/data/APAproject/pre_qual/apatlas/hglft_genome_2be15_ea7880.bed', sep='\\t')\n",
    "hg38_coordinates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "44558500",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/aiden/anaconda3/envs/r_seurat/lib/python3.9/site-packages/pandas/core/frame.py:3607: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self._set_item(key, value)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PAS_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Mean RPM</th>\n",
       "      <th>Intron/exon location</th>\n",
       "      <th>Ensemble ID</th>\n",
       "      <th>RefSeq Gene ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>Gene Name</th>\n",
       "      <th>...</th>\n",
       "      <th>FANTOM Category</th>\n",
       "      <th>Extension</th>\n",
       "      <th>PAS_type</th>\n",
       "      <th>PSE</th>\n",
       "      <th>PAS Signal</th>\n",
       "      <th>Conservation</th>\n",
       "      <th>intergenic_TE</th>\n",
       "      <th>pos1</th>\n",
       "      <th>identifier</th>\n",
       "      <th>hg38_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1:564599:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564599</td>\n",
       "      <td>+</td>\n",
       "      <td>3.333397</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>49.53%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564599</td>\n",
       "      <td>chr1564599</td>\n",
       "      <td>629219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1:564629:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564629</td>\n",
       "      <td>+</td>\n",
       "      <td>1.201574</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564629</td>\n",
       "      <td>chr1564629</td>\n",
       "      <td>629249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1:564664:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564664</td>\n",
       "      <td>+</td>\n",
       "      <td>14.856402</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564664</td>\n",
       "      <td>chr1564664</td>\n",
       "      <td>629284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1:564708:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564708</td>\n",
       "      <td>+</td>\n",
       "      <td>1.872293</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564708</td>\n",
       "      <td>chr1564708</td>\n",
       "      <td>629328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1:564952:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564952</td>\n",
       "      <td>+</td>\n",
       "      <td>1.991760</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>mitochondrially encoded NADH:ubiquinone oxidor...</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>YES</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>32.71%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564952</td>\n",
       "      <td>chr1564952</td>\n",
       "      <td>629572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311589</th>\n",
       "      <td>chrY:27546122:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27546122</td>\n",
       "      <td>-</td>\n",
       "      <td>5.550364</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>11.21%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>27546122</td>\n",
       "      <td>chrY27546122</td>\n",
       "      <td>25399975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311590</th>\n",
       "      <td>chrY:28536584:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28536584</td>\n",
       "      <td>-</td>\n",
       "      <td>1.619023</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>AUUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28536584</td>\n",
       "      <td>chrY28536584</td>\n",
       "      <td>26390437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311591</th>\n",
       "      <td>chrY:28548425:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28548425</td>\n",
       "      <td>-</td>\n",
       "      <td>0.809998</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.54%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>28548425</td>\n",
       "      <td>chrY28548425</td>\n",
       "      <td>26402278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311592</th>\n",
       "      <td>chrY:28671979:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28671979</td>\n",
       "      <td>-</td>\n",
       "      <td>0.985659</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>8.41%</td>\n",
       "      <td>Arich</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28671979</td>\n",
       "      <td>chrY28671979</td>\n",
       "      <td>26525832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311593</th>\n",
       "      <td>chrY:28772221:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28772221</td>\n",
       "      <td>-</td>\n",
       "      <td>11.041395</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>60.75%</td>\n",
       "      <td>AAUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28772221</td>\n",
       "      <td>chrY28772221</td>\n",
       "      <td>26626074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311518 rows Ã— 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 PAS_ID Chromosome  Position Strand   Mean RPM  \\\n",
       "0         chr1:564599:+       chr1    564599      +   3.333397   \n",
       "1         chr1:564629:+       chr1    564629      +   1.201574   \n",
       "2         chr1:564664:+       chr1    564664      +  14.856402   \n",
       "3         chr1:564708:+       chr1    564708      +   1.872293   \n",
       "4         chr1:564952:+       chr1    564952      +   1.991760   \n",
       "...                 ...        ...       ...    ...        ...   \n",
       "311589  chrY:27546122:-       chrY  27546122      -   5.550364   \n",
       "311590  chrY:28536584:-       chrY  28536584      -   1.619023   \n",
       "311591  chrY:28548425:-       chrY  28548425      -   0.809998   \n",
       "311592  chrY:28671979:-       chrY  28671979      -   0.985659   \n",
       "311593  chrY:28772221:-       chrY  28772221      -  11.041395   \n",
       "\n",
       "       Intron/exon location      Ensemble ID RefSeq Gene ID Gene Symbol  \\\n",
       "0               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "1               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "2               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "3               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "4               Single exon  ENSG00000225972             na    MTND1P23   \n",
       "...                     ...              ...            ...         ...   \n",
       "311589           intergenic              NaN            NaN         NaN   \n",
       "311590           intergenic              NaN            NaN         NaN   \n",
       "311591           intergenic              NaN            NaN         NaN   \n",
       "311592           intergenic              NaN            NaN         NaN   \n",
       "311593           intergenic              NaN            NaN         NaN   \n",
       "\n",
       "                                                Gene Name  ...  \\\n",
       "0       mitochondrially encoded NADH:ubiquinone oxidor...  ...   \n",
       "1       mitochondrially encoded NADH:ubiquinone oxidor...  ...   \n",
       "2       mitochondrially encoded NADH:ubiquinone oxidor...  ...   \n",
       "3       mitochondrially encoded NADH:ubiquinone oxidor...  ...   \n",
       "4       mitochondrially encoded NADH:ubiquinone oxidor...  ...   \n",
       "...                                                   ...  ...   \n",
       "311589                                                NaN  ...   \n",
       "311590                                                NaN  ...   \n",
       "311591                                                NaN  ...   \n",
       "311592                                                NaN  ...   \n",
       "311593                                                NaN  ...   \n",
       "\n",
       "       FANTOM Category Extension    PAS_type     PSE PAS Signal Conservation  \\\n",
       "0                   na        NO  Pseudogene  49.53%      NoPAS           No   \n",
       "1                   na        NO  Pseudogene   3.74%      NoPAS           No   \n",
       "2                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "3                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "4                   na       YES  Pseudogene  32.71%      NoPAS           No   \n",
       "...                ...       ...         ...     ...        ...          ...   \n",
       "311589              na        NO  intergenic  11.21%   OtherPAS           No   \n",
       "311590              na        NO  intergenic   3.74%     AUUAAA           No   \n",
       "311591              na        NO  intergenic   6.54%   OtherPAS           No   \n",
       "311592              na        NO  intergenic   8.41%      Arich           No   \n",
       "311593              na        NO  intergenic  60.75%     AAUAAA           No   \n",
       "\n",
       "       intergenic_TE      pos1    identifier  hg38_pos  \n",
       "0                 NO    564599    chr1564599    629219  \n",
       "1                 NO    564629    chr1564629    629249  \n",
       "2                 NO    564664    chr1564664    629284  \n",
       "3                 NO    564708    chr1564708    629328  \n",
       "4                 NO    564952    chr1564952    629572  \n",
       "...              ...       ...           ...       ...  \n",
       "311589           YES  27546122  chrY27546122  25399975  \n",
       "311590            NO  28536584  chrY28536584  26390437  \n",
       "311591           YES  28548425  chrY28548425  26402278  \n",
       "311592            NO  28671979  chrY28671979  26525832  \n",
       "311593            NO  28772221  chrY28772221  26626074  \n",
       "\n",
       "[311518 rows x 21 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "poly_atlas['hg38_pos'] = hg38_coordinates['pos1'].values\n",
    "poly_atlas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "4dde1947",
   "metadata": {},
   "outputs": [],
   "source": [
    "poly_atlas.to_csv('/data/APAproject/pre_qual/apatlas/PolyA_db_v3_hg38_mapped.bed', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "984f6779",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>UTR3</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>intron</th>\n",
       "      <th>exon</th>\n",
       "      <th>CDS</th>\n",
       "      <th>pA_motif</th>\n",
       "      <th>pA_stretch</th>\n",
       "      <th>pT_stretch</th>\n",
       "      <th>Junctions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARGLU1:chr13:106543856-106557719:-1</th>\n",
       "      <td>chr13</td>\n",
       "      <td>106543856</td>\n",
       "      <td>106557719</td>\n",
       "      <td>13864</td>\n",
       "      <td>-</td>\n",
       "      <td>ARGLU1</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DST:chr6:56699737-56852174:-1</th>\n",
       "      <td>chr6</td>\n",
       "      <td>56699737</td>\n",
       "      <td>56852174</td>\n",
       "      <td>152438</td>\n",
       "      <td>-</td>\n",
       "      <td>DST</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TERF2IP:chr16:75647773-75657331:1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>75647773</td>\n",
       "      <td>75657331</td>\n",
       "      <td>9559</td>\n",
       "      <td>+</td>\n",
       "      <td>TERF2IP</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF708:chr19:21293722-21309554:-1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>21293722</td>\n",
       "      <td>21309554</td>\n",
       "      <td>15833</td>\n",
       "      <td>-</td>\n",
       "      <td>ZNF708</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2E:chr7:105014280-105062294:1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>105014280</td>\n",
       "      <td>105062294</td>\n",
       "      <td>48015</td>\n",
       "      <td>+</td>\n",
       "      <td>KMT2E</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>YES</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TAS2R14:chr12:10948138-10948978:-1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>10948138</td>\n",
       "      <td>10948978</td>\n",
       "      <td>841</td>\n",
       "      <td>-</td>\n",
       "      <td>PRH1,TAS2R14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRH1:chr12:10948138-10948978:-1</th>\n",
       "      <td>chr12</td>\n",
       "      <td>10948138</td>\n",
       "      <td>10948978</td>\n",
       "      <td>841</td>\n",
       "      <td>-</td>\n",
       "      <td>PRH1,TAS2R14</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANK2:chr4:112982777-112983569:1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>112982777</td>\n",
       "      <td>112983569</td>\n",
       "      <td>793</td>\n",
       "      <td>+</td>\n",
       "      <td>ANK2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NBPF11:chr1:148127499-148128417:-1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>148127499</td>\n",
       "      <td>148128417</td>\n",
       "      <td>919</td>\n",
       "      <td>-</td>\n",
       "      <td>NBPF11</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FNBP1:chr9:129920974-129922138:-1</th>\n",
       "      <td>chr9</td>\n",
       "      <td>129920974</td>\n",
       "      <td>129922138</td>\n",
       "      <td>1165</td>\n",
       "      <td>-</td>\n",
       "      <td>FNBP1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>291409 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    seqnames      start        end   width  \\\n",
       "ARGLU1:chr13:106543856-106557719:-1    chr13  106543856  106557719   13864   \n",
       "DST:chr6:56699737-56852174:-1           chr6   56699737   56852174  152438   \n",
       "TERF2IP:chr16:75647773-75657331:1      chr16   75647773   75657331    9559   \n",
       "ZNF708:chr19:21293722-21309554:-1      chr19   21293722   21309554   15833   \n",
       "KMT2E:chr7:105014280-105062294:1        chr7  105014280  105062294   48015   \n",
       "...                                      ...        ...        ...     ...   \n",
       "TAS2R14:chr12:10948138-10948978:-1     chr12   10948138   10948978     841   \n",
       "PRH1:chr12:10948138-10948978:-1        chr12   10948138   10948978     841   \n",
       "ANK2:chr4:112982777-112983569:1         chr4  112982777  112983569     793   \n",
       "NBPF11:chr1:148127499-148128417:-1      chr1  148127499  148128417     919   \n",
       "FNBP1:chr9:129920974-129922138:-1       chr9  129920974  129922138    1165   \n",
       "\n",
       "                                    strand       gene_id UTR3 UTR5 intron  \\\n",
       "ARGLU1:chr13:106543856-106557719:-1      -        ARGLU1  YES  NaN    NaN   \n",
       "DST:chr6:56699737-56852174:-1            -           DST  NaN  YES    YES   \n",
       "TERF2IP:chr16:75647773-75657331:1        +       TERF2IP  YES  NaN    NaN   \n",
       "ZNF708:chr19:21293722-21309554:-1        -        ZNF708  YES  NaN    NaN   \n",
       "KMT2E:chr7:105014280-105062294:1         +         KMT2E  NaN  YES    YES   \n",
       "...                                    ...           ...  ...  ...    ...   \n",
       "TAS2R14:chr12:10948138-10948978:-1       -  PRH1,TAS2R14  NaN  NaN    YES   \n",
       "PRH1:chr12:10948138-10948978:-1          -  PRH1,TAS2R14  NaN  NaN    YES   \n",
       "ANK2:chr4:112982777-112983569:1          +          ANK2  NaN  NaN    YES   \n",
       "NBPF11:chr1:148127499-148128417:-1       -        NBPF11  NaN  NaN    YES   \n",
       "FNBP1:chr9:129920974-129922138:-1        -         FNBP1  NaN  NaN    YES   \n",
       "\n",
       "                                    exon  CDS  pA_motif  pA_stretch  \\\n",
       "ARGLU1:chr13:106543856-106557719:-1  NaN  NaN     False       False   \n",
       "DST:chr6:56699737-56852174:-1        YES  YES     False       False   \n",
       "TERF2IP:chr16:75647773-75657331:1    NaN  NaN     False       False   \n",
       "ZNF708:chr19:21293722-21309554:-1    NaN  NaN     False       False   \n",
       "KMT2E:chr7:105014280-105062294:1     YES  YES     False       False   \n",
       "...                                  ...  ...       ...         ...   \n",
       "TAS2R14:chr12:10948138-10948978:-1   NaN  NaN     False       False   \n",
       "PRH1:chr12:10948138-10948978:-1      NaN  NaN     False       False   \n",
       "ANK2:chr4:112982777-112983569:1      NaN  NaN     False        True   \n",
       "NBPF11:chr1:148127499-148128417:-1   NaN  NaN     False       False   \n",
       "FNBP1:chr9:129920974-129922138:-1    NaN  NaN     False       False   \n",
       "\n",
       "                                     pT_stretch         Junctions  \n",
       "ARGLU1:chr13:106543856-106557719:-1       False  across-junctions  \n",
       "DST:chr6:56699737-56852174:-1             False  across-junctions  \n",
       "TERF2IP:chr16:75647773-75657331:1         False  across-junctions  \n",
       "ZNF708:chr19:21293722-21309554:-1         False  across-junctions  \n",
       "KMT2E:chr7:105014280-105062294:1          False  across-junctions  \n",
       "...                                         ...               ...  \n",
       "TAS2R14:chr12:10948138-10948978:-1        False      no-junctions  \n",
       "PRH1:chr12:10948138-10948978:-1           False      no-junctions  \n",
       "ANK2:chr4:112982777-112983569:1           False      no-junctions  \n",
       "NBPF11:chr1:148127499-148128417:-1        False      no-junctions  \n",
       "FNBP1:chr9:129920974-129922138:-1         False      no-junctions  \n",
       "\n",
       "[291409 rows x 15 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "### lets read an example annotation table and map it to APA database\n",
    "polyA_atlas = pd.read_csv('/data/APAproject/pre_qual/apatlas/PolyA_db_v3_hg38_mapped.bed', sep='\\t')\n",
    "peaks = pd.read_csv('/data1/APA/Paul_ALS_Data/sierra_out/ALS_paul_peaks_annotated.txt', sep='\\t')\n",
    "peaks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "077c8a23",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1603.3375839455884"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks['width'].mean()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "f2c12ab9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "444.0"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks['width'].median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "a427ecc9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1109956"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "max(peaks['width'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "286325c8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min(peaks['width'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e2d4015d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "min(peaks[peaks['UTR3']=='YES']['width'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1aa7d7f1",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>UTR3</th>\n",
       "      <th>UTR5</th>\n",
       "      <th>intron</th>\n",
       "      <th>exon</th>\n",
       "      <th>CDS</th>\n",
       "      <th>pA_motif</th>\n",
       "      <th>pA_stretch</th>\n",
       "      <th>pT_stretch</th>\n",
       "      <th>Junctions</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ARGLU1:chr13:106543856-106557719:-1</th>\n",
       "      <td>chr13</td>\n",
       "      <td>106543856</td>\n",
       "      <td>106557719</td>\n",
       "      <td>13864</td>\n",
       "      <td>-</td>\n",
       "      <td>ARGLU1</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TERF2IP:chr16:75647773-75657331:1</th>\n",
       "      <td>chr16</td>\n",
       "      <td>75647773</td>\n",
       "      <td>75657331</td>\n",
       "      <td>9559</td>\n",
       "      <td>+</td>\n",
       "      <td>TERF2IP</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF708:chr19:21293722-21309554:-1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>21293722</td>\n",
       "      <td>21309554</td>\n",
       "      <td>15833</td>\n",
       "      <td>-</td>\n",
       "      <td>ZNF708</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KMT2E:chr7:105077148-105091364:1</th>\n",
       "      <td>chr7</td>\n",
       "      <td>105077148</td>\n",
       "      <td>105091364</td>\n",
       "      <td>14217</td>\n",
       "      <td>+</td>\n",
       "      <td>KMT2E</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>REX1BD:chr19:18589507-18592336:1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>18589507</td>\n",
       "      <td>18592336</td>\n",
       "      <td>2830</td>\n",
       "      <td>+</td>\n",
       "      <td>REX1BD</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TMEM165:chr4:55425535-55426436:1</th>\n",
       "      <td>chr4</td>\n",
       "      <td>55425535</td>\n",
       "      <td>55426436</td>\n",
       "      <td>902</td>\n",
       "      <td>+</td>\n",
       "      <td>TMEM165</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YWHAB:chr20:44907620-44908280:1</th>\n",
       "      <td>chr20</td>\n",
       "      <td>44907620</td>\n",
       "      <td>44908280</td>\n",
       "      <td>661</td>\n",
       "      <td>+</td>\n",
       "      <td>YWHAB</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GOLGA8B:chr15:34525631-34526133:-1</th>\n",
       "      <td>chr15</td>\n",
       "      <td>34525631</td>\n",
       "      <td>34526133</td>\n",
       "      <td>503</td>\n",
       "      <td>-</td>\n",
       "      <td>GOLGA8B</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>True</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>USP33:chr1:77711703-77715940:-1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>77711703</td>\n",
       "      <td>77715940</td>\n",
       "      <td>4238</td>\n",
       "      <td>-</td>\n",
       "      <td>USP33</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>across-junctions</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LIFR:chr5:38476383-38477217:-1</th>\n",
       "      <td>chr5</td>\n",
       "      <td>38476383</td>\n",
       "      <td>38477217</td>\n",
       "      <td>835</td>\n",
       "      <td>-</td>\n",
       "      <td>LIFR</td>\n",
       "      <td>YES</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>no-junctions</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>48612 rows Ã— 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                    seqnames      start        end  width  \\\n",
       "ARGLU1:chr13:106543856-106557719:-1    chr13  106543856  106557719  13864   \n",
       "TERF2IP:chr16:75647773-75657331:1      chr16   75647773   75657331   9559   \n",
       "ZNF708:chr19:21293722-21309554:-1      chr19   21293722   21309554  15833   \n",
       "KMT2E:chr7:105077148-105091364:1        chr7  105077148  105091364  14217   \n",
       "REX1BD:chr19:18589507-18592336:1       chr19   18589507   18592336   2830   \n",
       "...                                      ...        ...        ...    ...   \n",
       "TMEM165:chr4:55425535-55426436:1        chr4   55425535   55426436    902   \n",
       "YWHAB:chr20:44907620-44908280:1        chr20   44907620   44908280    661   \n",
       "GOLGA8B:chr15:34525631-34526133:-1     chr15   34525631   34526133    503   \n",
       "USP33:chr1:77711703-77715940:-1         chr1   77711703   77715940   4238   \n",
       "LIFR:chr5:38476383-38477217:-1          chr5   38476383   38477217    835   \n",
       "\n",
       "                                    strand  gene_id UTR3 UTR5 intron exon  \\\n",
       "ARGLU1:chr13:106543856-106557719:-1      -   ARGLU1  YES  NaN    NaN  NaN   \n",
       "TERF2IP:chr16:75647773-75657331:1        +  TERF2IP  YES  NaN    NaN  NaN   \n",
       "ZNF708:chr19:21293722-21309554:-1        -   ZNF708  YES  NaN    NaN  NaN   \n",
       "KMT2E:chr7:105077148-105091364:1         +    KMT2E  YES  NaN    NaN  NaN   \n",
       "REX1BD:chr19:18589507-18592336:1         +   REX1BD  YES  NaN    NaN  NaN   \n",
       "...                                    ...      ...  ...  ...    ...  ...   \n",
       "TMEM165:chr4:55425535-55426436:1         +  TMEM165  YES  NaN    NaN  NaN   \n",
       "YWHAB:chr20:44907620-44908280:1          +    YWHAB  YES  NaN    NaN  NaN   \n",
       "GOLGA8B:chr15:34525631-34526133:-1       -  GOLGA8B  YES  NaN    NaN  NaN   \n",
       "USP33:chr1:77711703-77715940:-1          -    USP33  YES  NaN    NaN  NaN   \n",
       "LIFR:chr5:38476383-38477217:-1           -     LIFR  YES  NaN    NaN  NaN   \n",
       "\n",
       "                                     CDS  pA_motif  pA_stretch  pT_stretch  \\\n",
       "ARGLU1:chr13:106543856-106557719:-1  NaN     False       False       False   \n",
       "TERF2IP:chr16:75647773-75657331:1    NaN     False       False       False   \n",
       "ZNF708:chr19:21293722-21309554:-1    NaN     False       False       False   \n",
       "KMT2E:chr7:105077148-105091364:1     NaN     False       False       False   \n",
       "REX1BD:chr19:18589507-18592336:1     NaN      True       False       False   \n",
       "...                                  ...       ...         ...         ...   \n",
       "TMEM165:chr4:55425535-55426436:1     NaN     False       False       False   \n",
       "YWHAB:chr20:44907620-44908280:1      NaN     False       False       False   \n",
       "GOLGA8B:chr15:34525631-34526133:-1   NaN      True       False       False   \n",
       "USP33:chr1:77711703-77715940:-1      NaN     False       False       False   \n",
       "LIFR:chr5:38476383-38477217:-1       NaN     False       False       False   \n",
       "\n",
       "                                            Junctions  \n",
       "ARGLU1:chr13:106543856-106557719:-1  across-junctions  \n",
       "TERF2IP:chr16:75647773-75657331:1    across-junctions  \n",
       "ZNF708:chr19:21293722-21309554:-1    across-junctions  \n",
       "KMT2E:chr7:105077148-105091364:1     across-junctions  \n",
       "REX1BD:chr19:18589507-18592336:1     across-junctions  \n",
       "...                                               ...  \n",
       "TMEM165:chr4:55425535-55426436:1         no-junctions  \n",
       "YWHAB:chr20:44907620-44908280:1          no-junctions  \n",
       "GOLGA8B:chr15:34525631-34526133:-1       no-junctions  \n",
       "USP33:chr1:77711703-77715940:-1      across-junctions  \n",
       "LIFR:chr5:38476383-38477217:-1           no-junctions  \n",
       "\n",
       "[48612 rows x 15 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks[peaks['UTR3']=='YES']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "b20d415f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "13666"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks[peaks['UTR3']=='YES']['gene_id'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b2cdc6ce",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "DLG2                          744\n",
       "CADM2                         725\n",
       "IL1RAPL1                      660\n",
       "CNTNAP2                       630\n",
       "CSMD1                         584\n",
       "                             ... \n",
       "ZNF467                          2\n",
       "H2AFJ                           2\n",
       "AC005871.2                      2\n",
       "BACH1,BACH1-IT1,AP000240.1      2\n",
       "BNIP1                           2\n",
       "Name: gene_id, Length: 15636, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_counts = peaks['gene_id'].value_counts()\n",
    "gene_counts[gene_counts > 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "090f2363",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "genes with one peak on their 3UTR....... 4217\n",
      "genes with two peaks on their 3UTR....... 2728\n",
      "genes with three peaks on their 3UTR....... 1581\n",
      "genes with four peaks on their 3UTR....... 1274\n",
      "genes with more than four peaks on their 3UTR....... 10053\n"
     ]
    }
   ],
   "source": [
    "print('genes with one peak on their 3UTR.......', len(gene_counts[gene_counts == 1]))\n",
    "print('genes with two peaks on their 3UTR.......',len(gene_counts[gene_counts == 2]))\n",
    "print('genes with three peaks on their 3UTR.......',len(gene_counts[gene_counts == 3]))\n",
    "print('genes with four peaks on their 3UTR.......',len(gene_counts[gene_counts == 4]))\n",
    "print('genes with more than four peaks on their 3UTR.......',len(gene_counts[gene_counts > 4]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "a9484f82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>PAS_ID</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Position</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Mean RPM</th>\n",
       "      <th>Intron/exon location</th>\n",
       "      <th>Ensemble ID</th>\n",
       "      <th>RefSeq Gene ID</th>\n",
       "      <th>Gene Symbol</th>\n",
       "      <th>...</th>\n",
       "      <th>FANTOM Category</th>\n",
       "      <th>Extension</th>\n",
       "      <th>PAS_type</th>\n",
       "      <th>PSE</th>\n",
       "      <th>PAS Signal</th>\n",
       "      <th>Conservation</th>\n",
       "      <th>intergenic_TE</th>\n",
       "      <th>pos1</th>\n",
       "      <th>identifier</th>\n",
       "      <th>hg38_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>chr1:564599:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564599</td>\n",
       "      <td>+</td>\n",
       "      <td>3.333397</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>49.53%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564599</td>\n",
       "      <td>chr1564599</td>\n",
       "      <td>629219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1:564629:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564629</td>\n",
       "      <td>+</td>\n",
       "      <td>1.201574</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564629</td>\n",
       "      <td>chr1564629</td>\n",
       "      <td>629249</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1:564664:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564664</td>\n",
       "      <td>+</td>\n",
       "      <td>14.856402</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564664</td>\n",
       "      <td>chr1564664</td>\n",
       "      <td>629284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1:564708:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564708</td>\n",
       "      <td>+</td>\n",
       "      <td>1.872293</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>4.67%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564708</td>\n",
       "      <td>chr1564708</td>\n",
       "      <td>629328</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1:564952:+</td>\n",
       "      <td>chr1</td>\n",
       "      <td>564952</td>\n",
       "      <td>+</td>\n",
       "      <td>1.991760</td>\n",
       "      <td>Single exon</td>\n",
       "      <td>ENSG00000225972</td>\n",
       "      <td>na</td>\n",
       "      <td>MTND1P23</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>YES</td>\n",
       "      <td>Pseudogene</td>\n",
       "      <td>32.71%</td>\n",
       "      <td>NoPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>564952</td>\n",
       "      <td>chr1564952</td>\n",
       "      <td>629572</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311513</th>\n",
       "      <td>311589</td>\n",
       "      <td>chrY:27546122:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>27546122</td>\n",
       "      <td>-</td>\n",
       "      <td>5.550364</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>11.21%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>27546122</td>\n",
       "      <td>chrY27546122</td>\n",
       "      <td>25399975</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311514</th>\n",
       "      <td>311590</td>\n",
       "      <td>chrY:28536584:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28536584</td>\n",
       "      <td>-</td>\n",
       "      <td>1.619023</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>3.74%</td>\n",
       "      <td>AUUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28536584</td>\n",
       "      <td>chrY28536584</td>\n",
       "      <td>26390437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311515</th>\n",
       "      <td>311591</td>\n",
       "      <td>chrY:28548425:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28548425</td>\n",
       "      <td>-</td>\n",
       "      <td>0.809998</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>6.54%</td>\n",
       "      <td>OtherPAS</td>\n",
       "      <td>No</td>\n",
       "      <td>YES</td>\n",
       "      <td>28548425</td>\n",
       "      <td>chrY28548425</td>\n",
       "      <td>26402278</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311516</th>\n",
       "      <td>311592</td>\n",
       "      <td>chrY:28671979:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28671979</td>\n",
       "      <td>-</td>\n",
       "      <td>0.985659</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>8.41%</td>\n",
       "      <td>Arich</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28671979</td>\n",
       "      <td>chrY28671979</td>\n",
       "      <td>26525832</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>311517</th>\n",
       "      <td>311593</td>\n",
       "      <td>chrY:28772221:-</td>\n",
       "      <td>chrY</td>\n",
       "      <td>28772221</td>\n",
       "      <td>-</td>\n",
       "      <td>11.041395</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>na</td>\n",
       "      <td>NO</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>60.75%</td>\n",
       "      <td>AAUAAA</td>\n",
       "      <td>No</td>\n",
       "      <td>NO</td>\n",
       "      <td>28772221</td>\n",
       "      <td>chrY28772221</td>\n",
       "      <td>26626074</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>311518 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        Unnamed: 0           PAS_ID Chromosome  Position Strand   Mean RPM  \\\n",
       "0                0    chr1:564599:+       chr1    564599      +   3.333397   \n",
       "1                1    chr1:564629:+       chr1    564629      +   1.201574   \n",
       "2                2    chr1:564664:+       chr1    564664      +  14.856402   \n",
       "3                3    chr1:564708:+       chr1    564708      +   1.872293   \n",
       "4                4    chr1:564952:+       chr1    564952      +   1.991760   \n",
       "...            ...              ...        ...       ...    ...        ...   \n",
       "311513      311589  chrY:27546122:-       chrY  27546122      -   5.550364   \n",
       "311514      311590  chrY:28536584:-       chrY  28536584      -   1.619023   \n",
       "311515      311591  chrY:28548425:-       chrY  28548425      -   0.809998   \n",
       "311516      311592  chrY:28671979:-       chrY  28671979      -   0.985659   \n",
       "311517      311593  chrY:28772221:-       chrY  28772221      -  11.041395   \n",
       "\n",
       "       Intron/exon location      Ensemble ID RefSeq Gene ID Gene Symbol  ...  \\\n",
       "0               Single exon  ENSG00000225972             na    MTND1P23  ...   \n",
       "1               Single exon  ENSG00000225972             na    MTND1P23  ...   \n",
       "2               Single exon  ENSG00000225972             na    MTND1P23  ...   \n",
       "3               Single exon  ENSG00000225972             na    MTND1P23  ...   \n",
       "4               Single exon  ENSG00000225972             na    MTND1P23  ...   \n",
       "...                     ...              ...            ...         ...  ...   \n",
       "311513           intergenic              NaN            NaN         NaN  ...   \n",
       "311514           intergenic              NaN            NaN         NaN  ...   \n",
       "311515           intergenic              NaN            NaN         NaN  ...   \n",
       "311516           intergenic              NaN            NaN         NaN  ...   \n",
       "311517           intergenic              NaN            NaN         NaN  ...   \n",
       "\n",
       "       FANTOM Category Extension    PAS_type     PSE PAS Signal Conservation  \\\n",
       "0                   na        NO  Pseudogene  49.53%      NoPAS           No   \n",
       "1                   na        NO  Pseudogene   3.74%      NoPAS           No   \n",
       "2                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "3                   na        NO  Pseudogene   4.67%      NoPAS           No   \n",
       "4                   na       YES  Pseudogene  32.71%      NoPAS           No   \n",
       "...                ...       ...         ...     ...        ...          ...   \n",
       "311513              na        NO  intergenic  11.21%   OtherPAS           No   \n",
       "311514              na        NO  intergenic   3.74%     AUUAAA           No   \n",
       "311515              na        NO  intergenic   6.54%   OtherPAS           No   \n",
       "311516              na        NO  intergenic   8.41%      Arich           No   \n",
       "311517              na        NO  intergenic  60.75%     AAUAAA           No   \n",
       "\n",
       "       intergenic_TE      pos1    identifier  hg38_pos  \n",
       "0                 NO    564599    chr1564599    629219  \n",
       "1                 NO    564629    chr1564629    629249  \n",
       "2                 NO    564664    chr1564664    629284  \n",
       "3                 NO    564708    chr1564708    629328  \n",
       "4                 NO    564952    chr1564952    629572  \n",
       "...              ...       ...           ...       ...  \n",
       "311513           YES  27546122  chrY27546122  25399975  \n",
       "311514            NO  28536584  chrY28536584  26390437  \n",
       "311515           YES  28548425  chrY28548425  26402278  \n",
       "311516            NO  28671979  chrY28671979  26525832  \n",
       "311517            NO  28772221  chrY28772221  26626074  \n",
       "\n",
       "[311518 rows x 22 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "polyA_atlas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b6699751",
   "metadata": {},
   "source": [
    "## dist plotting the REDU for MAAPER outputs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "5eb16389",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "os.chdir('/data1/APA/Paul_ALS_Data/bams_in/subscelltype_bamfiles/Mapper_outs/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "43988242",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['AST-FB_sALSvsCTRL',\n",
       " 'IN-VIP_C9ALSvsCTRL',\n",
       " 'Oligodendrocytes_sALSvsCTRL',\n",
       " 'IN-SST_C9ALSvsCTRL',\n",
       " 'Excitatory_sALSvsCTRL',\n",
       " 'L2-3_C9ALSvsCTRL',\n",
       " 'AST-FB_C9ALSvsCTRL',\n",
       " 'L4_C9ALSvsCTRL',\n",
       " 'Inhibitory_sALSvsCTRL',\n",
       " 'L2-3_sALSvsCTRL',\n",
       " 'L5-6_C9ALSvsCTRL',\n",
       " 'Endothelial_ALSvsCTRL',\n",
       " 'AST-PP_sALSvsCTRL',\n",
       " 'L5-6-CC_sALSvsCTRL',\n",
       " 'L5-6-CC_C9ALSvsCTRL',\n",
       " 'L4_sALSvsCTRL',\n",
       " 'OPC_sALSvsCTRL',\n",
       " 'IN-VIP_sALSvsCTRL',\n",
       " 'IN-PV_C9ALSvsCTRL',\n",
       " 'IN-SST_sALSvsCTRL',\n",
       " 'AST-PP_C9ALSvsCTRL',\n",
       " 'IN-PV_sALSvsCTRL',\n",
       " 'Microglia_sALSvsCTRL',\n",
       " 'Excitatory_C9ALSvsCTRL',\n",
       " 'OPC_C9ALSvsCTRL',\n",
       " 'Oligodendrocytes_C9ALSvsCTRL',\n",
       " 'Microglia_C9ALSvsCTRL',\n",
       " 'Inhibitory_C9ALSvsCTRL']"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CTs = [ct for ct in os.listdir() if 'CTRL' in ct]\n",
    "CTs = list(set(CTs))\n",
    "CTs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "e060a3b6",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>npas</th>\n",
       "      <th>pval</th>\n",
       "      <th>RLDu</th>\n",
       "      <th>RLDi</th>\n",
       "      <th>REDu</th>\n",
       "      <th>REDu.pval</th>\n",
       "      <th>REDi</th>\n",
       "      <th>REDi.pval</th>\n",
       "      <th>nread.c1.1</th>\n",
       "      <th>nread.c1.2</th>\n",
       "      <th>nread.c1.3</th>\n",
       "      <th>nread.c1.4</th>\n",
       "      <th>nread.c1.5</th>\n",
       "      <th>nread.c2.1</th>\n",
       "      <th>nread.c2.2</th>\n",
       "      <th>nread.c2.3</th>\n",
       "      <th>nread.c2.4</th>\n",
       "      <th>nread.c2.5</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>B3GALT6</td>\n",
       "      <td>2</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>1.600</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.720</td>\n",
       "      <td>6.070000e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>72</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>63</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>PUSL1</td>\n",
       "      <td>4</td>\n",
       "      <td>0.130000</td>\n",
       "      <td>0.212</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.964</td>\n",
       "      <td>1.220000e-02</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>37</td>\n",
       "      <td>193</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>88</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>40</td>\n",
       "      <td>125</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>MIB2</td>\n",
       "      <td>2</td>\n",
       "      <td>0.003960</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.917</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.917</td>\n",
       "      <td>1.150000e-03</td>\n",
       "      <td>61</td>\n",
       "      <td>315</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>327</td>\n",
       "      <td>467</td>\n",
       "      <td>75</td>\n",
       "      <td>384</td>\n",
       "      <td>931</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>PRKCZ</td>\n",
       "      <td>4</td>\n",
       "      <td>0.003760</td>\n",
       "      <td>-0.334</td>\n",
       "      <td>0.641</td>\n",
       "      <td>-0.564</td>\n",
       "      <td>1.610000e-02</td>\n",
       "      <td>0.836</td>\n",
       "      <td>2.510000e-08</td>\n",
       "      <td>170</td>\n",
       "      <td>589</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>491</td>\n",
       "      <td>488</td>\n",
       "      <td>57</td>\n",
       "      <td>446</td>\n",
       "      <td>848</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>SKI</td>\n",
       "      <td>5</td>\n",
       "      <td>0.394000</td>\n",
       "      <td>-0.143</td>\n",
       "      <td>-0.228</td>\n",
       "      <td>-0.556</td>\n",
       "      <td>5.310000e-02</td>\n",
       "      <td>0.177</td>\n",
       "      <td>6.460000e-01</td>\n",
       "      <td>34</td>\n",
       "      <td>150</td>\n",
       "      <td>0</td>\n",
       "      <td>34</td>\n",
       "      <td>162</td>\n",
       "      <td>293</td>\n",
       "      <td>34</td>\n",
       "      <td>126</td>\n",
       "      <td>220</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9250</td>\n",
       "      <td>LAGE3</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>264</td>\n",
       "      <td>0</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>134</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9251</td>\n",
       "      <td>FAM3A</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>105</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>36</td>\n",
       "      <td>30</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>42</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9252</td>\n",
       "      <td>G6PD</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>31</td>\n",
       "      <td>45</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>58</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9253</td>\n",
       "      <td>MPP1</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>34</td>\n",
       "      <td>159</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>93</td>\n",
       "      <td>79</td>\n",
       "      <td>0</td>\n",
       "      <td>45</td>\n",
       "      <td>133</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>9254</td>\n",
       "      <td>RAB39B</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "      <td>92</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>79</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>9255 rows Ã— 19 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene  npas      pval   RLDu   RLDi   REDu     REDu.pval   REDi  \\\n",
       "0     B3GALT6     2  0.000002  1.600    NaN  2.720  6.070000e-07    NaN   \n",
       "1       PUSL1     4  0.130000  0.212    NaN  0.964  1.220000e-02    NaN   \n",
       "2        MIB2     2  0.003960    NaN  0.917    NaN           NaN  0.917   \n",
       "3       PRKCZ     4  0.003760 -0.334  0.641 -0.564  1.610000e-02  0.836   \n",
       "4         SKI     5  0.394000 -0.143 -0.228 -0.556  5.310000e-02  0.177   \n",
       "...       ...   ...       ...    ...    ...    ...           ...    ...   \n",
       "9250    LAGE3     1       NaN    NaN    NaN    NaN           NaN    NaN   \n",
       "9251    FAM3A     1       NaN    NaN    NaN    NaN           NaN    NaN   \n",
       "9252     G6PD     1       NaN    NaN    NaN    NaN           NaN    NaN   \n",
       "9253     MPP1     1       NaN    NaN    NaN    NaN           NaN    NaN   \n",
       "9254   RAB39B     1       NaN    NaN    NaN    NaN           NaN    NaN   \n",
       "\n",
       "         REDi.pval  nread.c1.1  nread.c1.2  nread.c1.3  nread.c1.4  \\\n",
       "0              NaN           0          72           0           0   \n",
       "1              NaN          37         193           0           0   \n",
       "2     1.150000e-03          61         315           0           0   \n",
       "3     2.510000e-08         170         589           0         105   \n",
       "4     6.460000e-01          34         150           0          34   \n",
       "...            ...         ...         ...         ...         ...   \n",
       "9250           NaN           0         264           0          49   \n",
       "9251           NaN           0         105           0           0   \n",
       "9252           NaN           0          45           0           0   \n",
       "9253           NaN          34         159           0           0   \n",
       "9254           NaN           0          92           0           0   \n",
       "\n",
       "      nread.c1.5  nread.c2.1  nread.c2.2  nread.c2.3  nread.c2.4  nread.c2.5  \n",
       "0              0           0           0           0          63           0  \n",
       "1             88          79           0          40         125           0  \n",
       "2            327         467          75         384         931           0  \n",
       "3            491         488          57         446         848          29  \n",
       "4            162         293          34         126         220           0  \n",
       "...          ...         ...         ...         ...         ...         ...  \n",
       "9250           0           0           0           0         134           0  \n",
       "9251          36          30           0          31          42           0  \n",
       "9252          31          45           0           4          58           0  \n",
       "9253          93          79           0          45         133           0  \n",
       "9254          79          81           0          68          41           0  \n",
       "\n",
       "[9255 rows x 19 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "exc_df = pd.read_csv('Excitatory_C9ALSvsCTRL/gene.txt', sep='\\t')\n",
    "exc_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "id": "390695ef",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_redu(df, name):\n",
    "    dat = df['REDu']\n",
    "    dat.replace([np.inf, -np.inf], np.nan, inplace=True)\n",
    "    dat.dropna(inplace=True)\n",
    "    ax = sns.distplot(dat)\n",
    "    ax.set(xlabel='REDu', ylabel='Density')\n",
    "    plt.axvline(dat.median(),\n",
    "                color='red', label=round(dat.median(),2))\n",
    "    plt.legend()\n",
    "    outname = \"REDU_plots/\" + name + '_REDu_distplots.png'\n",
    "    plt.savefig(outname)\n",
    "    plt.clf()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "d2178f24",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AST-FB_sALSvsCTRL\n",
      "(983, 11)\n",
      "IN-VIP_C9ALSvsCTRL\n",
      "(7381, 18)\n",
      "Oligodendrocytes_sALSvsCTRL\n",
      "(11532, 18)\n",
      "IN-SST_C9ALSvsCTRL\n",
      "(6432, 20)\n",
      "Excitatory_sALSvsCTRL\n",
      "(4928, 16)\n",
      "L2-3_C9ALSvsCTRL\n",
      "(8735, 13)\n",
      "AST-FB_C9ALSvsCTRL\n",
      "(1583, 14)\n",
      "L4_C9ALSvsCTRL\n",
      "(3968, 13)\n",
      "Inhibitory_sALSvsCTRL\n",
      "(8511, 18)\n",
      "L2-3_sALSvsCTRL\n",
      "(2183, 13)\n",
      "L5-6_C9ALSvsCTRL\n",
      "(413, 11)\n",
      "Endothelial_ALSvsCTRL\n",
      "(127, 11)\n",
      "AST-PP_sALSvsCTRL\n",
      "(11450, 16)\n",
      "L5-6-CC_sALSvsCTRL\n",
      "(106, 11)\n",
      "L5-6-CC_C9ALSvsCTRL\n",
      "(151, 11)\n",
      "L4_sALSvsCTRL\n",
      "(1698, 11)\n",
      "OPC_sALSvsCTRL\n",
      "(9013, 15)\n",
      "IN-VIP_sALSvsCTRL\n",
      "(2836, 14)\n",
      "IN-PV_C9ALSvsCTRL\n",
      "(5089, 19)\n",
      "IN-SST_sALSvsCTRL\n",
      "(5022, 16)\n",
      "AST-PP_C9ALSvsCTRL\n",
      "(11007, 21)\n",
      "IN-PV_sALSvsCTRL\n",
      "(4253, 14)\n",
      "Microglia_sALSvsCTRL\n",
      "(6241, 18)\n",
      "Excitatory_C9ALSvsCTRL\n",
      "(9255, 19)\n",
      "OPC_C9ALSvsCTRL\n",
      "(9774, 19)\n",
      "Oligodendrocytes_C9ALSvsCTRL\n",
      "(12012, 21)\n",
      "Microglia_C9ALSvsCTRL\n",
      "(6917, 21)\n",
      "Inhibitory_C9ALSvsCTRL\n",
      "(10400, 21)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<Figure size 432x288 with 0 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "for ct_cn in CTs:\n",
    "    file = ct_cn + \"/gene.txt\"\n",
    "    df = pd.read_csv(file, sep='\\t')\n",
    "    print(ct_cn)\n",
    "    print(df.shape)\n",
    "    plot_redu(df, ct_cn)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "id": "f190ce31",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "-0.8879999999999999"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dat.median()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ee124e2",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
